<!DOCTYPE html>
<html lang="en"><head>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.361">

  <meta name="author" content="Neelanjan Mukherjee">
  <title>MOLB 7950 - Bootcamp: Stats class 4</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  
  <script src="../../../site_libs/htmlwidgets-1.6.2/htmlwidgets.js"></script>
  <link href="../../../site_libs/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet">
  <script src="../../../site_libs/datatables-binding-0.28/datatables.js"></script>
  <script src="../../../site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
  <link href="../../../site_libs/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet">
  <link href="../../../site_libs/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet">
  <script src="../../../site_libs/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
  <link href="../../../site_libs/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet">
  <script src="../../../site_libs/crosstalk-1.2.0/js/crosstalk.min.js"></script>
<meta property="og:title" content="MOLB 7950 - Bootcamp: Stats class 4">
<meta property="og:site-name" content="MOLB 7950">
<meta name="twitter:title" content="MOLB 7950 - Bootcamp: Stats class 4">
<meta name="twitter:image" content="https://rnabioco.github.io/molb-7950/content/bootcamp/stats/images/twitter-card.png">
<meta name="twitter:creator" content="@rnabioco">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Bootcamp: Stats class 4</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
<a href="neelanjan.mukherjee@cuanschutz.edu">Neelanjan Mukherjee</a> 
</div>
</div>
</div>

  <p class="date">Invalid Date</p>
</section>
<section class="slide level2">

</section>
<section>
<section id="learning-objectives" class="title-slide slide level1 center">
<h1>Learning Objectives</h1>
<ul>
<li><strong>Visualize</strong> and <strong>Summarize</strong> the data being compared</li>
<li><strong>Formulate</strong> and <strong>Execute</strong> null hypothesis testing</li>
<li><strong>Identify</strong> and <strong>Perform</strong> the proper statistical test for data type/comparison</li>
<li><strong>Calculate</strong> and <strong>Interpret</strong> p-values</li>
<li><strong>Prevent</strong> p-hacking and <strong>Recognize</strong> issues with simultaneously testing multiple hypotheses.</li>
</ul>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="outline" class="title-slide slide level1 center">
<h1>Outline</h1>
<ul>
<li>Concepts and Definitions
<ul>
<li>Simplifying principles: Common tests as linear models</li>
<li>Types of comparisons and statistical tests</li>
<li>Definitions<br>
</li>
</ul></li>
<li>Day 1: Relationship between two or more continuous variables
<ul>
<li>Correlation vs Regression</li>
<li>Fitting it a line to data</li>
<li>Linear regression concepts</li>
<li>Multiple regression<br>
</li>
</ul></li>
<li><strong>Day 2: Relationship between categorical and continuous variables</strong>
<ul>
<li><p><strong>Comparing means between two groups (t-Test)</strong></p></li>
<li><p><strong>Comparing means between 3 or more groups (ANOVA)<br>
</strong></p></li>
</ul></li>
<li>Day 3: Multiple test correction, Bayesian intro, History</li>
</ul>
</section>
<section id="provide-a-simple-and-flexible-framework" class="slide level2">
<h2>Provide a <strong>simple</strong> and <strong>flexible</strong> framework</h2>

<img data-src="img/easy_hard.jpg" class="r-stretch"></section>
<section class="slide level2">

</section></section>
<section>
<section id="cheatsheet" class="title-slide slide level1 center">
<h1>CHEATSHEET</h1>

<img data-src="img/linear_tests_cheat_sheet_adapted.png" class="r-stretch"></section>
<section id="variables-definitions" class="slide level2">
<h2>Variables definitions</h2>
<h3 id="random-variables-x-y">Random variables (x, y)</h3>
<p><strong>Response Variable</strong> ( <strong>y</strong> - aka dependent or outcome variable): this variable is predicted or its variation is explained by the explanatory variable. In an experiment, this is the outcome that is measured following manipulation of the explanatory variable.</p>
<p><strong>Explanatory Variable</strong> ( <strong>x</strong> - aka independent or predictor variable): explains variations in the response variable. In an experiment, it is manipulated by the researcher.</p>
<h3 id="quantitative-variables">Quantitative Variables</h3>
<p><strong>Discrete variable</strong>: numeric variables that have a countable number of values between any two values - <code>integer</code> in R (e.g., number of mice, read counts).</p>
<p><strong>Continuous variable</strong>: numeric variables that have an infinite number of values between any two values - <code>numeric</code> in R (e.g., normalized expression values, fluorescent intensity).</p>
<h3 id="categorical-variables">Categorical Variables</h3>
<p><strong>Nominal variable</strong>: (unordered) random variables have categories where order doesn’t matter - <code>factor</code> in R (e.g., country, type of gene, genotype).</p>
<p><strong>Ordinal variable</strong>: (ordered) random variables have ordered categories - order of <code>levels</code> in R ( e.g.&nbsp;grade of tumor).</p>
</section>
<section id="hypothesis-testing-definitions" class="slide level2">
<h2>Hypothesis testing definitions</h2>
<p><strong>Hypothesis testing</strong> is a statistical analysis that uses sample data to assess two mutually exclusive theories about the properties of a population. Statisticians call these theories the null hypothesis and the alternative hypothesis. A hypothesis test assesses your sample statistic and factors in an estimate of the sample error to determine which hypothesis the data support.</p>
<p>When you can reject the null hypothesis, the results are statistically significant, and your data support the theory that an effect exists at the population level.</p>
<p><a href="https://www.graphpad.com/guides/prism/latest/statistics/hypothesis_testing_and_statistical_significance.htm"><strong>A legal analogy: Guilty or not guilty?</strong></a><strong><br>
</strong>The statistical concept of ‘significant’ vs.&nbsp;‘not significant’ can be understood by comparing to the legal concept of ‘guilty’ vs.&nbsp;‘not guilty’.</p>
<p>In the American legal system (and much of the world) a criminal defendant is presumed innocent until proven guilty. If the evidence proves the defendant guilty beyond a reasonable doubt, the verdict is ‘guilty’. Otherwise the verdict is ‘not guilty’. In some countries, this verdict is ‘not proven’, which is a better description. A ‘not guilty’ verdict does not mean the judge or jury concluded that the defendant is innocent -- it just means that the evidence was not strong enough to persuade the judge or jury that the defendant was guilty.</p>
<p>In statistical hypothesis testing, you start with the null hypothesis (usually that there is no difference between groups). If the evidence produces a small enough P value, you reject that null hypothesis, and conclude that the difference is real. If the P value is higher than your threshold (usually 0.05), you don’t reject the null hypothesis. This doesn’t mean the evidence convinced you that the treatment had no effect, only that the evidence was not persuasive enough to convince you that there is an effect.</p>
<p><strong>Effect</strong> — the difference between the population value and the null hypothesis value. The effect is also known as population effect or the difference. Typically, you do not know the size of the actual effect. However, you can use a hypothesis test to help you determine whether an effect exists and to estimate its size.</p>
<p><strong>Null Hypothesis</strong> or <span class="math inline">\(\mathcal{H}_0\)</span> — one of two mutually exclusive theories about the properties of the population in hypothesis testing. Typically, the null hypothesis states that there is no effect (i.e., the effect size equals zero).</p>
<p><strong>Alternative Hypothesis</strong> or <span class="math inline">\(\mathcal{H}_1\)</span> — the other theory about the properties of the population in hypothesis testing. Typically, the alternative hypothesis states that a population parameter does not equal the null hypothesis value. In other words, there is a non-zero effect. If your sample contains sufficient evidence, you can reject the null and favor the alternative hypothesis.</p>
<p><strong>P-values</strong> — the probability of obtaining test results at least as extreme as the results actually observed, under the assumption that the null hypothesis is correct. Lower p-values represent stronger evidence against the null. P-values in conjunction with the significance level determines whether your data favor the null or alternative hypothesis.</p>
<p><a href="https://www.youtube.com/watch?v=5Z9OIYA8He8">StatQuest: P Values, clearly explained</a></p>
<p><a href="https://www.youtube.com/watch?v=JQc3yx0-Q9E">StatQuest: How to calculate p-values</a></p>
<p><strong>Significance Level</strong> or <span class="math inline">\(a\)</span> — an evidentiary standard set before the study. It is the probability that you say there is an effect when there is no effect (the probability of rejecting the null hypothesis given that it is true). Lower significance levels indicate that you require stronger evidence before you will reject the null.It is usually set at or below .05.</p>

<img data-src="https://upload.wikimedia.org/wikipedia/commons/3/35/Guinness_Glass_2010.jpg" style="width:10.0%" class="r-stretch quarto-figure-center"><p class="caption">Guinness</p></section>
<section id="null-hypothesis-testing" class="slide level2">
<h2>Null hypothesis testing</h2>
<ol type="1">
<li>Specify the variables</li>
<li>Declare null hypothesis <span class="math inline">\(\mathcal{H}_0\)</span></li>
<li>Calculate test-statistic, exact p-value</li>
<li><em>Generate and visualize data reflecting null-distribution</em></li>
<li><em>Calculate the p-value from the test statistic and null distribution</em></li>
</ol>
<p>*4-5: For calculating empirical p-value</p>
</section>
<section id="the-simplicity-underlying-common-tests" class="slide level2">
<h2>The simplicity underlying common tests</h2>
<p>Most of the common statistical models (t-test, correlation, ANOVA; chi-square, etc.) are special cases of linear models or a very close approximation. This simplicity means that there is less to learn. In particular, it all comes down to:<br>
<span class="math inline">\(y = a \cdot x + b\)</span></p>
<p>This needless complexity multiplies when students try to rote learn the parametric assumptions underlying each test separately rather than deducing them from the linear model.</p>
</section>
<section id="parametric-vs-non-parametric-tests" class="slide level2">
<h2>Parametric vs Non-Parametric tests</h2>
<p><strong>Parametric tests</strong> are suitable for normally distributed data.</p>
<p><strong>Non-Parametric tests</strong> are suitable for any continuous data. For the sake of simplicity and sticking with a consistent framework, we will consider Non-Parametric tests as the <strong>ranked versions of the corresponding parametric tests</strong>.</p>
<p><a href="https://statisticsbyjim.com/hypothesis-testing/nonparametric-parametric-tests/">More on choosing Parametric vs Non-Parametric</a></p>
<div class="cell">
<div class="cell-output-display">
<table>
<thead>
<tr class="header">
<th style="text-align: left;">Info</th>
<th style="text-align: left;">Parametric</th>
<th style="text-align: left;">Non-Parametric</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">better descriptor</td>
<td style="text-align: left;">mean</td>
<td style="text-align: left;">median</td>
</tr>
<tr class="even">
<td style="text-align: left;"># of samples (N)</td>
<td style="text-align: left;">many</td>
<td style="text-align: left;">few</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="import-and-tidy-data" class="title-slide slide level1 center">
<h1>Import and tidy data</h1>
<p>We will be using mouse data from <a href="https://wp.cs.ucl.ac.uk/outbredmice/">Resources for Outbred Mice</a>. The goal of the study was to establish genotype-phenotype relationships for highly recombinant outbred mouse populations. We will be using the phenotypic data for our exercises.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># we are reading the data directly from the internet</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>biochem <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"http://mtweb.cs.ucl.ac.uk/HSMICE/PHENOTYPES/Biochemistry.txt"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-3"><a href="#cb1-3"></a>  janitor<span class="sc">::</span><span class="fu">clean_names</span>()</span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co"># simplify names a bit more</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">colnames</span>(biochem) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"biochem_"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="fu">colnames</span>(biochem))</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co"># we are going to simplify this a bit and only keep some columns</span></span>
<span id="cb1-9"><a href="#cb1-9"></a>keep <span class="ot">&lt;-</span> <span class="fu">colnames</span>(biochem)[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">9</span>,<span class="dv">14</span>,<span class="dv">15</span>,<span class="dv">24</span><span class="sc">:</span><span class="dv">28</span>)]</span>
<span id="cb1-10"><a href="#cb1-10"></a>biochem <span class="ot">&lt;-</span> biochem[,keep]</span>
<span id="cb1-11"><a href="#cb1-11"></a></span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="co"># get weights for each individual mouse</span></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="co"># careful: did not come with column names</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>weight <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"http://mtweb.cs.ucl.ac.uk/HSMICE/PHENOTYPES/weight"</span>, <span class="at">col_names =</span> F, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-15"><a href="#cb1-15"></a></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="co"># add column names</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="fu">colnames</span>(weight) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"subject_name"</span>,<span class="st">"weight"</span>)</span>
<span id="cb1-18"><a href="#cb1-18"></a></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="co"># add weight to biochem table and get rid of NAs</span></span>
<span id="cb1-20"><a href="#cb1-20"></a>biochem <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(biochem, weight, <span class="at">by=</span><span class="st">"subject_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-21"><a href="#cb1-21"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb1-22"><a href="#cb1-22"></a></span>
<span id="cb1-23"><a href="#cb1-23"></a></span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="co"># explore the data a bit</span></span>
<span id="cb1-25"><a href="#cb1-25"></a><span class="co"># colnames(biochem)</span></span>
<span id="cb1-26"><a href="#cb1-26"></a><span class="co"># str(biochem)</span></span>
<span id="cb1-27"><a href="#cb1-27"></a><span class="co"># View(biochem)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="review-linear-regression-from-yesterday" class="title-slide slide level1 center">
<h1>Review linear regression from yesterday</h1>

</section>
<section id="equation-for-a-line-stats-version" class="slide level2">
<h2>Equation for a line (Stats version)</h2>
<p>Model: the recipe for <span class="math inline">\(y\)</span> is a slope (<span class="math inline">\(\beta_1\)</span>) times <span class="math inline">\(x\)</span> plus an intercept (<span class="math inline">\(\beta_0\)</span>, aka a straight line).</p>
<p><span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x \qquad \mathcal{H}_0: \beta_1 = 0 \qquad y = \beta_0 \cdot 1\)</span></p>
<p><span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x \qquad \mathcal{H}_0: \beta_1 \neq 0\)</span></p>
</section>
<section id="find-the-best-beta-coefficients" class="slide level2">
<h2>Find the best <span class="math inline">\(\beta\)</span> coefficients</h2>
<p>These <span class="math inline">\(\beta\)</span> coefficients are also called the paramters of the model. The <span class="math inline">\(\beta\)</span> coefficients returned are for the lineear model that best fits the data.</p>
</section>
<section id="specify-variables-and-hypothesis" class="slide level2">
<h2>Specify variables and hypothesis</h2>
<p>Remember: <span class="math inline">\(y = \beta_0 \cdot 1+ \beta_1 \cdot x\)</span></p>
<p>Null Hypothesis: <span class="math inline">\(\mathcal{H}_0: \beta_1 = 0\)</span></p>
<p><span class="math inline">\(\mathcal{H}_0:\)</span> mouse <span class="math inline">\(cholesterol\)</span> does NOT explain <span class="math inline">\(weight\)</span></p>
<p>Alternative Hypothesis: <span class="math inline">\(\mathcal{H}_1: \beta_1 \neq 0\)</span></p>
<p>Simple model: <span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x\)</span></p>
<p><span class="math inline">\(\mathcal{H}_1:\)</span> mouse <span class="math inline">\(cholesterol\)</span> does explain <span class="math inline">\(weight\)</span></p>
</section>
<section id="relationship-between-mouse-weight-and-cholesterol" class="slide level2">
<h2>Relationship between mouse weight and cholesterol</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co"># fitting a line weight vs intercept (mean weight)</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>fit_W <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> , <span class="at">data =</span> biochem)</span>
<span id="cb2-3"><a href="#cb2-3"></a></span>
<span id="cb2-4"><a href="#cb2-4"></a><span class="co"># augment data to add fit/residuals</span></span>
<span id="cb2-5"><a href="#cb2-5"></a>biochem_W <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_W, <span class="at">data =</span> biochem)</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co">#plot data</span></span>
<span id="cb2-8"><a href="#cb2-8"></a>p_wch <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> tot_cholesterol)) <span class="sc">+</span></span>
<span id="cb2-9"><a href="#cb2-9"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb2-10"><a href="#cb2-10"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm, <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb2-11"><a href="#cb2-11"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb2-12"><a href="#cb2-12"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb2-13"><a href="#cb2-13"></a></span>
<span id="cb2-14"><a href="#cb2-14"></a>p_WvInt_res <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem_W, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb2-15"><a href="#cb2-15"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> biochem_W<span class="sc">$</span>.fitted, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span> <span class="co"># plot linear model fit</span></span>
<span id="cb2-16"><a href="#cb2-16"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> <span class="co"># plot height as points and color code by the value of the residual</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb2-18"><a href="#cb2-18"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">25</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb2-19"><a href="#cb2-19"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a><span class="co"># fitting a line weight vs icholesterol</span></span>
<span id="cb2-22"><a href="#cb2-22"></a>fit_WvC <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">data =</span> biochem,</span>
<span id="cb2-23"><a href="#cb2-23"></a>              <span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> tot_cholesterol)</span>
<span id="cb2-24"><a href="#cb2-24"></a></span>
<span id="cb2-25"><a href="#cb2-25"></a><span class="co"># augment data to add fit/residuals</span></span>
<span id="cb2-26"><a href="#cb2-26"></a>biochem_WvC <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_WvC, <span class="at">data =</span> biochem)</span>
<span id="cb2-27"><a href="#cb2-27"></a></span>
<span id="cb2-28"><a href="#cb2-28"></a><span class="co"># plot data</span></span>
<span id="cb2-29"><a href="#cb2-29"></a>p_WvC_res <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb2-30"><a href="#cb2-30"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span></span>
<span id="cb2-31"><a href="#cb2-31"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm, <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb2-32"><a href="#cb2-32"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code</span></span>
<span id="cb2-33"><a href="#cb2-33"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">1</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb2-34"><a href="#cb2-34"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb2-35"><a href="#cb2-35"></a></span>
<span id="cb2-36"><a href="#cb2-36"></a></span>
<span id="cb2-37"><a href="#cb2-37"></a></span>
<span id="cb2-38"><a href="#cb2-38"></a>biochem_WvC_rsq <span class="ot">&lt;-</span> fit_WvC <span class="sc">%&gt;%</span> <span class="fu">glance</span>() <span class="sc">%&gt;%</span> <span class="fu">pull</span>(r.squared)</span>
<span id="cb2-39"><a href="#cb2-39"></a></span>
<span id="cb2-40"><a href="#cb2-40"></a>biochem_WvC_pval <span class="ot">&lt;-</span> fit_WvC <span class="sc">%&gt;%</span> <span class="fu">glance</span>() <span class="sc">%&gt;%</span> <span class="fu">pull</span>(p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">plot_grid</span>(p_WvInt_res, p_WvC_res, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"weight by intercept"</span>,<span class="st">"weight by cholesterol"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="class-04_files/figure-revealjs/review plot-1.png" width="960" class="r-stretch"></section>
<section class="slide level2">

<h3 id="to-what-extent-does-mouse-cholesterol-predict-mouse-weight">To what extent does mouse cholesterol predict mouse weight?</h3>
<p><span class="math inline">\(R^2\)</span> or coefficient of determination — the proportion of the variance in the dependent variable that is predictable from the independent variable(s).</p>
<p>13%</p>
<h3 id="probability-that-relationship-is-due-to-chance">Probability that relationship is due to chance?</h3>
<p><span class="math inline">\(p-value\)</span> — the probability of obtaining an <span class="math inline">\(F-statistic\)</span> in the null distribution at least as extreme as our observed <span class="math inline">\(F-statistic\)</span>.</p>
<p>8.9241551^{-54}</p>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="comparing-means-between-two-groups-students-t-test" class="title-slide slide level1 center">
<h1>Comparing means between two groups (Student’s t-test)</h1>
<p>The T-Distribution, also known as Student’s t-distribution, gets its name from William Sealy Gosset who first published it in English in 1908 in the scientific journal Biometrika using his pseudonym “Student”[9][10] because his employer preferred staff to use pen names when publishing scientific papers instead of their real name, so he used the name “Student” to hide his identity.[11]</p>
<p><a href="https://medium.com/value-stream-design/the-curious-tale-of-william-sealy-gosset-b3178a9f6ac8">The Curious Tale of William Sealy Gosset</a></p>
</section>
<section id="guinness-brewery-in-dublin" class="slide level2">
<h2>Guinness Brewery in Dublin</h2>
<p><img data-src="img/gosset.jpg" style="width:19.0%"> <img data-src="img/inara_guinness.jpg" style="width:40.0%"></p>
<p>We will compare mouse <span class="math inline">\(weight\)</span> by <span class="math inline">\(gender\)</span>.</p>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="class-04_files/figure-revealjs/unnamed-chunk-2-1.png" width="960"></p>
</div>
</div>
</section>
<section id="specify-variables-and-hypothesis-1" class="slide level2">
<h2>Specify variables and hypothesis</h2>
<p>Model: <span class="math inline">\(y_{i} = \beta_0 \cdot 1+ \beta_1 \cdot x_{i}\)</span></p>
<p>Null Hypothesis: <span class="math inline">\(\mathcal{H}_0: \beta_1 = 0\)</span></p>
<p><span class="math inline">\(\mathcal{H}_0:\)</span> mouse <span class="math inline">\(gender\)</span> does NOT explain <span class="math inline">\(weight\)</span></p>
<p>Alternative Hypothesis: <span class="math inline">\(\mathcal{H}_1: \beta_1 \neq 0\)</span></p>
<p><span class="math inline">\(\mathcal{H}_1:\)</span> mouse <span class="math inline">\(gender\)</span> does explain <span class="math inline">\(weight\)</span></p>
<p><strong>Important:</strong> <span class="math inline">\(x_{i}\)</span> is an indicator (0 or 1) saying whether data point i was sampled from one or the other group (female or male).</p>
<p>We will explore this in more detail soon.</p>
</section>
<section id="calculate-ss_mean-for-weight-by-gender-female-vs-male" class="slide level2">
<h2>1. Calculate <span class="math inline">\(SS_{mean}\)</span> for weight by gender (female vs male)</h2>
<h3 id="compare-ss_mean-for-weight-by-cholesterol-versus-weight-by-gender">Compare <span class="math inline">\(SS_{mean}\)</span> for weight by cholesterol <em>versus</em> weight by gender</h3>
<p><span class="math inline">\(SS_{mean}\)</span> — sum of squares around the overall mean of <span class="math inline">\(y\)</span></p>
<p><span class="math inline">\(SS_{mean} = \sum_{i=1}^{n} (data - mean)^2\)</span></p>
<p><span class="math inline">\(SS_{mean} = \sum_{i=1}^{n} (y_{i} - \overline{y})^2\)</span></p>
<p><br></p>
<p><strong>Residuals</strong>, <span class="math inline">\(e\)</span> — the difference between the observed value of the dependent variable <span class="math inline">\(y\)</span> and the predicted value <span class="math inline">\(\widehat{y}\)</span> is called the residual. Each data point has one residual.</p>
<p><span class="math inline">\(e = y_{i} - \widehat{y}\)</span></p>
<h3 id="class-exercise-1"><strong>Class exercise 1</strong>:</h3>
<p>Which of these are valid ways to calculate <span class="math inline">\(SS_{mean}\)</span> from <code>biochem_W</code>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="co"># A. sum((biochem_W$weight - biochem_W$.fitted)^2)  </span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co"># B. sum((biochem_W$weight - biochem_W$.resid)^2)    </span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co"># C. sum(biochem_W$.resid^2)  </span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co"># D. sum((biochem_W$weight - mean(biochem_W$weight))^2)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="calculate-ss_fit-for-weight-by-gender" class="slide level2">
<h2>2. Calculate <span class="math inline">\(SS_{fit}\)</span> for weight by gender</h2>
<h3 id="compare-ss_fit-vs-ss_fit-weight-by-gender">Compare <span class="math inline">\(SS_{fit}\)</span> vs <span class="math inline">\(SS_{fit}\)</span> weight by gender</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># fitting a line weight vs intercept + gender</span></span>
<span id="cb5-2"><a href="#cb5-2"></a>fit_WvG <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> gender , <span class="at">data =</span> biochem)</span>
<span id="cb5-3"><a href="#cb5-3"></a></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co"># augment (i.e. add fitted and residual values)</span></span>
<span id="cb5-5"><a href="#cb5-5"></a>biochem_WvG <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_WvG, <span class="at">data =</span> biochem)</span>
<span id="cb5-6"><a href="#cb5-6"></a></span>
<span id="cb5-7"><a href="#cb5-7"></a><span class="co"># plot of data with mean and colored by residuals</span></span>
<span id="cb5-8"><a href="#cb5-8"></a>p_WvG_res <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(biochem_WvG, <span class="fu">aes</span>(<span class="at">x =</span> gender, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb5-9"><a href="#cb5-9"></a>  <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(), <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span></span>
<span id="cb5-10"><a href="#cb5-10"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb5-11"><a href="#cb5-11"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span>.<span class="dv">5</span>, <span class="at">xend=</span><span class="fl">1.5</span>, <span class="at">y=</span>fit_WvG<span class="sc">$</span>coefficients[<span class="dv">1</span>], <span class="at">yend=</span>fit_WvG<span class="sc">$</span>coefficients[<span class="dv">1</span>]), <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb5-12"><a href="#cb5-12"></a>    <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fl">1.5</span>, <span class="at">xend=</span><span class="fl">2.5</span>, <span class="at">y=</span><span class="fu">sum</span>(fit_WvG<span class="sc">$</span>coefficients)), <span class="at">yend=</span><span class="fu">sum</span>(fit_WvG<span class="sc">$</span>coefficients), <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb5-13"><a href="#cb5-13"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<p><img data-src="class-04_files/figure-revealjs/unnamed-chunk-5-1.png" width="960"></p>
</div>
</div>
<h3 id="note-we-are-fitting-2-lines-to-the-data">NOTE: We are fitting 2 lines to the data</h3>
<p>For the weight by intercept model (right) we fit <strong>1</strong> line.<br>
For the weight by gender model (left) we fit <strong>2</strong> lines (i.e.&nbsp;male and female).</p>
<h3 id="exceptions-to-the-fit">Exceptions to the fit</h3>
<p>Check the residuals!</p>
<h3 id="matrices-interlude-begin"><em>Matrices Interlude Begin</em></h3>
<h4 id="how-do-we-go-from-2-fit-lines-to-1-equation">How do we go from <strong>2 fit lines</strong> to <strong>1 equation</strong></h4>
<p>Since we don’t want to calculate any of this by hand, the framework needs to be flexible such that a computer can execute for different flavors of comparison (cont y vs cont x, cont y vs 2 or more categorical x, …).</p>
<p>Let’s break this down and focus on just a few players.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>p_exc</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="class-04_files/figure-revealjs/unnamed-chunk-7-1.png" width="768"></p>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="co"># ggplot(data = biochem_WvG, aes(.resid, color=gender)) +</span></span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">#   geom_density() + </span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">#   theme_minimal()</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Remember that:<br>
<span class="math inline">\(weight\)</span> is <span class="math inline">\(y\)</span><br>
<span class="math inline">\(F_{avg}\)</span> is the average <span class="math inline">\(weight\)</span> of <span class="math inline">\(females\)</span><br>
<span class="math inline">\(M_{avg}\)</span> is the average <span class="math inline">\(weight\)</span> of <span class="math inline">\(males\)</span></p>
</section>
<section class="slide level2">

<p>A0480548<strong>85</strong>, female<br>
<span class="math inline">\(y_{85}= 0 \cdot F_{avg} + 1 \cdot M_{avg} + residual_{85}\)</span></p>
<p>A0671097<strong>71</strong>, female<br>
<span class="math inline">\(y_{71}= 0 \cdot F_{avg} + 1 \cdot M_{avg} + residual_{71}\)</span></p>
</section>
<section class="slide level2">

<p>A0668223<strong>51</strong>, male<br>
<span class="math inline">\(y_{51}= 0 \cdot F_{avg} + 1 \cdot M_{avg} + residual_{51}\)</span></p>
<p>A0482743<strong>62</strong>, male<br>
<span class="math inline">\(y_{62}= 0 \cdot F_{avg} + 1 \cdot M_{avg} + residual_{62}\)</span></p>
</section>
<section class="slide level2">

<div class="cell">
<div class="cell-output-display">

<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-66e8ddcc229d16dae9ee" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-66e8ddcc229d16dae9ee">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4"],["A048054885","A067109771","A048274362","A066822351"],[24,25.5,13.4,13.2],["F","F","M","M"],[18.07586980920314,18.07586980920314,22.40594837261504,22.40594837261504],[5.924130190796859,7.424130190796859,-9.005948372615036,-9.205948372615037]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>subject_name<\/th>\n      <th>weight<\/th>\n      <th>gender<\/th>\n      <th>.fitted<\/th>\n      <th>.resid<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"columnDefs":[{"className":"dt-right","targets":[2,4,5]},{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
<section id="need-a-volunteer" class="slide level2">
<h2>Need a volunteer</h2>
<p><strong>Me:</strong> Ooohh my, imagine how tedious it would be to do this for all 1782 mice…<br>
<strong>Volunteer:</strong> Wait a sec…isn’t there a way to formulate this as a matrix algebra problem.<br>
<strong>Me:</strong> You’re right - I’m so glad you asked! Let’s wield our matrix-magic at this problem and see what happens.</p>
<p><span class="math inline">\(f_{avg} = \beta_0\)</span> is the average <span class="math inline">\(weight\)</span> of <span class="math inline">\(female\)</span> mice<br>
<span class="math inline">\(m_{avg} = \beta_1\)</span> is the average <span class="math inline">\(weight\)</span> of <span class="math inline">\(male\)</span> mice</p>
<p><span class="math inline">\(\begin{bmatrix} y_{85} \\ y_{71} \\ y_{51} \\y_{62} \end{bmatrix} = \begin{bmatrix} 0 &amp; 1 \\ 1 &amp; 0 \\ 0 &amp; 1 \\ 0 &amp; 1 \end{bmatrix} \cdot \begin{bmatrix} \beta_0 \\ \beta_1 \end{bmatrix} + \begin{bmatrix} e_{85} \\ e_{71} \\ e_{51} \\e_{62} \end{bmatrix}\)</span></p>
<p><strong>So basically this looks like the same equation for fitting a line we’ve been discussing, just w/a few more dimensions :)</strong></p>
<p>This is a conceptual peak into the underbelly of how the <span class="math inline">\(\beta\)</span> cofficients and least squares is performed using matrix operations (remember linear algebra, maybe?). We will not go any deeper in this course, but if you are interested in learning more I recommend the following resources:<br>
</p>
<p><a href="https://www.youtube.com/watch?v=CqLGvwi-5Pc&amp;list=PLblh5JKOoLUIzaEkCLIUxQFjPIlapw8nU&amp;index=6">Linear Models Pt.3 - Design Matrices</a><br>
</p>
<p><a href="https://online.stat.psu.edu/stat462/node/132/">A Matrix Formulation of the Multiple Regression Model</a></p>
<h3 id="matrices-interlude-fin"><em>Matrices Interlude FIN</em></h3>
<h3 id="class-exercise-2"><strong>Class exercise 2</strong>:</h3>
<p>Which of these are valid ways to calculate <span class="math inline">\(SS_{mean}\)</span> from <code>biochem_W</code>?</p>
<p><span class="math inline">\(SS_{fit}\)</span> — sum of squares around the least-squares fit</p>
<p><span class="math inline">\(SS_{fit} = \sum_{i=1}^{n} (data - line)^2\)</span></p>
<p>Which of these are valid ways to calculate <span class="math inline">\(SS_{fit}\)</span> from <code>biochem_WvG</code>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a><span class="co"># A. sum(biochem_WvG$.resid^2)  </span></span>
<span id="cb8-2"><a href="#cb8-2"></a><span class="co"># B. sum((biochem_WvG$weight - biochem_WvG$.resid)^2)    </span></span>
<span id="cb8-3"><a href="#cb8-3"></a><span class="co"># C. sum((biochem_WvG$weight - biochem_WvG$.fitted)^2)  </span></span>
<span id="cb8-4"><a href="#cb8-4"></a><span class="co"># D. sum((biochem_WvG$weight - mean(biochem_WvG$weight))^2)</span></span>
<span id="cb8-5"><a href="#cb8-5"></a></span>
<span id="cb8-6"><a href="#cb8-6"></a>p_WvG_res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="class-04_files/figure-revealjs/exercise 2-1.png" width="960" class="r-stretch"></section>
<section id="calculate-f-statistic" class="slide level2">
<h2>3. Calculate <span class="math inline">\(F-statistic\)</span></h2>
<p><strong>F-statistic</strong> — the proportion of the variance in the dependent variable that is predictable from the independent variable(s).</p>
<p><span class="math inline">\(F = \displaystyle \frac{SS_{fit}/(p_{fit}-p_{mean})} {SS_{mean}/(n-p_{fit})}\)</span></p>
<p><span class="math inline">\(p_{fit}\)</span> — number of parameters in the fit line<br>
<span class="math inline">\(p_{mean}\)</span> — number of parameters in the mean line<br>
<span class="math inline">\(n\)</span> — number of data points</p>
</section>
<section id="sanity-check" class="slide level2">
<h2>4. Sanity check</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># sum of squares of the residuals for the simple model</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>ss.mean <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_W<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb9-3"><a href="#cb9-3"></a></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co"># sum of squares of the residuals for the complex model</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>ss.fit <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_WvG<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb9-6"><a href="#cb9-6"></a></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="do">#### COEFFICIENTS NEEL</span></span>
<span id="cb9-8"><a href="#cb9-8"></a></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co"># number of paramters in simple model</span></span>
<span id="cb9-10"><a href="#cb9-10"></a>pmean <span class="ot">&lt;-</span> <span class="dv">1</span> </span>
<span id="cb9-11"><a href="#cb9-11"></a></span>
<span id="cb9-12"><a href="#cb9-12"></a><span class="co"># number of paramters in complex model</span></span>
<span id="cb9-13"><a href="#cb9-13"></a>pfit <span class="ot">&lt;-</span> <span class="dv">2</span></span>
<span id="cb9-14"><a href="#cb9-14"></a></span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="co"># F-value</span></span>
<span id="cb9-16"><a href="#cb9-16"></a>biochem_WvG_F <span class="ot">&lt;-</span> ((ss.mean <span class="sc">-</span> ss.fit) <span class="sc">/</span> (pfit <span class="sc">-</span> <span class="dv">1</span>)) <span class="sc">/</span> </span>
<span id="cb9-17"><a href="#cb9-17"></a>  (ss.fit <span class="sc">/</span> (<span class="fu">nrow</span>(biochem_WvG) <span class="sc">-</span> pfit))</span>
<span id="cb9-18"><a href="#cb9-18"></a></span>
<span id="cb9-19"><a href="#cb9-19"></a>biochem_WvG_F </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1296.507</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu">glance</span>(fit_WvG) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(statistic)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   value 
1296.507 </code></pre>
</div>
</div>
</section>
<section class="slide level2">

<p>Let’s take a look at the statistic and p-values:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a>fit_WvG_stats <span class="ot">&lt;-</span> <span class="fu">tidy</span>(fit_WvG) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(term<span class="sc">==</span><span class="st">"genderM"</span>) <span class="sc">%&gt;%</span> <span class="fu">select</span>(statistic, p.value)</span>
<span id="cb13-2"><a href="#cb13-2"></a></span>
<span id="cb13-3"><a href="#cb13-3"></a></span>
<span id="cb13-4"><a href="#cb13-4"></a><span class="co"># to run a t.test in R we need numeric vectors for each of our groups of interest</span></span>
<span id="cb13-5"><a href="#cb13-5"></a>male_weight_t <span class="ot">&lt;-</span> biochem_WvG <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gender<span class="sc">==</span><span class="st">"M"</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(weight)</span>
<span id="cb13-6"><a href="#cb13-6"></a>  </span>
<span id="cb13-7"><a href="#cb13-7"></a>female_weight_t <span class="ot">&lt;-</span> biochem_WvG <span class="sc">%&gt;%</span> <span class="fu">filter</span>(gender<span class="sc">==</span><span class="st">"F"</span>) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(weight)</span>
<span id="cb13-8"><a href="#cb13-8"></a></span>
<span id="cb13-9"><a href="#cb13-9"></a></span>
<span id="cb13-10"><a href="#cb13-10"></a>trad_t <span class="ot">&lt;-</span> <span class="fu">t.test</span>(male_weight_t, female_weight_t, <span class="at">var.equal =</span> T)</span>
<span id="cb13-11"><a href="#cb13-11"></a></span>
<span id="cb13-12"><a href="#cb13-12"></a>trad_WvG_stats <span class="ot">&lt;-</span> <span class="fu">tidy</span>(trad_t) <span class="sc">%&gt;%</span> <span class="fu">select</span>(statistic, p.value)</span>
<span id="cb13-13"><a href="#cb13-13"></a></span>
<span id="cb13-14"><a href="#cb13-14"></a>fit_WvG_stats</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 2
  statistic   p.value
      &lt;dbl&gt;     &lt;dbl&gt;
1      36.0 9.25e-214</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a>trad_WvG_stats</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 2
  statistic   p.value
      &lt;dbl&gt;     &lt;dbl&gt;
1      36.0 9.25e-214</code></pre>
</div>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="comparing-means-of-3-or-more" class="title-slide slide level1 center">
<h1>Comparing means of 3 or more?</h1>

</section>
<section id="one-way-anova" class="slide level2">
<h2>One-way ANOVA</h2>
<p>Let’s compare the <span class="math inline">\(weight\)</span> by <span class="math inline">\(family\)</span>, but only for a few selected families.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="co"># biochem %&gt;% </span></span>
<span id="cb17-2"><a href="#cb17-2"></a><span class="co">#   group_by(family) %&gt;%</span></span>
<span id="cb17-3"><a href="#cb17-3"></a><span class="co">#   count(family) %&gt;% </span></span>
<span id="cb17-4"><a href="#cb17-4"></a><span class="co">#   arrage(-n) %&gt;%</span></span>
<span id="cb17-5"><a href="#cb17-5"></a><span class="co">#   View()</span></span>
<span id="cb17-6"><a href="#cb17-6"></a></span>
<span id="cb17-7"><a href="#cb17-7"></a>bigfams <span class="ot">&lt;-</span> biochem <span class="sc">%&gt;%</span> </span>
<span id="cb17-8"><a href="#cb17-8"></a>  <span class="fu">group_by</span>(family) <span class="sc">%&gt;%</span></span>
<span id="cb17-9"><a href="#cb17-9"></a>  <span class="fu">count</span>(family) <span class="sc">%&gt;%</span></span>
<span id="cb17-10"><a href="#cb17-10"></a>  <span class="fu">filter</span>(n<span class="sc">&gt;</span><span class="dv">10</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-11"><a href="#cb17-11"></a>  <span class="fu">pull</span>(family)</span>
<span id="cb17-12"><a href="#cb17-12"></a>  </span>
<span id="cb17-13"><a href="#cb17-13"></a>biochem_bigfams <span class="ot">&lt;-</span> biochem <span class="sc">%&gt;%</span></span>
<span id="cb17-14"><a href="#cb17-14"></a>  <span class="fu">filter</span>(family <span class="sc">%in%</span> bigfams)</span>
<span id="cb17-15"><a href="#cb17-15"></a></span>
<span id="cb17-16"><a href="#cb17-16"></a><span class="co"># i have pre-selected some families to compare</span></span>
<span id="cb17-17"><a href="#cb17-17"></a>myfams <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"B1.5:E1.4(4) B1.5:A1.4(5)"</span>,</span>
<span id="cb17-18"><a href="#cb17-18"></a><span class="st">"F1.3:A1.2(3) F1.3:E2.2(3)"</span>,</span>
<span id="cb17-19"><a href="#cb17-19"></a><span class="st">"A1.3:D1.2(3) A1.3:H1.2(3)"</span>,</span>
<span id="cb17-20"><a href="#cb17-20"></a><span class="st">"D5.4:G2.3(4) D5.4:C4.3(4)"</span>)</span>
<span id="cb17-21"><a href="#cb17-21"></a></span>
<span id="cb17-22"><a href="#cb17-22"></a><span class="co"># only keep the familys in myfams</span></span>
<span id="cb17-23"><a href="#cb17-23"></a>fam_data <span class="ot">&lt;-</span> biochem_bigfams <span class="sc">%&gt;%</span></span>
<span id="cb17-24"><a href="#cb17-24"></a>  <span class="fu">filter</span>(family <span class="sc">%in%</span> myfams) <span class="sc">%&gt;%</span></span>
<span id="cb17-25"><a href="#cb17-25"></a>  <span class="fu">droplevels</span>() </span>
<span id="cb17-26"><a href="#cb17-26"></a></span>
<span id="cb17-27"><a href="#cb17-27"></a><span class="co"># simplify family names and make factor</span></span>
<span id="cb17-28"><a href="#cb17-28"></a>fam_data<span class="sc">$</span>family <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"</span><span class="sc">\\</span><span class="st">..*"</span>, </span>
<span id="cb17-29"><a href="#cb17-29"></a>                        <span class="at">replacement =</span> <span class="st">""</span>,</span>
<span id="cb17-30"><a href="#cb17-30"></a>                        <span class="at">x =</span> fam_data<span class="sc">$</span>family) <span class="sc">%&gt;%</span></span>
<span id="cb17-31"><a href="#cb17-31"></a>  as.factor</span>
<span id="cb17-32"><a href="#cb17-32"></a></span>
<span id="cb17-33"><a href="#cb17-33"></a></span>
<span id="cb17-34"><a href="#cb17-34"></a><span class="co"># make B1 the reference (most similar to overall mean)</span></span>
<span id="cb17-35"><a href="#cb17-35"></a>fam_data<span class="sc">$</span>family <span class="ot">&lt;-</span> <span class="fu">relevel</span>(<span class="at">x =</span> fam_data<span class="sc">$</span>family, <span class="at">ref =</span> <span class="st">"B1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Model: <span class="math inline">\(y_{i} = \beta_0 \cdot 1+ \beta_1 \cdot x_{i}\)</span></p>
<p>Null Hypothesis: <span class="math inline">\(\mathcal{H}_0: \beta_1 = 0\)</span></p>
<p><span class="math inline">\(\mathcal{H}_0:\)</span> mouse <span class="math inline">\(family\)</span> does NOT explain <span class="math inline">\(weight\)</span></p>
<p>Alternative Hypothesis: <span class="math inline">\(\mathcal{H}_1: \beta_1 \neq 0\)</span></p>
<p><span class="math inline">\(\mathcal{H}_1:\)</span> mouse <span class="math inline">\(family\)</span> does explain <span class="math inline">\(weight\)</span></p>
<p><strong>Important:</strong> <span class="math inline">\(x_{i}\)</span> is an indicator (0 or 1) saying which group point <span class="math inline">\(i\)</span> was sampled from using the matrix encoding of 0s and 1s.</p>
<p>Below is an example depicting 6 observations with 2 from each of 3 families:</p>
<p><span class="math inline">\(\begin{bmatrix} y_{1} \\ y_{2} \\ y_{3} \\y_{4} \\y_{5} \\y_{5} \end{bmatrix} = \begin{bmatrix} 1 &amp; 0 &amp; 0 \\ 1 &amp; 0 &amp; 0 \\ 0 &amp; 1 &amp; 0 \\ 0 &amp; 1 &amp; 0 \\ 0 &amp; 0 &amp; 1 \\ 0 &amp; 0 &amp; 1 \\ \end{bmatrix} \cdot \begin{bmatrix} \beta_0 \\ \beta_1 \end{bmatrix} + \begin{bmatrix} e_{1} \\ e_{2} \\ e_{3} \\e_{4} \\e_{5} \\e_{6} \end{bmatrix}\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1"></a><span class="co"># fitting a line weight vs intercept + family</span></span>
<span id="cb18-2"><a href="#cb18-2"></a>fit_WvFam <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> weight <span class="sc">~</span> family, <span class="at">data =</span> fam_data)</span>
<span id="cb18-3"><a href="#cb18-3"></a></span>
<span id="cb18-4"><a href="#cb18-4"></a></span>
<span id="cb18-5"><a href="#cb18-5"></a><span class="co"># augment (i.e. add fitted and residual values)</span></span>
<span id="cb18-6"><a href="#cb18-6"></a>biochem_WvFam <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_WvFam, fam_data)</span>
<span id="cb18-7"><a href="#cb18-7"></a></span>
<span id="cb18-8"><a href="#cb18-8"></a>mean_B1 <span class="ot">&lt;-</span> fit_WvFam<span class="sc">$</span>coefficients[<span class="dv">1</span>]</span>
<span id="cb18-9"><a href="#cb18-9"></a>mean_A1 <span class="ot">&lt;-</span> fit_WvFam<span class="sc">$</span>coefficients[<span class="dv">1</span>] <span class="sc">+</span> fit_WvFam<span class="sc">$</span>coefficients[<span class="dv">2</span>]</span>
<span id="cb18-10"><a href="#cb18-10"></a>mean_D5 <span class="ot">&lt;-</span> fit_WvFam<span class="sc">$</span>coefficients[<span class="dv">1</span>] <span class="sc">+</span> fit_WvFam<span class="sc">$</span>coefficients[<span class="dv">3</span>]</span>
<span id="cb18-11"><a href="#cb18-11"></a>mean_F1 <span class="ot">&lt;-</span> fit_WvFam<span class="sc">$</span>coefficients[<span class="dv">1</span>] <span class="sc">+</span> fit_WvFam<span class="sc">$</span>coefficients[<span class="dv">4</span>]</span>
<span id="cb18-12"><a href="#cb18-12"></a></span>
<span id="cb18-13"><a href="#cb18-13"></a><span class="co"># plot of data with mean and colored by residuals</span></span>
<span id="cb18-14"><a href="#cb18-14"></a>p_WvFam_res <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(biochem_WvFam, <span class="fu">aes</span>(<span class="at">x =</span> family, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb18-15"><a href="#cb18-15"></a>  <span class="fu">geom_point</span>(<span class="at">position =</span> <span class="fu">position_jitter</span>(), <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span></span>
<span id="cb18-16"><a href="#cb18-16"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span>.<span class="dv">5</span>, <span class="at">xend=</span><span class="fl">1.5</span>, <span class="at">y=</span>mean_B1, <span class="at">yend=</span>mean_B1), <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb18-17"><a href="#cb18-17"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fl">1.5</span>, <span class="at">xend=</span><span class="fl">2.5</span>, <span class="at">y=</span>mean_A1, <span class="at">yend=</span>mean_A1), <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb18-18"><a href="#cb18-18"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fl">2.5</span>, <span class="at">xend=</span><span class="fl">3.5</span>, <span class="at">y=</span>mean_D5, <span class="at">yend=</span>mean_D5), <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb18-19"><a href="#cb18-19"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span><span class="fl">3.5</span>, <span class="at">xend=</span><span class="fl">4.5</span>, <span class="at">y=</span>mean_F1, <span class="at">yend=</span>mean_F1), <span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb18-20"><a href="#cb18-20"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">x=</span>.<span class="dv">5</span>, <span class="at">xend=</span><span class="fl">4.5</span>, <span class="at">y=</span><span class="fu">mean</span>(weight), <span class="at">yend=</span><span class="fu">mean</span>(weight)), <span class="at">color=</span><span class="st">"black"</span>) <span class="sc">+</span></span>
<span id="cb18-21"><a href="#cb18-21"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb18-22"><a href="#cb18-22"></a>  <span class="fu">theme_minimal</span>()  </span>
<span id="cb18-23"><a href="#cb18-23"></a></span>
<span id="cb18-24"><a href="#cb18-24"></a>p_WvFam_res</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="class-04_files/figure-revealjs/fit weight to families-1.png" width="960" class="r-stretch"></section>
<section class="slide level2">

<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a>anova_WvFam <span class="ot">&lt;-</span> <span class="fu">aov</span>(<span class="at">formula =</span> weight <span class="sc">~</span> family, <span class="at">data =</span> fam_data)</span>
<span id="cb19-2"><a href="#cb19-2"></a></span>
<span id="cb19-3"><a href="#cb19-3"></a><span class="co"># are the coefficients the same?</span></span>
<span id="cb19-4"><a href="#cb19-4"></a><span class="fu">bind_cols</span>(<span class="fu">tidy</span>(fit_WvFam) <span class="sc">%&gt;%</span> <span class="fu">select</span>(term,estimate),</span>
<span id="cb19-5"><a href="#cb19-5"></a>          <span class="at">ANOVA_coef=</span>anova_WvFam<span class="sc">$</span>coefficients</span>
<span id="cb19-6"><a href="#cb19-6"></a>          )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4 × 3
  term        estimate ANOVA_coef
  &lt;chr&gt;          &lt;dbl&gt;      &lt;dbl&gt;
1 (Intercept)   18.1       18.1  
2 familyA1      -3.66      -3.66 
3 familyD5       5.86       5.86 
4 familyF1      -0.682     -0.682</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">glance</span>(fit_WvFam) <span class="sc">%&gt;%</span> <span class="fu">select</span>(p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 1
   p.value
     &lt;dbl&gt;
1 7.62e-13</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a><span class="fu">tidy</span>(anova_WvFam) <span class="sc">%&gt;%</span> <span class="fu">select</span>(p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 1
    p.value
      &lt;dbl&gt;
1  7.62e-13
2 NA       </code></pre>
</div>
</div>
</section>
<section class="slide level2">

<h3 id="class-exercise-3"><strong>Class exercise 3</strong>:</h3>

<img data-src="class-04_files/figure-revealjs/unnamed-chunk-10-1.png" width="960" class="r-stretch"><p>What are the <span class="math inline">\(p_{mean}\)</span> and <span class="math inline">\(p_{fit}\)</span> for the models above?</p>
<div class="cell">
<div class="cell-output-display">
<table>
<thead>
<tr class="header">
<th style="text-align: left;">Parameter</th>
<th style="text-align: left;">fit cholesterol</th>
<th style="text-align: left;">fit gender</th>
<th style="text-align: left;">fit family</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><span class="math inline">\(p_{mean}\)</span></td>
<td style="text-align: left;">?</td>
<td style="text-align: left;">?</td>
<td style="text-align: left;">?</td>
</tr>
<tr class="even">
<td style="text-align: left;"><span class="math inline">\(p_{fit}\)</span></td>
<td style="text-align: left;">?</td>
<td style="text-align: left;">?</td>
<td style="text-align: left;">?</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="comparing-2-continuouse-variables-for-2-groups" class="title-slide slide level1 center">
<h1>Comparing 2 continuouse variables for 2 groups?</h1>

</section>
<section id="ancova-analysis-of-covariance" class="slide level2">
<h2>ANCOVA, Analysis of Covariance</h2>
<p>ANOVA with more than one independent variable. In this case, what is the impact of mouse age on mouse weight for males vs females.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1"></a><span class="co"># more info here https://towardsdatascience.com/doing-and-reporting-your-first-anova-and-ancova-in-r-1d820940f2ef</span></span>
<span id="cb25-2"><a href="#cb25-2"></a></span>
<span id="cb25-3"><a href="#cb25-3"></a>p_WvA_gender <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> age, <span class="at">color=</span>gender)) <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm) <span class="sc">+</span></span>
<span id="cb25-6"><a href="#cb25-6"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb25-7"><a href="#cb25-7"></a></span>
<span id="cb25-8"><a href="#cb25-8"></a>p_WvA_gender</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>WvA_gender <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> age <span class="sc">+</span> gender, <span class="at">data =</span> biochem)</span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a>WvA_gender <span class="sc">%&gt;%</span> <span class="fu">glance</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 12
  r.squared adj.r.squared sigma statistic   p.value    df logLik   AIC   BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     0.475         0.475  2.42      805. 8.37e-250     2 -4100. 8209. 8231.
# ℹ 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a><span class="fu">aov</span>(<span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> age <span class="sc">+</span> gender, <span class="at">data =</span> biochem) <span class="sc">%&gt;%</span> <span class="fu">glance</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 6
  logLik   AIC   BIC deviance  nobs r.squared
   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt; &lt;int&gt;     &lt;dbl&gt;
1 -4100. 8209. 8231.   10402.  1782     0.475</code></pre>
</div>
</div>
<img data-src="class-04_files/figure-revealjs/ANCOVA-1.png" width="960" class="r-stretch"></section></section>
<section id="references-and-resources" class="title-slide slide level1 center">
<h1>References and resources</h1>
<p>I have borrowed heavily from, directly taken, and/or highly recommend the following fantastic resources:</p>
<ol type="1">
<li><a href="https://lindeloev.github.io/tests-as-linear/">Common statistical tests are linear models</a> from Jonas Kristoffer Lindeløv<br>
</li>
<li><a href="https://www.youtube.com/channel/UCtYLUTtgS3k1Fg4y5tAhLbw">Statquest</a></li>
<li><a href="https://www.rapidtables.com/math/symbols/Statistical_Symbols.html">Stats gobbledygook</a></li>
<li><a href="http://www.sthda.com/english/articles/39-regression-model-diagnostics/161-linear-regression-assumptions-and-diagnostics-in-r-essentials/">Linear Regression Assumptions and Diagnostics in R: Essentials</a><br>
</li>
<li><a href="https://www.graphpad.com/guides/prism/latest/statistics/stat_---_principles_of_statistics_-.htm">PRINCIPLES OF STATISTICS</a> from GraphPad/SAS.</li>
</ol>

<div class="footer footer-default">

</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>