<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.361">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Neelanjan Mukherjee">

<title>MOLB 7950 - Bootcamp: Stats class 3</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../../">
<link href="../../../content/bootcamp/stats/class-04.html" rel="next">
<link href="../../../content/bootcamp/stats/class-02.html" rel="prev">
<script src="../../../site_libs/quarto-html/quarto.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<meta property="og:title" content="MOLB 7950 - Bootcamp: Stats class 3">
<meta property="og:description" content="">
<meta property="og:image" content="https://rnabioco.github.io/molb-7950/content/bootcamp/stats/img/easy_hard.jpg">
<meta property="og:site-name" content="MOLB 7950">
<meta name="twitter:title" content="MOLB 7950 - Bootcamp: Stats class 3">
<meta name="twitter:description" content="">
<meta name="twitter:image" content="https://rnabioco.github.io/molb-7950/content/bootcamp/stats/images/twitter-card.png">
<meta name="twitter:creator" content="@rnabioco">
<meta name="twitter:card" content="summary_large_image">
</head>

<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../../../content/bootcamp/stats/class-01.html">Bootcamp — Statistics</a></li><li class="breadcrumb-item"><a href="../../../content/bootcamp/stats/class-03.html">Bootcamp Stats Day 3</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="pt-lg-2 mt-2 text-center sidebar-header">
      <a href="../../../index.html" class="sidebar-logo-link">
      <img src="../../../content/img/logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none">
      </a>
      <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/rnabioco/molb-7950" title="Course GitHub page" class="quarto-navigation-tool px-1" aria-label="Course GitHub page"><i class="bi bi-github"></i></a>
    <a href="https://sso.posit.cloud/cu-anschutz" title="Posit Cloud SSO login" class="quarto-navigation-tool px-1" aria-label="Posit Cloud SSO login"><i class="bi bi-cloud"></i></a>
    <a href="https://calendly.com/molb7950" title="Office hours scheduling" class="quarto-navigation-tool px-1" aria-label="Office hours scheduling"><i class="bi bi-calendar"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Course information</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/course-info/syllabus.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/course-info/support.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Support</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/course-info/team.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Teaching team</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/course-info/links.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Useful links</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false">
 <span class="menu-text">Bootcamp — R</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/r/class-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp R Day 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/r/class-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp R Day 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/r/class-03.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp R Day 3</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/r/class-04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp R Day 4</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/r/class-05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp R Day 5</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true">
 <span class="menu-text">Bootcamp — Statistics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/stats/class-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp Stats Day 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/stats/class-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp Stats Day 2</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/stats/class-03.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Bootcamp Stats Day 3</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/stats/class-04.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp Stats Day 4</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/bootcamp/stats/class-05.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp Stats Day 5</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Block DNA</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/block-dna/variant-calling/class-01.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant Calling Day 1</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/block-dna/variant-calling/class-02.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Variant Calling Day 2</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Assignments</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/course-info/problem-sets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Problet Set Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../../../content/course-info/final-projects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Final Project Overview</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#outline" id="toc-outline" class="nav-link" data-scroll-target="#outline">Outline</a>
  <ul class="collapse">
  <li><a href="#provide-a-simple-and-flexible-framework" id="toc-provide-a-simple-and-flexible-framework" class="nav-link" data-scroll-target="#provide-a-simple-and-flexible-framework">Provide a <strong>simple</strong> and <strong>flexible</strong> framework</a></li>
  </ul></li>
  <li><a href="#cheatsheet" id="toc-cheatsheet" class="nav-link" data-scroll-target="#cheatsheet">CHEATSHEET</a>
  <ul class="collapse">
  <li><a href="#variables-definitions" id="toc-variables-definitions" class="nav-link" data-scroll-target="#variables-definitions">Variables definitions</a>
  <ul class="collapse">
  <li><a href="#random-variables-x-y" id="toc-random-variables-x-y" class="nav-link" data-scroll-target="#random-variables-x-y">Random variables (x, y)</a></li>
  <li><a href="#quantitative-variables" id="toc-quantitative-variables" class="nav-link" data-scroll-target="#quantitative-variables">Quantitative Variables</a></li>
  <li><a href="#categorical-variables" id="toc-categorical-variables" class="nav-link" data-scroll-target="#categorical-variables">Categorical Variables</a></li>
  </ul></li>
  <li><a href="#hypothesis-testing-definitions" id="toc-hypothesis-testing-definitions" class="nav-link" data-scroll-target="#hypothesis-testing-definitions">Hypothesis testing definitions</a></li>
  <li><a href="#null-hypothesis-testing" id="toc-null-hypothesis-testing" class="nav-link" data-scroll-target="#null-hypothesis-testing">Null hypothesis testing</a></li>
  <li><a href="#the-simplicity-underlying-common-tests" id="toc-the-simplicity-underlying-common-tests" class="nav-link" data-scroll-target="#the-simplicity-underlying-common-tests">The simplicity underlying common tests</a></li>
  <li><a href="#parametric-vs-non-parametric-tests" id="toc-parametric-vs-non-parametric-tests" class="nav-link" data-scroll-target="#parametric-vs-non-parametric-tests">Parametric vs Non-Parametric tests</a></li>
  </ul></li>
  <li><a href="#import-and-tidy-data" id="toc-import-and-tidy-data" class="nav-link" data-scroll-target="#import-and-tidy-data">Import and tidy data</a></li>
  <li><a href="#relationship-between-two-or-more-continuous-variables" id="toc-relationship-between-two-or-more-continuous-variables" class="nav-link" data-scroll-target="#relationship-between-two-or-more-continuous-variables">Relationship between two or more continuous variables?</a>
  <ul class="collapse">
  <li><a href="#correlation-vs-regression" id="toc-correlation-vs-regression" class="nav-link" data-scroll-target="#correlation-vs-regression">Correlation vs Regression</a>
  <ul class="collapse">
  <li><a href="#pearson-correlation" id="toc-pearson-correlation" class="nav-link" data-scroll-target="#pearson-correlation">Pearson Correlation</a></li>
  <li><a href="#spearman-correlation-nonparametric" id="toc-spearman-correlation-nonparametric" class="nav-link" data-scroll-target="#spearman-correlation-nonparametric">Spearman Correlation (nonparametric)</a></li>
  </ul></li>
  <li><a href="#regression" id="toc-regression" class="nav-link" data-scroll-target="#regression">Regression</a></li>
  <li><a href="#equation-for-a-line" id="toc-equation-for-a-line" class="nav-link" data-scroll-target="#equation-for-a-line">Equation for a line</a></li>
  <li><a href="#stats-equation-for-a-line" id="toc-stats-equation-for-a-line" class="nav-link" data-scroll-target="#stats-equation-for-a-line">Stats equation for a line</a></li>
  <li><a href="#linear-regression" id="toc-linear-regression" class="nav-link" data-scroll-target="#linear-regression">Linear Regression</a>
  <ul class="collapse">
  <li><a href="#step-1-can-mouse-cholesterol-levels-help-explain-mouse-weight" id="toc-step-1-can-mouse-cholesterol-levels-help-explain-mouse-weight" class="nav-link" data-scroll-target="#step-1-can-mouse-cholesterol-levels-help-explain-mouse-weight">STEP 1: Can mouse cholesterol levels help explain mouse weight?</a></li>
  <li><a href="#step-2-do-the-regression" id="toc-step-2-do-the-regression" class="nav-link" data-scroll-target="#step-2-do-the-regression">STEP 2: Do the regression</a></li>
  <li><a href="#step-4-look-at-the-r2-f-value-and-p-value" id="toc-step-4-look-at-the-r2-f-value-and-p-value" class="nav-link" data-scroll-target="#step-4-look-at-the-r2-f-value-and-p-value">STEP 4: Look at the <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value</a></li>
  <li><a href="#step-5-visualize-fit-and-errors" id="toc-step-5-visualize-fit-and-errors" class="nav-link" data-scroll-target="#step-5-visualize-fit-and-errors">STEP 5: Visualize fit and errors</a></li>
  <li><a href="#how-to-find-the-best-least-squares-fit" id="toc-how-to-find-the-best-least-squares-fit" class="nav-link" data-scroll-target="#how-to-find-the-best-least-squares-fit">How to find the best (least squares) fit?</a></li>
  <li><a href="#non-parametric-version" id="toc-non-parametric-version" class="nav-link" data-scroll-target="#non-parametric-version">Non-parametric version</a></li>
  </ul></li>
  <li><a href="#multiple-regression" id="toc-multiple-regression" class="nav-link" data-scroll-target="#multiple-regression">Multiple regression</a></li>
  <li><a href="#non-linear-regression" id="toc-non-linear-regression" class="nav-link" data-scroll-target="#non-linear-regression">Non-linear regression</a></li>
  </ul></li>
  <li><a href="#references-and-resources" id="toc-references-and-resources" class="nav-link" data-scroll-target="#references-and-resources">References and resources</a></li>
  </ul>
<div class="toc-actions"><div><i class="bi bi-github"></i></div><div class="action-links"><p><a href="https://github.com/rnabioco/molb-7950/issues/new" class="toc-action">Report an issue</a></p><p><a href="https://github.com/rnabioco/molb-7950/edit/main/content/bootcamp/stats/class-03.qmd" class="toc-action">Edit this page</a></p></div></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Bootcamp: Stats class 3</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p><a href="neelanjan.mukherjee@cuanschutz.edu">Neelanjan Mukherjee</a> </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">Invalid Date</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<hr>
<section id="learning-objectives" class="level1">
<h1>Learning Objectives</h1>
<ul>
<li><strong>Visualize</strong> and <strong>Summarize</strong> the data being compared</li>
<li><strong>Formulate</strong> and <strong>Execute</strong> null hypothesis testing</li>
<li><strong>Identify</strong> and <strong>Perform</strong> the proper statistical test for data type/comparison</li>
<li><strong>Calculate</strong> and <strong>Interpret</strong> p-values</li>
<li><strong>Prevent</strong> p-hacking and <strong>Recognize</strong> issues with simultaneously testing multiple hypotheses.</li>
</ul>
<hr>
</section>
<section id="outline" class="level1">
<h1>Outline</h1>
<ul>
<li><strong>Concepts and Definitions</strong>
<ul>
<li><strong>Simplifying principles: Common tests as linear models</strong></li>
<li><strong>Types of comparisons and statistical tests</strong></li>
<li><strong>Definitions<br>
</strong></li>
</ul></li>
<li><strong>Day 1: Relationship between two or more continuous variables</strong>
<ul>
<li><strong>Correlation vs Regression</strong></li>
<li><strong>Fitting it a line to data</strong></li>
<li><strong>Linear regression concepts</strong></li>
<li><strong>Multiple regression</strong><br>
</li>
</ul></li>
<li>Day 2: Relationship between categorical and continuous variables
<ul>
<li><p>Comparing means between two groups (t-Test)</p></li>
<li><p>Comparing means between 3 or more groups (ANOVA)<br>
</p></li>
</ul></li>
<li>Day 3: Multiple test correction, Bayesian intro, History</li>
</ul>
<hr>
<section id="provide-a-simple-and-flexible-framework" class="level2">
<h2 class="anchored" data-anchor-id="provide-a-simple-and-flexible-framework">Provide a <strong>simple</strong> and <strong>flexible</strong> framework</h2>
<p><img src="img/easy_hard.jpg" class="img-fluid"></p>
<hr>
</section>
</section>
<section id="cheatsheet" class="level1">
<h1>CHEATSHEET</h1>
<p><img src="img/linear_tests_cheat_sheet_adapted.png" class="img-fluid"></p>
<hr>
<section id="variables-definitions" class="level2">
<h2 class="anchored" data-anchor-id="variables-definitions">Variables definitions</h2>
<section id="random-variables-x-y" class="level3">
<h3 class="anchored" data-anchor-id="random-variables-x-y">Random variables (x, y)</h3>
<p><strong>Response Variable</strong> ( <strong>y</strong> - aka dependent or outcome variable): this variable is predicted or its variation is explained by the explanatory variable. In an experiment, this is the outcome that is measured following manipulation of the explanatory variable.</p>
<p><strong>Explanatory Variable</strong> ( <strong>x</strong> - aka independent or predictor variable): explains variations in the response variable. In an experiment, it is manipulated by the researcher.</p>
</section>
<section id="quantitative-variables" class="level3">
<h3 class="anchored" data-anchor-id="quantitative-variables">Quantitative Variables</h3>
<p><strong>Discrete variable</strong>: numeric variables that have a countable number of values between any two values - <code>integer</code> in R (e.g., number of mice, read counts).</p>
<p><strong>Continuous variable</strong>: numeric variables that have an infinite number of values between any two values - <code>numeric</code> in R (e.g., normalized expression values, fluorescent intensity).</p>
</section>
<section id="categorical-variables" class="level3">
<h3 class="anchored" data-anchor-id="categorical-variables">Categorical Variables</h3>
<p><strong>Nominal variable</strong>: (unordered) random variables have categories where order doesn’t matter - <code>factor</code> in R (e.g., country, type of gene, genotype).</p>
<p><strong>Ordinal variable</strong>: (ordered) random variables have ordered categories - order of <code>levels</code> in R ( e.g.&nbsp;grade of tumor).</p>
<hr>
</section>
</section>
<section id="hypothesis-testing-definitions" class="level2">
<h2 class="anchored" data-anchor-id="hypothesis-testing-definitions">Hypothesis testing definitions</h2>
<p><strong>Hypothesis testing</strong> is a statistical analysis that uses sample data to assess two mutually exclusive theories about the properties of a population. Statisticians call these theories the null hypothesis and the alternative hypothesis. A hypothesis test assesses your sample statistic and factors in an estimate of the sample error to determine which hypothesis the data support.</p>
<p>When you can reject the null hypothesis, the results are statistically significant, and your data support the theory that an effect exists at the population level.</p>
<p><a href="https://www.graphpad.com/guides/prism/latest/statistics/hypothesis_testing_and_statistical_significance.htm"><strong>A legal analogy: Guilty or not guilty?</strong></a><strong><br>
</strong>The statistical concept of ‘significant’ vs.&nbsp;‘not significant’ can be understood by comparing to the legal concept of ‘guilty’ vs.&nbsp;‘not guilty’.</p>
<p>In the American legal system (and much of the world) a criminal defendant is presumed innocent until proven guilty. If the evidence proves the defendant guilty beyond a reasonable doubt, the verdict is ‘guilty’. Otherwise the verdict is ‘not guilty’. In some countries, this verdict is ‘not proven’, which is a better description. A ‘not guilty’ verdict does not mean the judge or jury concluded that the defendant is innocent -- it just means that the evidence was not strong enough to persuade the judge or jury that the defendant was guilty.</p>
<p>In statistical hypothesis testing, you start with the null hypothesis (usually that there is no difference between groups). If the evidence produces a small enough P value, you reject that null hypothesis, and conclude that the difference is real. If the P value is higher than your threshold (usually 0.05), you don’t reject the null hypothesis. This doesn’t mean the evidence convinced you that the treatment had no effect, only that the evidence was not persuasive enough to convince you that there is an effect.</p>
<p><strong>Effect</strong> — the difference between the population value and the null hypothesis value. The effect is also known as population effect or the difference. Typically, you do not know the size of the actual effect. However, you can use a hypothesis test to help you determine whether an effect exists and to estimate its size.</p>
<p><strong>Null Hypothesis</strong> or <span class="math inline">\(\mathcal{H}_0\)</span> — one of two mutually exclusive theories about the properties of the population in hypothesis testing. Typically, the null hypothesis states that there is no effect (i.e., the effect size equals zero).</p>
<p><strong>Alternative Hypothesis</strong> or <span class="math inline">\(\mathcal{H}_1\)</span> — the other theory about the properties of the population in hypothesis testing. Typically, the alternative hypothesis states that a population parameter does not equal the null hypothesis value. In other words, there is a non-zero effect. If your sample contains sufficient evidence, you can reject the null and favor the alternative hypothesis.</p>
<p><strong>P-values</strong> — the probability of obtaining test results at least as extreme as the results actually observed, under the assumption that the null hypothesis is correct. Lower p-values represent stronger evidence against the null. P-values in conjunction with the significance level determines whether your data favor the null or alternative hypothesis.</p>
<p><a href="https://www.youtube.com/watch?v=5Z9OIYA8He8">StatQuest: P Values, clearly explained</a></p>
<p><a href="https://www.youtube.com/watch?v=JQc3yx0-Q9E">StatQuest: How to calculate p-values</a></p>
<p><strong>Significance Level</strong> or <span class="math inline">\(a\)</span> — an evidentiary standard set before the study. It is the probability that you say there is an effect when there is no effect (the probability of rejecting the null hypothesis given that it is true). Lower significance levels indicate that you require stronger evidence before you will reject the null.It is usually set at or below .05.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="https://upload.wikimedia.org/wikipedia/commons/3/35/Guinness_Glass_2010.jpg" class="img-fluid figure-img" style="width:10.0%"></p>
<figcaption class="figure-caption">Guinness</figcaption>
</figure>
</div>
<hr>
</section>
<section id="null-hypothesis-testing" class="level2">
<h2 class="anchored" data-anchor-id="null-hypothesis-testing">Null hypothesis testing</h2>
<ol type="1">
<li>Specify the variables</li>
<li>Declare null hypothesis <span class="math inline">\(\mathcal{H}_0\)</span></li>
<li>Calculate test-statistic, exact p-value</li>
<li><em>Generate and visualize data reflecting null-distribution</em></li>
<li><em>Calculate the p-value from the test statistic and null distribution</em></li>
</ol>
<p>*4-5: For calculating empirical p-value</p>
<hr>
</section>
<section id="the-simplicity-underlying-common-tests" class="level2">
<h2 class="anchored" data-anchor-id="the-simplicity-underlying-common-tests">The simplicity underlying common tests</h2>
<p>Most of the common statistical models (t-test, correlation, ANOVA; chi-square, etc.) are special cases of linear models or a very close approximation. This simplicity means that there is less to learn. In particular, it all comes down to:<br>
<span class="math inline">\(y = a \cdot x + b\)</span></p>
<p>This needless complexity multiplies when students try to rote learn the parametric assumptions underlying each test separately rather than deducing them from the linear model.</p>
<hr>
</section>
<section id="parametric-vs-non-parametric-tests" class="level2">
<h2 class="anchored" data-anchor-id="parametric-vs-non-parametric-tests">Parametric vs Non-Parametric tests</h2>
<p><strong>Parametric tests</strong> are suitable for normally distributed data.</p>
<p><strong>Non-Parametric tests</strong> are suitable for any continuous data. For the sake of simplicity and sticking with a consistent framework, we will consider Non-Parametric tests as the <strong>ranked versions of the corresponding parametric tests</strong>.</p>
<p><a href="https://statisticsbyjim.com/hypothesis-testing/nonparametric-parametric-tests/">More on choosing Parametric vs Non-Parametric</a></p>
<div class="cell">
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Info</th>
<th style="text-align: left;">Parametric</th>
<th style="text-align: left;">Non-Parametric</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">better descriptor</td>
<td style="text-align: left;">mean</td>
<td style="text-align: left;">median</td>
</tr>
<tr class="even">
<td style="text-align: left;"># of samples (N)</td>
<td style="text-align: left;">many</td>
<td style="text-align: left;">few</td>
</tr>
</tbody>
</table>
</div>
</div>
<hr>
</section>
</section>
<section id="import-and-tidy-data" class="level1">
<h1>Import and tidy data</h1>
<p>We will be using mouse data from <a href="https://wp.cs.ucl.ac.uk/outbredmice/">Resources for Outbred Mice</a>. The goal of the study was to establish genotype-phenotype relationships for highly recombinant outbred mouse populations. We will be using the phenotypic data for our exercises.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we are reading the data directly from the internet</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>biochem <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"http://mtweb.cs.ucl.ac.uk/HSMICE/PHENOTYPES/Biochemistry.txt"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  janitor<span class="sc">::</span><span class="fu">clean_names</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00C4&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00D6&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00DC&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00E4&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00F6&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00FC&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00DF&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00C6&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00E6&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00D8&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00F8&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00C5&gt;' to native encoding</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in FUN(X[[i]], ...): unable to translate '&lt;U+00E5&gt;' to native encoding</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simplify names a bit more</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(biochem) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"biochem_"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="fu">colnames</span>(biochem))</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># we are going to simplify this a bit and only keep some columns</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>keep <span class="ot">&lt;-</span> <span class="fu">colnames</span>(biochem)[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">6</span>, <span class="dv">9</span>, <span class="dv">14</span>, <span class="dv">15</span>, <span class="dv">24</span><span class="sc">:</span><span class="dv">28</span>)]</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>biochem <span class="ot">&lt;-</span> biochem[, keep]</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co"># get weights for each individual mouse</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co"># careful: did not come with column names</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>weight <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"http://mtweb.cs.ucl.ac.uk/HSMICE/PHENOTYPES/weight"</span>, <span class="at">col_names =</span> F, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># add column names</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(weight) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"subject_name"</span>, <span class="st">"weight"</span>)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a><span class="co"># add weight to biochem table and get rid of NAs</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>biochem <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(biochem, weight, <span class="at">by =</span> <span class="st">"subject_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a><span class="co"># explore the data a bit</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "subject_name"    "calcium"         "glucose"         "sodium"         
 [5] "tot_cholesterol" "family"          "gender"          "age"            
 [9] "cage_density"    "litter"          "weight"         </code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [1,782 x 11] (S3: tbl_df/tbl/data.frame)
 $ subject_name   : chr [1:1782] "A048005080" "A048005112" "A048006555" "A048007096" ...
 $ calcium        : num [1:1782] 1.97 2.11 1.71 2.49 2.14 2.16 2.29 1.69 1.84 2.31 ...
 $ glucose        : num [1:1782] 12.23 10.99 5.97 10.61 11.88 ...
 $ sodium         : num [1:1782] 123 133 119 148 131 134 146 117 121 144 ...
 $ tot_cholesterol: num [1:1782] 3.01 2.46 3.57 2.61 2.04 2.86 3.22 3.47 3.35 2.29 ...
 $ family         : chr [1:1782] "H2.3:C5.2(3) H2.3:G2.2(3)" "H2.2:C3.1(4) H2.2:G3.1(3)" "E1.3:H1.2(3) E1.3:D1.2(3)" "D3.2:G2.1(5) D3.2:C5.1(5)" ...
 $ gender         : chr [1:1782] "F" "F" "M" "M" ...
 $ age            : num [1:1782] 66 70 72 66 63 72 66 72 64 66 ...
 $ cage_density   : num [1:1782] 5 3 4 4 6 7 4 5 6 3 ...
 $ litter         : num [1:1782] 2 4 1 4 1 3 1 1 1 1 ...
 $ weight         : num [1:1782] 20.3 16.1 19.5 22.2 17.3 18.1 25.6 18.6 23.1 17.3 ...
 - attr(*, "na.action")= 'omit' Named int [1:237] 3 8 11 17 18 23 25 35 38 40 ...
  ..- attr(*, "names")= chr [1:237] "3" "8" "11" "17" ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View(biochem)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="relationship-between-two-or-more-continuous-variables" class="level1">
<h1>Relationship between two or more continuous variables?</h1>
<p>What is the relationship between weight and cholesterol?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> tot_cholesterol)) <span class="sc">+</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<section id="correlation-vs-regression" class="level2">
<h2 class="anchored" data-anchor-id="correlation-vs-regression">Correlation vs Regression</h2>
<p><strong>Correlation</strong> is primarily used to quickly and concisely summarize the direction and strength of the relationships between a set of 2 or more numeric variables.<c2><a0></a0></c2></p>
<p><strong>Regression</strong> is primarily used to build models/equations to predict a key response, Y, from a set of predictor (X) variables.</p>
<table class="table">
<colgroup>
<col style="width: 25%">
<col style="width: 37%">
<col style="width: 37%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Correlation</th>
<th>Regression</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Description</strong></td>
<td>Association between 2 or more variables</td>
<td>How an independent variable is numerically related to the dependent variable</td>
</tr>
<tr class="even">
<td><strong>Usage</strong></td>
<td>To represent linear relationship between two variables</td>
<td>To fit a best line and estimate one variable on the basis of another variable</td>
</tr>
<tr class="odd">
<td><strong>Dependent vs Independent variables</strong></td>
<td>Doesn’t matter</td>
<td>must define (i.e.&nbsp;order of relationship matters)</td>
</tr>
<tr class="even">
<td><strong>Interpretation</strong></td>
<td>Correlation coefficient indicates the extent to which two variables move together</td>
<td>Regression indicates the impact of a unit change in the known variable (x) on the estimated variable (y)</td>
</tr>
<tr class="odd">
<td><strong>Goal</strong></td>
<td>To find a numerical value expressing the relationship between variables</td>
<td>To estimate values of random variable on the basis of the values of fixed variable</td>
</tr>
</tbody>
</table>
<p>*Borrowed from and more info available <a href="https://www.graphpad.com/support/faq/what-is-the-difference-between-correlation-and-linear-regression/">here</a> and <a href="https://keydifferences.com/difference-between-correlation-and-regression.html">here</a>.</p>
<section id="pearson-correlation" class="level3">
<h3 class="anchored" data-anchor-id="pearson-correlation">Pearson Correlation</h3>
<p>It was developed by <a href="https://en.wikipedia.org/wiki/Karl_Pearson" title="Karl Pearson">Karl Pearson</a> from a related idea introduced by <a href="https://en.wikipedia.org/wiki/Francis_Galton" title="Francis Galton">Francis Galton</a> in the 1880s, and for which the mathematical formula was derived and published by <a href="https://en.wikipedia.org/wiki/Auguste_Bravais" title="Auguste Bravais">Auguste Bravais</a> in 1844.<sup><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-6">[a]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-7">[6]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-8">[7]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-9">[8]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-10">[9]</a></sup> The naming of the coefficient is thus an example of <a href="https://en.wikipedia.org/wiki/Stigler%27s_Law" title="Stigler's Law">Stigler’s Law</a> (see list of examples <a href="https://en.wikipedia.org/wiki/List_of_examples_of_Stigler%27s_law">here</a>).</p>
<p>Interpretation of coefficient:</p>
<p>1 = perfect linear correlation</p>
<p>0 = no correlation</p>
<p>-1 = perfect linear anti-correlation</p>
<p><span class="math inline">\(Corr(x,y) = \displaystyle \frac {\sum_{i=1}^{n} (x_{i} - \overline{x})(y_{i} - \overline{y})}{\sum_{i=1}^{n} \sqrt(x_{i} - \overline{x})^2 \sqrt(y_{i} - \overline{y})^2}\)</span></p>
<p><span class="math inline">\(x_{i}\)</span> = the “i-th” observation of the variable <span class="math inline">\(x\)</span></p>
<p><span class="math inline">\(\overline{x}\)</span> = mean of all observations of <span class="math inline">\(x\)</span></p>
<p><span class="math inline">\(y_{i}\)</span> = the “i-th” observation of the variable <span class="math inline">\(y\)</span></p>
<p><span class="math inline">\(\overline{y}\)</span> = mean of all observations of <span class="math inline">\(y\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>r_pearson <span class="ot">&lt;-</span> <span class="fu">cor</span>(<span class="at">x =</span> biochem<span class="sc">$</span>tot_cholesterol, <span class="at">y =</span> biochem<span class="sc">$</span>weight)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>r_pearson</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3540731</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># average total cholesterol</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>avg_chol <span class="ot">&lt;-</span> <span class="fu">mean</span>(biochem<span class="sc">$</span>tot_cholesterol)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a><span class="co"># average weight</span></span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>avg_weight <span class="ot">&lt;-</span> <span class="fu">mean</span>(biochem<span class="sc">$</span>weight)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="co"># difference from mean total cholesterol</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>diff_chol <span class="ot">&lt;-</span> biochem<span class="sc">$</span>tot_cholesterol <span class="sc">-</span> avg_chol</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="co"># difference from mean total cholesterol</span></span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>diff_weight <span class="ot">&lt;-</span> biochem<span class="sc">$</span>weight <span class="sc">-</span> avg_weight</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a><span class="co"># follow formula above</span></span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>manual_pearson <span class="ot">&lt;-</span> <span class="fu">sum</span>(diff_chol <span class="sc">*</span> diff_weight) <span class="sc">/</span> (</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sqrt</span>(<span class="fu">sum</span>(diff_chol<span class="sc">^</span><span class="dv">2</span>)) <span class="sc">*</span> <span class="fu">sqrt</span>(<span class="fu">sum</span>(diff_weight<span class="sc">^</span><span class="dv">2</span>)))</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>manual_pearson</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3540731</code></pre>
</div>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">identical</span>(manual_pearson, r_pearson)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cov(x = biochem$tot_cholesterol, y = biochem$weight)/(sd(biochem$tot_cholesterol)*sd(biochem$weight))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
<section id="spearman-correlation-nonparametric" class="level3">
<h3 class="anchored" data-anchor-id="spearman-correlation-nonparametric">Spearman Correlation (nonparametric)</h3>
<p><a href="https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient">Spearman’s rank correlation coefficient</a>or Spearman’s <cf>&lt;81&gt;, named after Charles Spearman is a <strong>nonparametric</strong> measure of <strong>rank</strong> correlation (statistical dependence between the rankings of two variables). It assesses how well the relationship between two variables can be described using a monotonic function.</cf></p>
<p>More info <a href="https://towardsdatascience.com/clearly-explained-pearson-v-s-spearman-correlation-coefficient-ada2f473b8#:~:text=The%20fundamental%20difference%20between%20the,with%20monotonic%20relationships%20as%20well.">here</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">30</span>, <span class="dv">1</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span>x</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(x, y, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">las =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/spearman-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(x, y, <span class="at">method =</span> <span class="st">"pearson"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5199042</code></pre>
</div>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(x, y, <span class="at">method =</span> <span class="st">"spearman"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(<span class="at">x =</span> biochem<span class="sc">$</span>tot_cholesterol, <span class="at">y =</span> biochem<span class="sc">$</span>weight, <span class="at">method =</span> <span class="st">"spearman"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3596281</code></pre>
</div>
</div>
</section>
</section>
<section id="regression" class="level2">
<h2 class="anchored" data-anchor-id="regression">Regression</h2>
</section>
<section id="equation-for-a-line" class="level2">
<h2 class="anchored" data-anchor-id="equation-for-a-line">Equation for a line</h2>
<p>Remember:<br>
<span class="math inline">\(y = a \cdot x + b\)</span><br>
OR<br>
<span class="math inline">\(y = b + a \cdot x\)</span></p>
<p><span class="math inline">\(a\)</span> is the <strong>SLOPE</strong><br>
<span class="math inline">\(b\)</span> is the <strong>y-intercept</strong></p>
<div class="cell">
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="192"></p>
</div>
</div>
<p><span class="math inline">\(a\)</span> = 2 (the slope)<br>
<span class="math inline">\(b\)</span> = 0.5 (the y-intercept)</p>
<hr>
</section>
<section id="stats-equation-for-a-line" class="level2">
<h2 class="anchored" data-anchor-id="stats-equation-for-a-line">Stats equation for a line</h2>
<p>Model: the recipe for <span class="math inline">\(y\)</span> is a slope (<span class="math inline">\(\beta_1\)</span>) times <span class="math inline">\(x\)</span> plus an intercept (<span class="math inline">\(\beta_0\)</span>).</p>
<p><span class="math inline">\(y = \beta_0 + \beta_1 x \qquad \qquad \mathcal{H}_0: \beta_1 = 0\)</span></p>
<p>… which is the same has <span class="math inline">\(y = a \cdot x + b\)</span> (here ordered as <span class="math inline">\(y = b + a \cdot x\)</span>). In R we are lazy and write <code>y ~ 1 + x</code> which R reads like <code>y = 1*number + x*othernumber</code> and the task of t-tests, lm, etc., is simply to find the numbers that best predict <span class="math inline">\(y\)</span>.</p>
<p>Either way you write it, it’s an intercept (<span class="math inline">\(\beta_0\)</span>) and a slope (<span class="math inline">\(\beta_1\)</span>) yielding a straight line:</p>
<div class="cell">
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="192"></p>
</div>
</div>
<p><span class="math inline">\(\beta_0\)</span> = 0.5 (the y-intercept)<br>
<span class="math inline">\(\beta_1\)</span> = 2 (the slope)</p>
<p><span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x\)</span><br>
<span class="math inline">\(y = .5 \cdot 1 + 2 \cdot x\)</span></p>
<p>Our mission: <strong>FIND THE BEST</strong> <span class="math inline">\(\beta\)</span> coefficients</p>
<hr>
</section>
<section id="linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="linear-regression">Linear Regression</h2>
<ul>
<li>STEP 1: Make a scatter plot visualize the linear relationship between x and y.</li>
<li>STEP 2: Perform the regression</li>
<li>STEP 3: Look at the <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value</li>
<li>STEP 4: Visualize fit and errors</li>
<li>STEP 5: Calculate <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value ourselves</li>
</ul>
<hr>
<section id="step-1-can-mouse-cholesterol-levels-help-explain-mouse-weight" class="level3">
<h3 class="anchored" data-anchor-id="step-1-can-mouse-cholesterol-levels-help-explain-mouse-weight">STEP 1: Can mouse cholesterol levels help explain mouse weight?</h3>
<p>Plot weight (y, response variable) and cholesterol levels (x, explanatory variable) of the mice.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> tot_cholesterol)) <span class="sc">+</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
<section id="step-2-do-the-regression" class="level3">
<h3 class="anchored" data-anchor-id="step-2-do-the-regression">STEP 2: Do the regression</h3>
<p>Let’s fit a line (linear model).</p>
<p>Remember: <span class="math inline">\(y = \beta_0 \cdot 1+ \beta_1 \cdot x\)</span></p>
<p>linear model equation: <span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol\)</span></p>
<p><span class="math inline">\(\mathcal{H}_0:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> does NOT explain <span class="math inline">\(weight\)</span></p>
<p>Null Hypothesis: <span class="math inline">\(\mathcal{H}_0: \beta_1 = 0\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + 0 \cdot cholesterol\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1\)</span></p>
<p>Alternative Hypothesis: <span class="math inline">\(\mathcal{H}_1: \beta_1 \neq 0\)</span><br>
</p>
<p>Full model: <span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x\)</span></p>
<p><span class="math inline">\(\mathcal{H}_1:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> does explain <span class="math inline">\(weight\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol\)</span></p>
<p>The cool thing here is that we can assess and compare our null and alternative hypothesis by learning and examining the model coefficients (intercept and slope). Essentially, we are comparing a complex model (including cholesterol) to a simple model (weight).</p>
<p><a href="https://statisticsbyjim.com/regression/interpret-constant-y-intercept-regression/" class="uri">https://statisticsbyjim.com/regression/interpret-constant-y-intercept-regression/</a></p>
</section>
<section id="step-4-look-at-the-r2-f-value-and-p-value" class="level3">
<h3 class="anchored" data-anchor-id="step-4-look-at-the-r2-f-value-and-p-value">STEP 4: Look at the <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># fitting a line</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>fit_WvC <span class="ot">&lt;-</span> <span class="fu">lm</span>(</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> biochem,</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> tot_cholesterol</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb36-8"><a href="#cb36-8" aria-hidden="true" tabindex="-1"></a><span class="co"># base R summary of fit</span></span>
<span id="cb36-9"><a href="#cb36-9" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ 1 + tot_cholesterol, data = biochem)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.9624 -2.1349 -0.2627  2.0113 10.2927 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      14.5560     0.3635   40.04   &lt;2e-16 ***
tot_cholesterol   1.8516     0.1159   15.97   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.121 on 1780 degrees of freedom
Multiple R-squared:  0.1254,    Adjusted R-squared:  0.1249 
F-statistic: 255.1 on 1 and 1780 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>That’s a lot of info, but how would I access it? Time to meet your new best friend — <a href="&quot;https://cran.r-project.org/web/packages/broom/vignettes/broom.html&quot;">Broom</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># information about the model fit</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">glance</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 x 12
  r.squared adj.r.squared sigma statistic  p.value    df logLik   AIC   BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     0.125         0.125  3.12      255. 8.92e-54     1 -4556. 9117. 9134.
# i 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># information about the intercept and coefficients</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a><span class="fu">tidy</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 x 5
  term            estimate std.error statistic   p.value
  &lt;chr&gt;              &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)        14.6      0.363      40.0 1.48e-250
2 tot_cholesterol     1.85     0.116      16.0 8.92e- 54</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>chol_intercept <span class="ot">&lt;-</span> <span class="fu">tidy</span>(fit_WvC)[<span class="dv">1</span>, <span class="dv">2</span>]</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>chol_slope <span class="ot">&lt;-</span> <span class="fu">tidy</span>(fit_WvC)[<span class="dv">2</span>, <span class="dv">2</span>]</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a><span class="co"># for every 1 unit increase in cholesterol there is a 1.85 unit increase weight</span></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="co"># add residuals and other information</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="fu">augment</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,782 x 8
   weight tot_cholesterol .fitted .resid     .hat .sigma     .cooksd .std.resid
    &lt;dbl&gt;           &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;
 1   20.3            3.01    20.1  0.171 0.000566   3.12 0.000000846     0.0547
 2   16.1            2.46    19.1 -3.01  0.00107    3.12 0.000501       -0.965 
 3   19.5            3.57    21.2 -1.67  0.000906   3.12 0.000129       -0.534 
 4   22.2            2.61    19.4  2.81  0.000853   3.12 0.000347        0.901 
 5   17.3            2.04    18.3 -1.03  0.00203    3.12 0.000111       -0.331 
 6   18.1            2.86    19.9 -1.75  0.000622   3.12 0.0000981      -0.561 
 7   25.6            3.22    20.5  5.08  0.000592   3.12 0.000786        1.63  
 8   18.6            3.47    21.0 -2.38  0.000782   3.12 0.000228       -0.763 
 9   23.1            3.35    20.8  2.34  0.000669   3.12 0.000189        0.750 
10   17.3            2.29    18.8 -1.50  0.00140    3.12 0.000161       -0.480 
# i 1,772 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add residuals and other information into the biochem object</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>biochem_WvC <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_WvC, <span class="at">data =</span> biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For every 1 unit increase in <span class="math inline">\(cholesterol\)</span> there is a 1.8516281 increase in <span class="math inline">\(weight\)</span>.</p>
<hr>
</section>
<section id="step-5-visualize-fit-and-errors" class="level3">
<h3 class="anchored" data-anchor-id="step-5-visualize-fit-and-errors">STEP 5: Visualize fit and errors</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> tot_cholesterol)) <span class="sc">+</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>, <span class="at">col =</span> <span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="fu">pull</span>(chol_intercept), <span class="at">slope =</span> <span class="fu">pull</span>(chol_slope), <span class="at">col =</span> <span class="st">"pink"</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> lm, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">se =</span> F, <span class="at">size =</span> <span class="dv">1</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
i Please use `linewidth` instead.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/visualize lm-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb48-2"><a href="#cb48-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span></span>
<span id="cb48-3"><a href="#cb48-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="fu">pull</span>(chol_intercept), <span class="at">slope =</span> <span class="fu">pull</span>(chol_slope), <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb48-4"><a href="#cb48-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code</span></span>
<span id="cb48-5"><a href="#cb48-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">1</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb48-6"><a href="#cb48-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/visualize lm-2.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
<p>STEP 6: Calculate <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value ourselves</p>
<section id="definitions" class="level4">
<h4 class="anchored" data-anchor-id="definitions">DEFINITIONS</h4>
<p><span class="math inline">\(SS_{mean}\)</span> — sum of squared errors around the mean of <span class="math inline">\(y\)</span></p>
<p><span class="math inline">\(SS_{mean} = \sum_{i=1}^{n} (data - mean)^2\)</span><br>
<span class="math inline">\(SS_{mean} = \sum_{i=1}^{n} (y_{i} - \overline{y})^2\)</span></p>
<p><span class="math inline">\(Var_{mean}\)</span> — think of it like the average of the sum of squared error around the mean of <span class="math inline">\(y\)</span></p>
<p><span class="math inline">\(Var_{mean} = \displaystyle \frac {\sum_{i=1}^{n} (y_{i} - \overline{y})^2}{n}\)</span></p>
<p><span class="math inline">\(SS_{fit}\)</span> — sum of squared errors around the least-squares fit</p>
<p><span class="math inline">\(SS_{fit} = \sum_{i=1}^{n} (data - line)^2\)</span></p>
<p><span class="math inline">\(SS_{fit} = \sum_{i=1}^{n} (y_{i} - (\beta_0 \cdot 1+ \beta_1 \cdot x)^2\)</span></p>
<p><strong>Residuals</strong>, <span class="math inline">\(e\)</span> — the difference between the observed value of the dependent variable <span class="math inline">\(y\)</span> and the predicted value <span class="math inline">\(\widehat{y}\)</span> is called the residual. Each data point has one residual.</p>
<p><span class="math inline">\(e = y_{i} - \widehat{y}\)</span></p>
<p><span class="math inline">\(Var_{fit}\)</span> — think of it like the average of the sum of squared errors around the least-squares fit</p>
<p><span class="math inline">\(Var_{fit} = \displaystyle \frac {\sum_{i=1}^{n} (y_{i} - (\beta_0 \cdot 1+ \beta_1 \cdot x))}{n}\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>fit_W <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span>, <span class="at">data =</span> biochem)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit_W)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ 1, data = biochem)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.3409 -2.4409 -0.4409  2.2591  9.9591 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) 20.24091    0.07903   256.1   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.336 on 1781 degrees of freedom</code></pre>
</div>
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>biochem_W <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_W, <span class="at">data =</span> biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="super-nice-way-of-visualizing" class="level4">
<h4 class="anchored" data-anchor-id="super-nice-way-of-visualizing">Super nice way of <a href="https://drsimonj.svbtle.com/visualising-residuals">visualizing</a></h4>
<p>First, let’s look at <span class="math inline">\(SS_{mean}\)</span> for avg weight:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>p_W <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem_W, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> biochem_W<span class="sc">$</span>.fitted, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">size =</span> .<span class="dv">5</span>) <span class="sc">+</span> <span class="co"># plot linear model fit</span></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> <span class="co"># plot height as points and color code by the value of the residual</span></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">25</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s look at <span class="math inline">\(SS_{fit}\)</span> for the <span class="math inline">\(weight\)</span> ~ <span class="math inline">\(cholesterol\)</span> :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>p_WvC <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="fu">pull</span>(chol_intercept), <span class="at">slope =</span> <span class="fu">pull</span>(chol_slope), <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> <span class="co"># plot height as points and color code by the value of the residual</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># guides(color = FALSE) + # no legend for color code</span></span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">1</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_grid</span>(p_W, p_WvC, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"weight by intercept"</span>, <span class="st">"weight by cholesterol"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid" width="960"></p>
</div>
</div>
<hr>
</section>
<section id="for-which-mice-does-the-model-fit-perform-the-most-poorly" class="level4">
<h4 class="anchored" data-anchor-id="for-which-mice-does-the-model-fit-perform-the-most-poorly">For which mice does the model fit perform the most poorly?</h4>
<p>Gotta check residuals!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># make new variable exception = absolute value of resid &gt; 9 then subject id</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>biochem_WvC<span class="sc">$</span>exceptions <span class="ot">&lt;-</span> <span class="fu">if_else</span>(</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">condition =</span> <span class="fu">abs</span>(biochem_WvC<span class="sc">$</span>.resid) <span class="sc">&lt;</span> <span class="dv">9</span>,</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">true =</span> <span class="st">""</span>,</span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">false =</span> biochem_WvC<span class="sc">$</span>subject_name</span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb54-10"><a href="#cb54-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-11"><a href="#cb54-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight, <span class="at">label =</span> exceptions)) <span class="sc">+</span></span>
<span id="cb54-12"><a href="#cb54-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="fu">ifelse</span>(biochem_WvC<span class="sc">$</span>exceptions <span class="sc">==</span> <span class="st">""</span>, <span class="st">"grey50"</span>, <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb54-13"><a href="#cb54-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>() <span class="sc">+</span></span>
<span id="cb54-14"><a href="#cb54-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> <span class="co"># plot height as points and color code by the value of the residual</span></span>
<span id="cb54-15"><a href="#cb54-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb54-16"><a href="#cb54-16" aria-hidden="true" tabindex="-1"></a>  <span class="co"># guides(color = FALSE) + # no legend for color code</span></span>
<span id="cb54-17"><a href="#cb54-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">1</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb54-18"><a href="#cb54-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
<section id="big-baby-and-kryptonate" class="level4">
<h4 class="anchored" data-anchor-id="big-baby-and-kryptonate">Big Baby and KryptoNate</h4>
<p>Nate Robinson aka “KryptoNate”<br>
5’ 9”<br>
180 lbs</p>
<p><img src="img/Baby_Nate.jpg" class="img-fluid" style="width:15.0%"></p>
<p>Glen Davis aka “Big Baby”<br>
6’ 9”<br>
280 lbs<br>
<img src="img/Baby_Nate2.jpg" class="img-fluid" style="width:15.0%"></p>
<hr>
<p><span class="math inline">\(R^2\)</span> or coefficient of determination — the proportion of the variance in the dependent variable that is predictable from the independent variable(s).</p>
<p><span class="math inline">\(R^2 = \displaystyle \frac {SS_{mean} - SS_{fit}}{SS_{mean}}\)</span></p>
<p><span class="math inline">\(R^2 = \displaystyle \frac {SS_{w} - SS_{wc}}{SS_{w}}\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>ss.fit <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_WvC<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>ss.mean <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_W<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb55-7"><a href="#cb55-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Calc R^2 value</span></span>
<span id="cb55-8"><a href="#cb55-8" aria-hidden="true" tabindex="-1"></a>biochem_WvC_rsq <span class="ot">&lt;-</span> (ss.mean <span class="sc">-</span> ss.fit) <span class="sc">/</span> ss.mean</span>
<span id="cb55-9"><a href="#cb55-9" aria-hidden="true" tabindex="-1"></a>biochem_WvC_rsq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1253678</code></pre>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glance</span>(fit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(r.squared)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1253678</code></pre>
</div>
</div>
</section>
<section id="interpretation-of-r2" class="level4">
<h4 class="anchored" data-anchor-id="interpretation-of-r2">Interpretation of <span class="math inline">\(R^2\)</span></h4>
<p>There is a 13% reduction in the variance when we take mouse <span class="math inline">\(cholesterol\)</span> into account<br>
OR<br>
Mouse <span class="math inline">\(cholesterol\)</span> explains 13% in player <span class="math inline">\(weight\)</span></p>
<p>By the way, this is the same <span class="math inline">\(R\)</span> as from the Pearson correlation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Pearson correlation R value</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(biochem<span class="sc">$</span>weight, biochem<span class="sc">$</span>tot_cholesterol, <span class="at">method =</span> <span class="st">"pearson"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3540731</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Pearson correlation R^2 value</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(biochem<span class="sc">$</span>weight, biochem<span class="sc">$</span>tot_cholesterol, <span class="at">method =</span> <span class="st">"pearson"</span>)<span class="sc">^</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1253678</code></pre>
</div>
</div>
<hr>
</section>
<section id="the-f-statistic" class="level4">
<h4 class="anchored" data-anchor-id="the-f-statistic">The F-statistic</h4>
<p><strong>F-statistic</strong> — the proportion of the variance in the dependent variable that is predictable from the independent variable(s).</p>
<p><span class="math inline">\(F = \displaystyle \frac{SS_{fit}/(p_{fit}-p_{mean})} {SS_{mean}/(n-p_{fit})}\)</span></p>
<p><span class="math inline">\(p_{fit}\)</span> — number of parameters in the fit line<br>
<span class="math inline">\(p_{mean}\)</span> — number of parameters in the mean line<br>
<span class="math inline">\(n\)</span> — number of data points</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="co"># F-value</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>biochem_WvC_F <span class="ot">&lt;-</span> ((ss.mean <span class="sc">-</span> ss.fit) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">-</span> <span class="dv">1</span>)) <span class="sc">/</span></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>  (ss.fit <span class="sc">/</span> (<span class="fu">nrow</span>(biochem_WvC) <span class="sc">-</span> <span class="dv">2</span>))</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>biochem_WvC_F</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 255.1411</code></pre>
</div>
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glance</span>(fit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(statistic)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   value 
255.1411 </code></pre>
</div>
</div>
</section>
<section id="p-value-from-the-f-statistic" class="level4">
<h4 class="anchored" data-anchor-id="p-value-from-the-f-statistic">P-value from the F-statistic</h4>
<p>We need to generate a null distribution of <span class="math inline">\(F-statistic\)</span> values to compare to our observed <span class="math inline">\(F-statistic\)</span>.</p>
<p>Therefore, we will randomize the player Height and Weight and then calculate the <span class="math inline">\(F-statistic\)</span>.</p>
<p><img src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" class="img-fluid" style="width:10.0%"></p>
<p><br></p>
<p>We will do this many many times to generate a null distribution of <span class="math inline">\(F-statistic\)</span>s.<br>
<br></p>
<p><img src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" class="img-fluid" style="width:5.0%"> <img src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" class="img-fluid" style="width:5.0%"><br>
<img src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" class="img-fluid" style="width:5.0%"> <img src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" class="img-fluid" style="width:5.0%"> <br></p>
<p>The p-value will be the probability of obtaining an <span class="math inline">\(F-statistic\)</span> in the null distribution at least as extreme as our observed <span class="math inline">\(F-statistic\)</span>.</p>
<p>Another beautiful Statquest explaining <a href="https://www.youtube.com/watch?v=nk2CQITm_eo">how to go from F-statistic to p-value</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set up an empty tibble to hold our null distribution</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>fake_biochem <span class="ot">&lt;-</span> <span class="fu">tribble</span>()</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-4"><a href="#cb67-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb67-5"><a href="#cb67-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sample function to randomize/permute data</span></span>
<span id="cb67-6"><a href="#cb67-6" aria-hidden="true" tabindex="-1"></a><span class="fu">sample</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  3  8  2  9  1  6  7 10  5  4</code></pre>
</div>
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># we will perform 100 permutations</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>myPerms <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>myPerms) {</span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">bind_cols</span>(</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>    biochem_WvC[<span class="fu">sample</span>(<span class="fu">nrow</span>(biochem_WvC)), <span class="st">"weight"</span>],</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a>    biochem_WvC[<span class="fu">sample</span>(<span class="fu">nrow</span>(biochem_WvC)), <span class="st">"tot_cholesterol"</span>],</span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>    <span class="st">"perm"</span> <span class="ot">=</span> <span class="fu">factor</span>(<span class="fu">rep</span>(i, <span class="fu">nrow</span>(biochem_WvC)))</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>  fake_biochem <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(fake_biochem, tmp)</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rm</span>(tmp)</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a><span class="co"># let's look at permutations 1 and 2</span></span>
<span id="cb69-17"><a href="#cb69-17" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(fake_biochem <span class="sc">%&gt;%</span> <span class="fu">filter</span>(perm <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)), <span class="fu">aes</span>(<span class="at">x =</span> weight, <span class="at">y =</span> tot_cholesterol, <span class="at">color =</span> perm)) <span class="sc">+</span></span>
<span id="cb69-18"><a href="#cb69-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb69-19"><a href="#cb69-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="remember-your-best-friend" class="level4">
<h4 class="anchored" data-anchor-id="remember-your-best-friend">Remember your best friend</h4>
<p><a href="https://cran.r-project.org/web/packages/broom/vignettes/broom_and_dplyr.html">BROOM</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># here we will calculate and extract linear model results for each permutation individualy using nest, mutate, and map functions</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>fake_biochem_lms <span class="ot">&lt;-</span> fake_biochem <span class="sc">%&gt;%</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">nest</span>(<span class="at">data =</span> <span class="sc">-</span>perm) <span class="sc">%&gt;%</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">fit =</span> <span class="fu">map</span>(data, <span class="sc">~</span> <span class="fu">lm</span>(weight <span class="sc">~</span> tot_cholesterol, <span class="at">data =</span> .x)),</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">glanced =</span> <span class="fu">map</span>(fit, glance)</span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unnest</span>(glanced)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="lets-take-a-look-at-the-null-distribution-of-f-statistics-from-the-randomized-values" class="level4">
<h4 class="anchored" data-anchor-id="lets-take-a-look-at-the-null-distribution-of-f-statistics-from-the-randomized-values">Let’s take a look at the null distribution of F-statistics from the randomized values</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>fake_biochem_lms <span class="sc">%&gt;%</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(<span class="at">x =</span> statistic)) <span class="sc">+</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>remember that the <span class="math inline">\(F-statistic\)</span> we observed was ~255!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>fake_biochem_lms <span class="sc">%&gt;%</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(<span class="at">x =</span> statistic)) <span class="sc">+</span></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>, biochem_WvC_F <span class="sc">*</span> <span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>(<span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> biochem_WvC_F, <span class="at">color =</span> <span class="st">"blue"</span>) <span class="sc">+</span></span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>  <span class="co">#  scale_x_log10() +</span></span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-16-1.png" class="img-fluid" width="672"></p>
</div>
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glance</span>(fit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       value 
8.924155e-54 </code></pre>
</div>
</div>
</section>
<section id="interpretation-of-p-value" class="level4">
<h4 class="anchored" data-anchor-id="interpretation-of-p-value">Interpretation of <span class="math inline">\(p-value\)</span></h4>
<p>There is no value more extreme than our observed <span class="math inline">\(F-statistic\)</span>.<br>
Therefore, the empirical <span class="math inline">\(p-value &lt; 0.001\)</span> — empirical what we calculated by randomizing our data.</p>
<p>The exact <span class="math inline">\(p-value\)</span> is 8.9241551^{-54} AKA <em>“a ridiculously small number”</em>.</p>
<p><strong>Correct</strong> — Assuming that <span class="math inline">\(cholesterol\)</span> has zero effect on <span class="math inline">\(weight\)</span> in the population, you’d obtain the sample effect, or larger, in 8.9241551^{-54} AKA <em>“a ridiculously small number”</em> of studies because of random sample error.</p>
<p><strong>Incorrect</strong> —There’s a 8.9241551^{-54} AKA <em>“a ridiculously small number”</em> chance of making a mistake by rejecting the null hypothesis.</p>
<p>More on <a href="https://statisticsbyjim.com/hypothesis-testing/interpreting-p-values/">p-value (mis)interpretation</a></p>
<hr>
</section>
</section>
<section id="how-to-find-the-best-least-squares-fit" class="level3">
<h3 class="anchored" data-anchor-id="how-to-find-the-best-least-squares-fit">How to find the best (least squares) fit?</h3>
<ol type="1">
<li>Rotate the line of fit<br>
</li>
<li>Find the fit that minimizes the Sum of Squared Residuals or <span class="math inline">\(SS_{fit}\)</span><br>
</li>
<li>This is the derivative (slope of tangent at best point = 0) of the function describing the <span class="math inline">\(SS_{fit}\)</span> and the next rotation is 0.</li>
</ol>
<p><a href="https://youtu.be/PaFPbb66DxQ">StatQuest: Fitting a line to data, aka least squares, aka linear regression.</a></p>
<p><a href="https://youtu.be/sDv4f4s2SB8">StatQuest: Gradient Descent, Step-by-Step</a></p>
<hr>
</section>
<section id="non-parametric-version" class="level3">
<h3 class="anchored" data-anchor-id="non-parametric-version">Non-parametric version</h3>
<section id="rank" class="level4">
<h4 class="anchored" data-anchor-id="rank">Theory: rank-transformation</h4>
<p><code>rank</code> simply takes a list of numbers and “replace” them with the integers of their rank (1st smallest, 2nd smallest, 3rd smallest, etc.). So the result of the rank-transformation <code>rank(c(3.6, 3.4, -5.0, 8.2))</code> is <code>3, 2, 1, 4</code>. See that in the figure above?</p>
<p>A <em>signed</em> rank is the same, just where we rank according to absolute size first and then add in the sign second. So the signed rank here would be <code>2, 1, -3, 4</code>. Or in code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>npfit_WvC <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> <span class="fu">rank</span>(weight) <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fu">rank</span>(tot_cholesterol), <span class="at">data =</span> biochem)</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="fu">tidy</span>(npfit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 x 5
  term                  estimate std.error statistic   p.value
  &lt;chr&gt;                    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)            571.      22.8         25.1 4.01e-119
2 rank(tot_cholesterol)    0.360    0.0221      16.3 1.54e- 55</code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">glance</span>(npfit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(r.squared)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1293324</code></pre>
</div>
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="co"># cor(biochem$weight, biochem$tot_cholesterol, method = "spearman")^2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
</section>
<section id="multiple-regression" class="level2">
<h2 class="anchored" data-anchor-id="multiple-regression">Multiple regression</h2>
<p>Remember:</p>
<p><span class="math inline">\(y = \beta_0 \cdot 1+ \beta_1 \cdot x\)</span></p>
<p>Let’s add an explanatory variable:</p>
<p><span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x + \beta_2 \cdot z\)</span></p>
<p>linear model equation: <span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol + \beta_2 \cdot sodium\)</span></p>
<p><span class="math inline">\(\mathcal{H}_0:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> and <span class="math inline">\(sodium\)</span> does NOT explain <span class="math inline">\(weight\)</span></p>
<p>Null Hypothesis: <span class="math inline">\(\mathcal{H}_0: \beta_1, \beta_2 = 0\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + 0 \cdot cholesterol + 0 \cdot sodium\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1\)</span></p>
<p>Alternative Hypothesis: <span class="math inline">\(\mathcal{H}_1: \beta_1,\beta_2 \neq 0\)</span><br>
</p>
<p>Full model: <span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x + \beta_2 \cdot z\)</span><br>
<span class="math inline">\(\mathcal{H}_1:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> and <span class="math inline">\(sodium\)</span> does explain <span class="math inline">\(weight\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol + \beta_2 \cdot sodium\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="co"># how do different variables correlate with weight?</span></span>
<span id="cb80-2"><a href="#cb80-2" aria-hidden="true" tabindex="-1"></a>biochem <span class="sc">%&gt;%</span></span>
<span id="cb80-3"><a href="#cb80-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select_if</span>(is.numeric) <span class="sc">%&gt;%</span></span>
<span id="cb80-4"><a href="#cb80-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">cor</span>() <span class="sc">%&gt;%</span></span>
<span id="cb80-5"><a href="#cb80-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="sc">%&gt;%</span></span>
<span id="cb80-6"><a href="#cb80-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(weight) <span class="sc">%&gt;%</span></span>
<span id="cb80-7"><a href="#cb80-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(<span class="sc">-</span>weight) <span class="sc">%&gt;%</span></span>
<span id="cb80-8"><a href="#cb80-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "weight"          "tot_cholesterol" "age"             "sodium"         
[5] "glucose"         "calcium"         "litter"          "cage_density"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="co"># does sodium + cholesterol predict weight better than cholesterol alone?</span></span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>fit_WvC_S <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">data =</span> biochem, <span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> tot_cholesterol <span class="sc">+</span> sodium)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>fit_WvC_S <span class="sc">%&gt;%</span> <span class="fu">glance</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 x 12
  r.squared adj.r.squared sigma statistic  p.value    df logLik   AIC   BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     0.171         0.170  3.04      183. 3.72e-73     2 -4508. 9024. 9046.
# i 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>biochem_WvC_S <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_WvC_S, <span class="at">data =</span> biochem)</span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>ss.fit_C_S <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_WvC_S<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>var.fit_C_S <span class="ot">&lt;-</span> ss.fit_C_S <span class="sc">/</span> <span class="fu">nrow</span>(biochem_WvC_S)</span>
<span id="cb84-6"><a href="#cb84-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-7"><a href="#cb84-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-8"><a href="#cb84-8" aria-hidden="true" tabindex="-1"></a>ss.fit_C <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_WvC<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb84-9"><a href="#cb84-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-10"><a href="#cb84-10" aria-hidden="true" tabindex="-1"></a>var.fit_C <span class="ot">&lt;-</span> ss.fit_C <span class="sc">/</span> <span class="fu">nrow</span>(biochem_WvC)</span>
<span id="cb84-11"><a href="#cb84-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-12"><a href="#cb84-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-13"><a href="#cb84-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-14"><a href="#cb84-14" aria-hidden="true" tabindex="-1"></a>fit_Wvall <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">data =</span> biochem, <span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> tot_cholesterol <span class="sc">+</span> age <span class="sc">+</span> sodium <span class="sc">+</span> glucose <span class="sc">+</span> calcium <span class="sc">+</span> litter <span class="sc">+</span> cage_density)</span>
<span id="cb84-15"><a href="#cb84-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb84-16"><a href="#cb84-16" aria-hidden="true" tabindex="-1"></a>fit_Wvall <span class="sc">%&gt;%</span> <span class="fu">glance</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 x 12
  r.squared adj.r.squared sigma statistic   p.value    df logLik   AIC   BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     0.253         0.250  2.89      85.7 1.33e-107     7 -4415. 8849. 8898.
# i 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>fit_Wvall <span class="sc">%&gt;%</span> <span class="fu">tidy</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 x 5
  term            estimate std.error statistic  p.value
  &lt;chr&gt;              &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      -6.53      1.68       -3.90 9.98e- 5
2 tot_cholesterol   1.97      0.109      18.2  9.08e-68
3 age               0.133     0.0161      8.28 2.49e-16
4 sodium            0.179     0.0145     12.4  1.12e-33
5 glucose           0.159     0.0284      5.59 2.59e- 8
6 calcium          -5.65      0.692      -8.16 6.13e-16
7 litter           -0.0557    0.0523     -1.06 2.88e- 1
8 cage_density     -0.316     0.0648     -4.87 1.22e- 6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> calcium)) <span class="sc">+</span></span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/multiple regression-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
<section id="non-linear-regression" class="level2">
<h2 class="anchored" data-anchor-id="non-linear-regression">Non-linear regression</h2>
<p><strong>Non-linear regression</strong> — observational data are modeled by a function which is a nonlinear combination of the model parameters and depends on one or more independent variables. The data are fitted by a method of successive approximations.</p>
<p><strong>DANGER OVERFITTING</strong></p>
<section id="perform-loess-regression" class="level4">
<h4 class="anchored" data-anchor-id="perform-loess-regression">Perform loess regression</h4>
<p><strong>Loess regression</strong> — a non-parametric technique that uses local weighted regression to fit a smooth curve through points in a scatter plot. LOESS combines much of the simplicity of linear least squares regression with the flexibility of nonlinear regression. It does this by fitting simple models to localized subsets of the data to build up a function that describes the deterministic part of the variation in the data, point by point. In fact, one of the chief attractions of this method is that the data analyst is not required to specify a global function of any form to fit a model to the data, only to fit segments of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="co"># local-weighted regression fit</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>loessfit_WvC <span class="ot">&lt;-</span> <span class="fu">loess</span>(<span class="at">formula =</span> weight <span class="sc">~</span> tot_cholesterol, <span class="at">data =</span> biochem)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>loess_biochem_WvC <span class="ot">&lt;-</span> <span class="fu">augment</span>(loessfit_WvC, biochem)</span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(loessfit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call:
loess(formula = weight ~ tot_cholesterol, data = biochem)

Number of Observations: 1782 
Equivalent Number of Parameters: 5.24 
Residual Standard Error: 3.112 
Trace of smoother matrix: 5.73  (exact)

Control settings:
  span     :  0.75 
  degree   :  2 
  family   :  gaussian
  surface  :  interpolate     cell = 0.2
  normalize:  TRUE
 parametric:  FALSE
drop.square:  FALSE </code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ 1 + tot_cholesterol, data = biochem)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.9624 -2.1349 -0.2627  2.0113 10.2927 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      14.5560     0.3635   40.04   &lt;2e-16 ***
tot_cholesterol   1.8516     0.1159   15.97   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.121 on 1780 degrees of freedom
Multiple R-squared:  0.1254,    Adjusted R-squared:  0.1249 
F-statistic: 255.1 on 1 and 1780 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
</section>
<section id="plot-loess-fit-depicting-residuals" class="level4">
<h4 class="anchored" data-anchor-id="plot-loess-fit-depicting-residuals">Plot loess fit depicting residuals</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> loess_biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> lm, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">se =</span> F, <span class="at">size =</span> .<span class="dv">25</span>, <span class="at">linetype =</span> <span class="st">"dashed"</span>) <span class="sc">+</span> <span class="co"># linear fit black dashed line</span></span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_smooth</span>(<span class="at">method =</span> loess, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">se =</span> F, <span class="at">size =</span> .<span class="dv">25</span>) <span class="sc">+</span> <span class="co"># loess fit red line</span></span>
<span id="cb93-4"><a href="#cb93-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span></span>
<span id="cb93-5"><a href="#cb93-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span></span>
<span id="cb93-6"><a href="#cb93-6" aria-hidden="true" tabindex="-1"></a>  <span class="co"># guides(color = FALSE) +</span></span>
<span id="cb93-7"><a href="#cb93-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">25</span>) <span class="sc">+</span></span>
<span id="cb93-8"><a href="#cb93-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>`geom_smooth()` using formula = 'y ~ x'
`geom_smooth()` using formula = 'y ~ x'</code></pre>
</div>
<div class="cell-output-display">
<p><img src="class-03_files/figure-html/unnamed-chunk-19-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<hr>
</section>
</section>
</section>
<section id="references-and-resources" class="level1">
<h1>References and resources</h1>
<p>I have borrowed heavily from, directly taken, and/or highly recommend the following fantastic resources:</p>
<ol type="1">
<li><a href="https://lindeloev.github.io/tests-as-linear/">Common statistical tests are linear models</a> from Jonas Kristoffer Lindel<c3><b8>v<br>
</b8></c3></li>
<li><a href="https://www.youtube.com/channel/UCtYLUTtgS3k1Fg4y5tAhLbw">Statquest</a></li>
<li><a href="https://www.rapidtables.com/math/symbols/Statistical_Symbols.html">Stats gobbledygook</a></li>
<li><a href="http://www.sthda.com/english/articles/39-regression-model-diagnostics/161-linear-regression-assumptions-and-diagnostics-in-r-essentials/">Linear Regression Assumptions and Diagnostics in R: Essentials</a><br>
</li>
<li><a href="https://www.graphpad.com/guides/prism/latest/statistics/stat_---_principles_of_statistics_-.htm">PRINCIPLES OF STATISTICS</a> from GraphPad/SAS.</li>
</ol>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../../../content/bootcamp/stats/class-02.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Bootcamp Stats Day 2</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../../../content/bootcamp/stats/class-04.html" class="pagination-link">
        <span class="nav-page-text">Bootcamp Stats Day 4</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">This page is built with <a href="https://quarto.org/">Quarto</a>.</div>
  </div>
</footer>



</body></html>