<!DOCTYPE html>
<html lang="en"><head>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.361">

  <meta name="author" content="Neelanjan Mukherjee">
  <title>MOLB 7950 - Bootcamp: Stats class 5</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
<meta property="og:title" content="MOLB 7950 - Bootcamp: Stats class 5">
<meta property="og:site-name" content="MOLB 7950">
<meta name="twitter:title" content="MOLB 7950 - Bootcamp: Stats class 5">
<meta name="twitter:image" content="https://rnabioco.github.io/molb-7950/content/bootcamp/stats/images/twitter-card.png">
<meta name="twitter:creator" content="@rnabioco">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Bootcamp: Stats class 5</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
<a href="neelanjan.mukherjee@cuanschutz.edu">Neelanjan Mukherjee</a> 
</div>
</div>
</div>

  <p class="date">Invalid Date</p>
</section>
<section class="slide level2">

</section>
<section>
<section id="learning-objectives" class="title-slide slide level1 center">
<h1>Learning Objectives</h1>
<ul>
<li><strong>Visualize</strong> and <strong>Summarize</strong> the data being compared</li>
<li><strong>Formulate</strong> and <strong>Execute</strong> null hypothesis testing</li>
<li><strong>Identify</strong> and <strong>Perform</strong> the proper statistical test for data type/comparison</li>
<li><strong>Calculate</strong> and <strong>Interpret</strong> p-values</li>
<li><strong>Prevent</strong> p-hacking and <strong>Recognize</strong> issues with simultaneously testing multiple hypotheses.</li>
</ul>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="outline" class="title-slide slide level1 center">
<h1>Outline</h1>
<ul>
<li>Concepts and Definitions
<ul>
<li>Simplifying principles: Common tests as linear models</li>
<li>Types of comparisons and statistical tests</li>
<li>Definitions<br>
</li>
</ul></li>
<li>Day 1: Relationship between two or more continuous variables
<ul>
<li>Correlation vs Regression</li>
<li>Fitting it a line to data</li>
<li>Linear regression concepts</li>
<li>Multiple regression<br>
</li>
</ul></li>
<li>Day 2: Relationship between categorical and continuous variables
<ul>
<li><p>Comparing means between two groups (t-Test)</p></li>
<li><p>Comparing means between 3 or more groups (ANOVA)<strong><br>
</strong></p></li>
</ul></li>
<li><strong>Day 3: Multiple test correction, Bayesian intro, History</strong></li>
</ul>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="multiple-testing-correction" class="title-slide slide level1 center">
<h1>Multiple testing correction</h1>

</section>
<section id="genomics---lots-of-data---lots-of-hypothesis-tests" class="slide level2">
<h2>Genomics -&gt; Lots of Data -&gt; Lots of Hypothesis Tests</h2>
<p>In a typical RNA-seq experiment, we test ~10K different hypotheses. For example, you have 10K genes and for each gene you test whether the mean expression changed in condition A vs condition B. Using a standard p-value cut-off of 0.05, we’d expect <strong>500 genes</strong> to be deemed “significant” by chance. Thus, we concerned about <strong>False Positives or Type I Errors</strong>.</p>
<p><img data-src="img/type-i-and-ii-error-2.png"></p>
<p>So, we want to control our type 1 error. We can approach this in two different ways.</p>
<ol type="1">
<li><p>Control overall α (also known as family-wise error rate or <a href="https://en.wikipedia.org/wiki/Family-wise_error_rate">FWER</a>), which will affect the α* for each test. That is, we are controlling the overall probability of making <em>at least one</em> false discovery. Bonferroni and Sidak corrections all control FWER.</p></li>
<li><p>Control <a href="https://en.wikipedia.org/wiki/False_discovery_rate">false discovery rate</a> (FDR). Where FWER controls for the probability for making a type 1 error <em>at all</em>, these procedures allow for type 1 errors (false positives) but control the proportion of these false positives in relation to true positives. This is done by adjusting the decision made for the p-value associated with each individual test to decide rejection or not. Because this will result in a higher type 1 error rate, it has higher <a href="https://en.wikipedia.org/wiki/Power_(statistics)">power</a>. This affords a higher probability of <em>true discoveries.</em> The step procedures control for FDR.</p></li>
</ol>
<h3 id="bonferroni-correction">Bonferroni Correction</h3>
<p>**The most conservative of corrections, the Bonferroni correction is also perhaps the most straightforward in its approach. Simply divide α by the number of tests (<em>m</em>).</p>
<p>However, with many tests, α* will become very small. This reduces power, which means that we are very unlikely to make any true discoveries.</p>
<h3 id="sidak-correction">Sidak Correction</h3>
<p>**α* = 1-(1-α)^(1/<em>m</em>)</p>
<h3 id="holms-step-down-procedure">Holm’s Step-Down Procedure</h3>
<p>**An update of the Bonferroni correction, this procedure is more powerful. Rather than controlling the FWER, Holm’s procedure controls for the <a href="https://en.wikipedia.org/wiki/False_discovery_rate">false discovery rate</a> (FDR) and is performed after conducting all hypothesis tests and finding associated p-values at α within a set.</p>
<p>The step-down procedure is best illustrated with an example. Say we have three hypotheses, each with the associated p-value:</p>
<p>H1: 0.025<br>
H2: 0.003<br>
H3: 0.01</p>
<p>Step 1: Order p-values from smallest to greatest</p>
<p>H2: 0.003<br>
H3: 0.01<br>
H1: 0.025</p>
<p>Step 2: Use the Holm-Bonferroni formula for the first-ranked (smallest) p-value</p>
<pre><code>α* = α/(n-rank+1)</code></pre>
<p>α* = 0.05/(3–1+1) = 0.0167</p>
<p>Step 3: Compare the first-ranked p-value with the α* calculated from Step 2</p>
<p>0.003 &lt; 0.0167</p>
<p>Because the p-value for H2 is less than the calculated α*, we can reject H2.</p>
<p>Move onto the next ranked p-value and repeat steps 2–3, calculating the α* for its respective rank and comparing it to that p-value. Continue until you reach the first non-rejected hypothesis. You would then fail to reject all following hypotheses.</p>
<h3 id="hochbergs-step-up-procedure">Hochberg’s Step-Up Procedure</h3>
<p>**More powerful than Holm’s step-down procedure, Hochberg’s step-up procedure also seeks to control the FDR and follows a similar process, only p-values are ranked from largest to smallest.</p>
<p>For each ranked p-value, it is compared to the α* calculated for its respective rank (same formula as Holm’s procedure). Testing continues until you reach the first non-rejected hypothesis. You would then fail to reject all following hypotheses.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>rna <span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="at">file =</span> <span class="st">"data/deltaRNA_test.csv"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>  <span class="fu">select</span>(gene_id, pvalue) <span class="sc">%&gt;%</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a>rna<span class="sc">$</span>fdr <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(<span class="at">p =</span> rna<span class="sc">$</span>pvalue, <span class="at">method =</span> <span class="st">"fdr"</span>, <span class="at">n =</span> <span class="fu">nrow</span>(rna))</span>
<span id="cb2-6"><a href="#cb2-6"></a></span>
<span id="cb2-7"><a href="#cb2-7"></a>rna<span class="sc">$</span>BH <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(<span class="at">p =</span> rna<span class="sc">$</span>pvalue, <span class="at">method =</span> <span class="st">"BH"</span>, <span class="at">n =</span> <span class="fu">nrow</span>(rna))</span>
<span id="cb2-8"><a href="#cb2-8"></a></span>
<span id="cb2-9"><a href="#cb2-9"></a>rna<span class="sc">$</span>bon <span class="ot">&lt;-</span> <span class="fu">p.adjust</span>(<span class="at">p =</span> rna<span class="sc">$</span>pvalue, <span class="at">method =</span> <span class="st">"bonferroni"</span>, <span class="at">n =</span> <span class="fu">nrow</span>(rna))</span>
<span id="cb2-10"><a href="#cb2-10"></a></span>
<span id="cb2-11"><a href="#cb2-11"></a></span>
<span id="cb2-12"><a href="#cb2-12"></a>rna_long <span class="ot">&lt;-</span> rna <span class="sc">%&gt;%</span> <span class="fu">pivot_longer</span>(<span class="at">cols =</span> pvalue<span class="sc">:</span>bon, <span class="at">names_to =</span> <span class="st">"type"</span>)</span>
<span id="cb2-13"><a href="#cb2-13"></a></span>
<span id="cb2-14"><a href="#cb2-14"></a></span>
<span id="cb2-15"><a href="#cb2-15"></a></span>
<span id="cb2-16"><a href="#cb2-16"></a>p_none <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> rna, <span class="fu">aes</span>(<span class="at">x =</span> pvalue, <span class="at">y =</span> pvalue)) <span class="sc">+</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb2-18"><a href="#cb2-18"></a>  <span class="fu">ggtitle</span>(<span class="st">"None"</span>) <span class="sc">+</span></span>
<span id="cb2-19"><a href="#cb2-19"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb2-20"><a href="#cb2-20"></a></span>
<span id="cb2-21"><a href="#cb2-21"></a></span>
<span id="cb2-22"><a href="#cb2-22"></a>p_FDR <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> rna, <span class="fu">aes</span>(<span class="at">x =</span> pvalue, <span class="at">y =</span> fdr)) <span class="sc">+</span></span>
<span id="cb2-23"><a href="#cb2-23"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb2-24"><a href="#cb2-24"></a>  <span class="fu">ggtitle</span>(<span class="st">"FDR"</span>) <span class="sc">+</span></span>
<span id="cb2-25"><a href="#cb2-25"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb2-26"><a href="#cb2-26"></a></span>
<span id="cb2-27"><a href="#cb2-27"></a>p_BH <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> rna, <span class="fu">aes</span>(<span class="at">x =</span> pvalue, <span class="at">y =</span> BH)) <span class="sc">+</span></span>
<span id="cb2-28"><a href="#cb2-28"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb2-29"><a href="#cb2-29"></a>  <span class="fu">ggtitle</span>(<span class="st">"BH"</span>) <span class="sc">+</span></span>
<span id="cb2-30"><a href="#cb2-30"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb2-31"><a href="#cb2-31"></a></span>
<span id="cb2-32"><a href="#cb2-32"></a>p_bon <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> rna, <span class="fu">aes</span>(<span class="at">x =</span> pvalue, <span class="at">y =</span> bon)) <span class="sc">+</span></span>
<span id="cb2-33"><a href="#cb2-33"></a>  <span class="fu">geom_point</span>(<span class="at">size =</span> .<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb2-34"><a href="#cb2-34"></a>  <span class="fu">ggtitle</span>(<span class="st">"Bonferroni"</span>) <span class="sc">+</span></span>
<span id="cb2-35"><a href="#cb2-35"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb2-36"><a href="#cb2-36"></a></span>
<span id="cb2-37"><a href="#cb2-37"></a><span class="fu">plot_grid</span>(p_none, p_bon, p_BH, p_FDR, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">nrow =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="stats-class-05_files/figure-revealjs/unnamed-chunk-1-1.png" width="960"></p>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> rna_long, <span class="fu">aes</span>(<span class="at">x =</span> value, <span class="at">color =</span> type)) <span class="sc">+</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>  <span class="fu">stat_ecdf</span>() <span class="sc">+</span></span>
<span id="cb3-3"><a href="#cb3-3"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb3-4"><a href="#cb3-4"></a>  <span class="fu">xlab</span>(<span class="st">"p-values"</span>) <span class="sc">+</span></span>
<span id="cb3-5"><a href="#cb3-5"></a>  <span class="fu">ylab</span>(<span class="st">"cumulative fraction"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="stats-class-05_files/figure-revealjs/unnamed-chunk-1-2.png" width="960"></p>
</div>
</div>
</section></section>
<section>
<section id="bayesian-statistics" class="title-slide slide level1 center">
<h1>Bayesian Statistics</h1>

</section>
<section id="bayes-rule" class="slide level2">
<h2>Bayes’ Rule</h2>
<p>The conditional probability of the event <span class="math inline">\(A\)</span> conditional on the event <span class="math inline">\(B\)</span> is given by <span class="math display">\[
  P(A \mid B) = \frac{P(A \,\&amp;\, B)}{P(B)}.
\]</span></p>
<p>This section introduces how the Bayes’ rule is applied to calculating conditional probability, and several real-life examples are demonstrated. Finally, we compare the Bayesian and frequentist definition of probability.</p>
<h3 id="bayes-rule-and-diagnostic-testing">Bayes’ Rule and Diagnostic Testing</h3>
<p>To better understand conditional probabilities and their importance, let us consider an example involving the human immunodeficiency virus (HIV). In the early 1980s, HIV had just been discovered and was rapidly expanding. There was major concern with the safety of the blood supply. Also, virtually no cure existed making an HIV diagnosis basically a death sentence, in addition to the stigma that was attached to the disease.</p>
<p>These made false positives and false negatives in HIV testing highly undesirable. A <strong>false positive</strong> is when a test returns postive while the truth is negative. That would for instance be that someone without HIV is wrongly diagnosed with HIV, wrongly telling that person they are going to die and casting the stigma on them. A <strong>false negative</strong> is when a test returns negative while the truth is positive. That is when someone with HIV undergoes an HIV test which wrongly comes back negative. The latter poses a threat to the blood supply if that person is about to donate blood.</p>
<p>The probability of a false positive if the truth is negative is called the false positive rate. Similarly, the false negative rate is the probability of a false negative if the truth is positive. Note that both these rates are conditional probabilities: The false positive rate of an HIV test is the probability of a positive result <strong>conditional on</strong> the person tested having no HIV.</p>
<p>The HIV test we consider is an enzyme-linked immunosorbent assay, commonly known as an ELISA. We would like to know the probability that someone (in the early 1980s) has HIV if ELISA tests positive. For this, we need the following information. ELISA’s true positive rate (one minus the false negative rate), also referred to as sensitivity, recall, or probability of detection, is estimated as</p>
<p><span class="math display">\[
  P(\text{ELISA is positive} \mid \text{Person tested has HIV}) = 93\% = 0.93.
\]</span></p>
<p>Its true negative rate (one minus the false positive rate), also referred to as specificity, is estimated as</p>
<p><span class="math display">\[
  P(\text{ELISA is negative} \mid \text{Person tested has no HIV}) = 99\% = 0.99.
\]</span></p>
<p>Also relevant to our question is the prevalence of HIV in the overall population, which is estimated to be 1.48 out of every 1000 American adults. We therefore assume</p>
<span class="math display">\[\begin{equation}
  P(\text{Person tested has HIV}) = \frac{1.48}{1000} = 0.00148.
  (\#eq:HIVpositive)
\end{equation}\]</span>
<p>Note that the above numbers are estimates. For our purposes, however, we will treat them as if they were exact.</p>
<p>Our goal is to compute the probability of HIV if ELISA is positive, that is <span class="math inline">\(P(\text{Person tested has HIV} \mid \text{ELISA is positive})\)</span>. In none of the above numbers did we condition on the outcome of ELISA. Fortunately, Bayes’ rule allows is to use the above numbers to compute the probability we seek. Bayes’ rule states that</p>
<span class="math display">\[\begin{equation}
  P(\text{Person tested has HIV}  \mid \text{ELISA is positive}) = \frac{P(\text{Person tested has HIV} \,\&amp;\, \text{ELISA is positive})}{P(\text{ELISA is positive})}.
   (\#eq:HIVconditional)
\end{equation}\]</span>
<p>This can be derived as follows. For someone to test positive and be HIV positive, that person first needs to be HIV positive and then secondly test positive. The probability of the first thing happening is <span class="math inline">\(P(\text{HIV positive}) = 0.00148\)</span>. The probability of then testing positive is <span class="math inline">\(P(\text{ELISA is positive} \mid \text{Person tested has HIV}) = 0.93\)</span>, the true positive rate. This yields for the numerator</p>
<span class="math display">\[\begin{multline}
  P(\text{Person tested has HIV} \,\&amp;\, \text{ELISA is positive}) \\
  \begin{split}
  &amp;= P(\text{Person tested has HIV}) P(\text{ELISA is positive} \mid \text{Person tested has HIV}) \\
  &amp;= 0.00148 \cdot 0.93
  = 0.0013764.
  \end{split}
  (\#eq:HIVjoint)
\end{multline}\]</span>
<p>The first step in the above equation is implied by Bayes’ rule: By multiplying the left- and right-hand side of Bayes’ rule by <span class="math inline">\(P(B)\)</span>, we obtain <span class="math display">\[
  P(A \mid B) P(B) = P(A \,\&amp;\, B).
\]</span></p>
<p>The denominator in @ref(eq:HIVconditional) can be expanded as</p>
<span class="math display">\[\begin{multline*}
  P(\text{ELISA is positive}) \\
  \begin{split}
  &amp;= P(\text{Person tested has HIV} \,\&amp;\, \text{ELISA is positive}) + P(\text{Person tested has no HIV} \,\&amp;\, \text{ELISA is positive}) \\
  &amp;= 0.0013764 + 0.0099852 = 0.0113616
  \end{split}
\end{multline*}\]</span>
<p>where we used @ref(eq:HIVjoint) and</p>
<span class="math display">\[\begin{multline*}
  P(\text{Person tested has no HIV} \,\&amp;\, \text{ELISA is positive}) \\
  \begin{split}
  &amp;= P(\text{Person tested has no HIV}) P(\text{ELISA is positive} \mid \text{Person tested has no HIV}) \\
  &amp;= \left(1 - P(\text{Person tested has HIV})\right) \cdot \left(1 - P(\text{ELISA is negative} \mid \text{Person tested has no HIV})\right) \\
  &amp;= \left(1 - 0.00148\right) \cdot \left(1 - 0.99\right) = 0.0099852.
  \end{split}
\end{multline*}\]</span>
<p>Putting this all together and inserting into @ref(eq:HIVconditional) reveals <span class="math display">\[\begin{equation}
  P(\text{Person tested has HIV} \mid \text{ELISA is positive}) = \frac{0.0013764}{0.0113616} \approx 0.12.
  (\#eq:HIVresult)
\end{equation}\]</span> So even when the ELISA returns positive, the probability of having HIV is only 12%. An important reason why this number is so low is due to the prevalence of HIV. Before testing, one’s probability of HIV was 0.148%, so the positive test changes that probability dramatically, but it is still below 50%. That is, it is more likely that one is HIV negative rather than positive after one positive ELISA test.</p>
<p>Questions like the one we just answered (What is the probability of a disease if a test returns positive?) are crucial to make medical diagnoses. As we saw, just the true positive and true negative rates of a test do not tell the full story, but also a disease’s prevalence plays a role. Bayes’ rule is a tool to synthesize such numbers into a more useful probability of having a disease after a test result.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode numberSource example number-lines code-with-copy"><code class="sourceCode"><span id="cb4-1"><a href="#cb4-1"></a>What is the probability that someone who tests positive does not actually have HIV?</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We found in @ref(eq:HIVresult) that someone who tests positive has a <span class="math inline">\(0.12\)</span> probability of having HIV. That implies that the same person has a <span class="math inline">\(1-0.12=0.88\)</span> probability of not having HIV, despite testing positive.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource example number-lines code-with-copy"><code class="sourceCode"><span id="cb5-1"><a href="#cb5-1"></a>If the individual is at a higher risk for having HIV than a randomly sampled person from the population considered, how, if at all, would you expect $P(\text{Person tested has HIV} \mid \text{ELISA is positive})$ to change?</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If the person has a priori a higher risk for HIV and tests positive, then the probability of having HIV must be higher than for someone not at increased risk who also tests positive. Therefore, <span class="math inline">\(P(\text{Person tested has HIV} \mid \text{ELISA is positive}) &gt; 0.12\)</span> where <span class="math inline">\(0.12\)</span> comes from @ref(eq:HIVresult).</p>
<p>One can derive this mathematically by plugging in a larger number in @ref(eq:HIVpositive) than 0.00148, as that number represents the prior risk of HIV. Changing the calculations accordingly shows <span class="math inline">\(P(\text{Person tested has HIV} \mid \text{ELISA is positive}) &gt; 0.12\)</span>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource example number-lines code-with-copy"><code class="sourceCode"><span id="cb6-1"><a href="#cb6-1"></a>If the false positive rate of the test is higher than 1%, how, if at all, would you expect $P(\text{Person tested has HIV} \mid \text{ELISA is positive})$ to change?</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If the false positive rate increases, the probability of a wrong positive result increases. That means that a positive test result is more likely to be wrong and thus less indicative of HIV. Therefore, the probability of HIV after a positive ELISA goes down such that <span class="math inline">\(P(\text{Person tested has HIV} \mid \text{ELISA is positive}) &lt; 0.12\)</span>.</p>
<h3 id="bayes-updating">Bayes Updating</h3>
<p>In the previous section, we saw that one positive ELISA test yields a probability of having HIV of 12%. To obtain a more convincing probability, one might want to do a second ELISA test after a first one comes up positive. What is the probability of being HIV positive if also the second ELISA test comes back positive?</p>
<p>To solve this problem, we will assume that the correctness of this second test is not influenced by the first ELISA, that is, the tests are independent from each other. This assumption probably does not hold true as it is plausible that if the first test was a false positive, it is more likely that the second one will be one as well. Nonetheless, we stick with the independence assumption for simplicity.</p>
<p>In the last section, we used <span class="math inline">\(P(\text{Person tested has HIV}) = 0.00148\)</span>, see @ref(eq:HIVpositive), to compute the probability of HIV after one positive test. If we repeat those steps but now with <span class="math inline">\(P(\text{Person tested has HIV}) = 0.12\)</span>, the probability that a person with one positive test has HIV, we exactly obtain the probability of HIV after two positive tests. Repeating the maths from the previous section, involving Bayes’ rule, gives</p>
<span class="math display">\[\begin{multline}
  P(\text{Person tested has HIV} \mid \text{Second ELISA is also positive}) \\
  \begin{split}
  &amp;= \frac{P(\text{Person tested has HIV}) P(\text{Second ELISA is positive} \mid \text{Person tested has HIV})}{P(\text{Second ELISA is also positive})} \\
  &amp;= \frac{0.12 \cdot 0.93}{
  \begin{split}
  &amp;P(\text{Person tested has HIV}) P(\text{Second ELISA is positive} \mid \text{Has HIV}) \\
  &amp;+ P(\text{Person tested has no HIV}) P(\text{Second ELISA is positive} \mid \text{Has no HIV})
  \end{split}
  } \\
  &amp;= \frac{0.1116}{0.12 \cdot 0.93 + (1 - 0.12)\cdot (1 - 0.99)} \approx 0.93.
  \end{split}
  (\#eq:Bayes-updating)
\end{multline}\]</span>
<p>Since we are considering the same ELISA test, we used the same true positive and true negative rates as in Section @ref(sec:diagnostic-testing). We see that two positive tests makes it much more probable for someone to have HIV than when only one test comes up positive.</p>
<p>This process, of using Bayes’ rule to update a probability based on an event affecting it, is called Bayes’ updating. More generally, the what one tries to update can be considered ‘prior’ information, sometimes simply called the <strong>prior</strong>. The event providing information about this can also be data. Then, updating this prior using Bayes’ rule gives the information conditional on the data, also known as the <strong>posterior</strong>, as in the information <strong>after</strong> having seen the data. Going from the prior to the posterior is Bayes updating.</p>
<p>The probability of HIV after one positive ELISA, 0.12, was the posterior in the previous section as it was an update of the overall prevalence of HIV, @ref(eq:HIVpositive). However, in this section we answered a question where we used this posterior information as the prior. This process of using a posterior as prior in a new problem is natural in the Bayesian framework of updating knowledge based on the data.</p>
</section></section>
<section>
<section id="modern-statistics-beer-and-eugenics" class="title-slide slide level1 center">
<h1>Modern Statistics, Beer, and Eugenics</h1>

<img data-src="https://media.springernature.com/full/springer-static/image/art%253A10.1038%252F35038589/MediaObjects/41576_2000_Article_BF35038589_Fig1_HTML.gif?as=webp" class="r-stretch quarto-figure-center"><p class="caption">Image courtesy of the American Philosophical Society, Philadelphia, USA.</p></section>
<section id="fathers-of-statistics" class="slide level2">
<h2>Fathers of statistics</h2>
<p>The torch was passed within the triumvirate of Galton, Pearson, and Fisher.</p>
<h3 id="sir-francis-galton-1822-1911">Sir Francis Galton (1822-1911)</h3>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://galton.org/photos/standard-photo.gif"></p>
<figcaption>from galton.org</figcaption>
</figure>
</div>
<ul>
<li><p>Discovered regression to the mean</p></li>
<li><p>Re-discovered correlation and regression and discovered how to apply these in anthropology, psychology, and more</p></li>
<li><p>Defined the concept of standard deviation</p></li>
<li><p>Established the field of Eugenics in 1883</p></li>
<li><p>Darwin’s cousin.</p></li>
</ul>
<p>Galton’s reasoning for coining the term eugenics:</p>
<blockquote>
<p>“We greatly want a brief word to express the science of improving stock, which…takes cognisance of all influences that tend in however remote a degree to give the more suitable races or strains of blood a better chance of prevailing speedily over the less suitable than they otherwise would have had.”</p>
</blockquote>
<h3 id="karl-pearson-1857-1936">Karl Pearson (1857-1936)</h3>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://cdn.britannica.com/32/38832-004-24FF6E46/Karl-Pearson-pencil-drawing-FA-de-Biden-1924.jpg?s=1500x700&amp;q=85" width="350"></p>
<figcaption><a href="https://www.britannica.com/biography/Karl-Pearson" class="uri">https://www.britannica.com/biography/Karl-Pearson</a></figcaption>
</figure>
</div>
<p>Karl Pearson was Galton’s protégé and directly or contributed to:</p>
<ul>
<li><p>Developed hypothesis testing</p></li>
<li><p>Developed the use of p-values</p></li>
<li><p>Defined the Chi-Squared test</p></li>
<li><p>Correlation coefficient</p></li>
<li><p>Principle components analysis</p></li>
</ul>
<p>Also authors of timeless “classics” such as:<br>
<a href="https://archive.org/details/ethicoffreethoug00pear/page/370/mode/2up?view=theater">The Woman’s Question</a></p>
<p><a href="https://archive.org/details/nationallifefro00peargoog/page/n12/mode/2up?view=theater">National Life from the standpoint of science</a></p>
<p>In the year Mein Kampf was published, Pearson wrote an article called:<br>
<a href="https://onlinelibrary.wiley.com/doi/10.1111/j.1469-1809.1925.tb02037.x"><strong>THE PROBLEM OF ALIEN IMMIGRATION INTO GREAT BRITAIN, ILLUSTRATED BY AN EXAMINATION OF RUSSIAN AND POLISH JEWISH CHILDREN</strong></a></p>
<p>Here is an excerpt:</p>
<blockquote>
<p>“[they] will develop into a parasitic race…Taken <em>on the average</em>, and regarding both sexes, this alien Jewish population is somewhat inferior physically and mentally to the native population.”</p>
</blockquote>
<h3 id="sir-ronald-aylmer-fisher-1890-1962">Sir Ronald Aylmer Fisher (1890-1962)</h3>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://www.42evolution.org/wp-content/uploads/2014/07/Ronald-Fisher-from-Royal-Society.jpg"></p>
<figcaption><a href="https://www.42evolution.org/ronald-a-fisher/" class="uri">https://www.42evolution.org/ronald-a-fisher/</a></figcaption>
</figure>
</div>
<p>Fisher’s work in statistics established and promoted many important methods of statistical inference. His contributions include:</p>
<ul>
<li><p>Establishing p = 0.05 as the normal threshold for significant p-values</p></li>
<li><p>Promoting Maximum Likelihood Estimation</p></li>
<li><p>Developing the ANalysis Of VAriance (ANOVA) The iris dataset (this seems an incredibly minor contribution but I use it daily)</p></li>
<li><p><a href="https://en.wikipedia.org/wiki/The_Genetical_Theory_of_Natural_Selection">The Genetical Theory of Natural Selection</a>, which blended the work of Mendel and Darwin.</p></li>
</ul>
<p>There is no lack of Fisher’s strong and consistent support for eugenics. Here is an example from as late as 1954.</p>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://thisviewoflife.com/wp-content/uploads/2021/04/Fisher-to-Gates-08.27.54-768x534.png"></p>
<figcaption>Letter from R.A. Fisher to R. Ruggles Gates. Ronald Fisher Archive. University of Adelaide.</figcaption>
</figure>
</div>
</section>
<section id="storytime-galton-laboratory" class="slide level2">
<h2>Storytime: Galton Laboratory</h2>
<p>Galton founded the Eugenics Record Office (1904)</p>
<p>Galton Eugenics Laboratory as part of University College London (UCL). Created by Pearson and funded by Galton. (1907)</p>
<p>Galton left UCL enough money to create a Chair in National Eugenics, filled by Pearson and then Fisher. Hell of a name for an endowed chair!</p>
<p>Annals of Human Genetics: It was established in 1925 Pearson as the Annals of Eugenics, and obtained its current name in 1954.</p>
<p>Galton laboratory was incorporated into the Department of Eugenics, Biometry and Genetic at UCL in 1944.</p>
<p>Renamed to the Department of Human Genetics and Biometry in 1966.</p>
<p>Became part of the Department of Biology at UCL in 1996.</p>
<p>In 2020: <a href="https://www.theguardian.com/education/2020/jun/19/ucl-renames-three-facilities-that-honoured-prominent-eugenicists">UCL renames three facilities that honoured prominent eugenicists</a></p>
<p><strong>These horrendous views did not appear to be common at UCL in the 1930s. For example, they were not held by JBS Haldane, Egon Pearson (son of Karl), and Lionel Penrose.</strong></p>
<h3 id="what-about-in-the-us">What about in the US?</h3>
<p><a href="http://eugenicsarchive.ca/discover/tree/5233cfa15c2ec500000000ad">CSHL - Eugenics Archive</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2757926/">U.S. Scientists’ Role in the Eugenics Movement (1907–1939): A Contemporary Biologist’s Perspective</a></p>
<p><a href="https://carnegiescience.edu/carnegie-institution-science-statement-eugenics-research">Charles Davenport (first director of CSHL) and the Carnegie Insitution</a></p>
<p><a href="https://www.cshl.edu/archives/expanded-commentary/cold-spring-harbor-and-german-eugenics-in-the-1930s/">Cold Spring Harbor and German Eugenics in the 1930s</a></p>
<p><a href="https://www.science.org/content/blog-post/eugenics-and-history-science-and-aaas">Eugenics and the history of Science and AAAS</a></p>
</section>
<section id="section" class="slide level2">
<h2></h2>

<img data-src="https://i0.wp.com/www.theifod.com/wp-content/uploads/2021/02/image.png?w=1267&amp;ssl=1" class="r-stretch quarto-figure-center"><p class="caption">from “America’s Shameful History of Eugenics and Forced Sterilizations”</p></section>
<section id="modern-day-eugenics-and-beyond" class="slide level2">
<h2>Modern day: Eugenics and beyond</h2>
<h3 id="sordid-genealogies-a-conjectural-history-of-cambridge-analyticas">[Sordid genealogies: a conjectural history of Cambridge Analytica’s</h3>
<p>eugenic roots](https://www.nature.com/articles/s41599-020-0505-5)</p>
<h3 id="american-renaissance"><a href="https://www.amren.com/about/">American Renaissance</a></h3>
<p><img data-src="https://i0.wp.com/sitn.hms.harvard.edu/wp-content/uploads/2017/04/Fig1-raceConception-2.png" width="350" alt="https://sitn.hms.harvard.edu/flash/2017/science-genetics-reshaping-race-debate-21st-century/"><br>
‘Race’ cannot be biologically defined due to genetic variation among human individuals and populations. (A) The old concept of the “five races:” African, Asian, European, Native American, and Oceanian. (B) Actual genetic variation in humans.</p>
<h3 id="polygenic-traits-human-embryos-and-eugenic-dreams"><strong>Polygenic Traits, Human Embryos, and Eugenic Dreams</strong></h3>
<p>An academic study debunked the idea of “Screening Human Embryos for Polygenic Traits,” but the CEO of the company Stephen Hsu cofounded announced that they had screened human embryos for polygenic traits.<br>
<img data-src="https://www.geneticsandsociety.org/sites/default/files/styles/teaser/public/screen_shot_2021-09-26_at_5.48.10_pm.png?itok=p9514DCD" width="350" alt="https://www.geneticsandsociety.org/biopolitical-times/polygenic-traits-human-embryos-and-eugenic-dreams"></p>
<p><a href="https://liorpachter.wordpress.com/2021/04/12/the-amoral-nonsense-of-orchids-embryo-selection/">The amoral nonsense of Orchid’s embryo selection</a><br>
</p>
<h3 id="superior-the-return-of-race-science"><strong>“Superior: The Return of Race Science”</strong></h3>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://upload.wikimedia.org/wikipedia/en/9/90/Superior_(book)_cover.jpeg"></p>
<figcaption><a href="https://en.wikipedia.org/wiki/Superior:_The_Return_of_Race_Science" class="uri">https://en.wikipedia.org/wiki/Superior:_The_Return_of_Race_Science</a></figcaption>
</figure>
</div>
<h3 id="weapons-of-math-destruction"><strong>“Weapons of Math Destruction”</strong></h3>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="https://upload.wikimedia.org/wikipedia/en/0/0b/Weapons_of_Math_Destruction.jpg"></p>
<figcaption><a href="https://en.wikipedia.org/wiki/Weapons_of_Math_Destruction" class="uri">https://en.wikipedia.org/wiki/Weapons_of_Math_Destruction</a></figcaption>
</figure>
</div>
</section></section>
<section>
<section id="references-and-resources" class="title-slide slide level1 center">
<h1>References and resources</h1>

</section>
<h3 id="i-have-borrowed-heavily-from-directly-taken-andor-highly-recommend-the-following-fantastic-resources">I have borrowed heavily from, directly taken, and/or highly recommend the following fantastic resources:</h3>
<section id="multiple-testing" class="slide level2">
<h2>Multiple testing</h2>
<ul>
<li><a href="https://www.nature.com/articles/nbt1209-1135">How does multiple testing correction work?</a><br>
</li>
<li><a href="https://towardsdatascience.com/multiple-testing-how-should-you-adjust-41048eab4a3d">Multiple Testing — How Should You Adjust?</a><br>
</li>
<li><a href="https://benwhalley.github.io/just-enough-r/multiple-comparisons.html">Multiple Comparisons</a><br>
</li>
<li><a href="https://towardsdatascience.com/an-overview-of-methods-to-address-the-multiple-comparison-problem-310427b3ba92">An Overview of Methods to Address the Multiple Comparison Problem</a></li>
</ul>
</section>
<section id="bayesian-statistics-1" class="slide level2">
<h2>Bayesian statistics</h2>
<ul>
<li><p><a href="https://arbital.com/p/bayes_rule/?l=693">High-speed intro to Bayes’s rule</a></p></li>
<li><p><a href="https://statswithr.github.io/book/">An Introduction to Bayesian Thinking</a></p></li>
<li><p><a href="https://www.youtube.com/watch?v=9wCnvr7Xw4E">Bayes’ Theorem, Clearly Explained!!!!</a></p></li>
</ul>
</section>
<section id="modern-statistics-beer-and-eugenics-1" class="slide level2">
<h2>Modern Statistics, Beer, and Eugenics</h2>
<ul>
<li><p><a href="https://www.youtube.com/watch?v=U9Wr7VEPGXA">Why We Might Not Have Statistics Without Guinness Brewery</a></p></li>
<li><p><a href="https://towardsdatascience.com/statistics-eugenics-and-me-29eaf43efac7">Statistics, Eugenics, and Me</a></p></li>
<li><p><a href="https://medium.com/swlh/is-statistics-racist-59cd4ddb5fa9">Is Statistics Racist?</a></p></li>
<li><p><a href="https://www.forbes.com/sites/jerrybowyer/2016/01/06/beer-vs-eugenics-the-good-and-the-bad-uses-of-statistics/#64831bd32a14">Beer Vs. Eugenics: The Good And The Bad Uses Of Statistics</a></p></li>
<li><p><a href="https://www.nature.com/articles/35038589">Engineering American society: the lesson of eugenics</a></p></li>
<li><p><a href="https://www.kdnuggets.com/2016/04/eugenics-journey-dark-side-statistics.html">Eugenics – journey to the dark side at the dawn of statistics</a></p></li>
<li><p><a href="https://nautil.us/how-eugenics-shaped-statistics-9365/">How Eugenics Shaped Statistics</a></p></li>
<li><p><a href="https://www.jstor.org/stable/229774">Francis Galton’s Statistical Ideas: The Influence of Eugenics</a></p></li>
<li><p><a href="https://doi.org/10.1016/j.shpsc.2006.12.007">R. A. Fisher: a faith fit for eugenics</a></p></li>
<li><p><a href="https://www.nature.com/articles/s41599-020-0505-5">Sordid genealogies: a conjectural history of Cambridge Analytica’s eugenic roots</a></p></li>
<li><p><a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2757926/">U.S. Scientists’ Role in the Eugenics Movement (1907–1939): A Contemporary Biologist’s Perspective</a></p></li>
<li><p><a href="https://www.nature.com/articles/35002739">Biomedical centre memorial to victims of Nazi research</a></p></li>
<li><p><a href="https://link.springer.com/article/10.1007/s00109-017-1532-6">Berlin Wild—and the Max Delbrück Center for Molecular Medicine</a></p></li>
<li><p><a href="https://eugenicsarchive.ca/discover/timeline/543d5ab028f51f0000000003">Eugenics timeline</a></p></li>
<li><p><a href="https://profjoecain.net/karl-pearson-praised-hitler-nazi-race-hygiene/">Karl Pearson praised Hitler and Nazi Race Hygiene</a></p></li>
<li><p><a href="https://thisviewoflife.com/ronald-fisher-is-not-being-cancelled-but-his-eugenic-advocacy-should-have-consequences/">Ronald Fisher Is Not Being ‘Cancelled’, But His Eugenic Advocacy Should Have Consequences</a></p></li>
</ul>

<div class="footer footer-default">

</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>