<!DOCTYPE html>
<html lang="en"><head>
<script src="../../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../../site_libs/quarto-html/tabby.min.js"></script>
<script src="../../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../../../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.3.361">

  <meta name="author" content="Neelanjan Mukherjee">
  <title>MOLB 7950 - Bootcamp: Stats class 3</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../../../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../../../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
<meta property="og:title" content="MOLB 7950 - Bootcamp: Stats class 3">
<meta property="og:site-name" content="MOLB 7950">
<meta name="twitter:title" content="MOLB 7950 - Bootcamp: Stats class 3">
<meta name="twitter:image" content="https://rnabioco.github.io/molb-7950/content/bootcamp/stats/images/twitter-card.png">
<meta name="twitter:creator" content="@rnabioco">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Bootcamp: Stats class 3</h1>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
<a href="neelanjan.mukherjee@cuanschutz.edu">Neelanjan Mukherjee</a> 
</div>
</div>
</div>

  <p class="date">Invalid Date</p>
</section>
<section class="slide level2">

</section>
<section>
<section id="learning-objectives" class="title-slide slide level1 center">
<h1>Learning Objectives</h1>
<ul>
<li><strong>Visualize</strong> and <strong>Summarize</strong> the data being compared</li>
<li><strong>Formulate</strong> and <strong>Execute</strong> null hypothesis testing</li>
<li><strong>Identify</strong> and <strong>Perform</strong> the proper statistical test for data type/comparison</li>
<li><strong>Calculate</strong> and <strong>Interpret</strong> p-values</li>
<li><strong>Prevent</strong> p-hacking and <strong>Recognize</strong> issues with simultaneously testing multiple hypotheses.</li>
</ul>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="outline" class="title-slide slide level1 center">
<h1>Outline</h1>
<ul>
<li><strong>Concepts and Definitions</strong>
<ul>
<li><strong>Simplifying principles: Common tests as linear models</strong></li>
<li><strong>Types of comparisons and statistical tests</strong></li>
<li><strong>Definitions<br>
</strong></li>
</ul></li>
<li><strong>Day 1: Relationship between two or more continuous variables</strong>
<ul>
<li><strong>Correlation vs Regression</strong></li>
<li><strong>Fitting it a line to data</strong></li>
<li><strong>Linear regression concepts</strong></li>
<li><strong>Multiple regression</strong><br>
</li>
</ul></li>
<li>Day 2: Relationship between categorical and continuous variables
<ul>
<li><p>Comparing means between two groups (t-Test)</p></li>
<li><p>Comparing means between 3 or more groups (ANOVA)<br>
</p></li>
</ul></li>
<li>Day 3: Multiple test correction, Bayesian intro, History</li>
</ul>
</section>
<section id="provide-a-simple-and-flexible-framework" class="slide level2">
<h2>Provide a <strong>simple</strong> and <strong>flexible</strong> framework</h2>

<img data-src="img/easy_hard.jpg" class="r-stretch"></section>
<section class="slide level2">

</section></section>
<section>
<section id="cheatsheet" class="title-slide slide level1 center">
<h1>CHEATSHEET</h1>

<img data-src="img/linear_tests_cheat_sheet_adapted.png" class="r-stretch"></section>
<section id="variables-definitions" class="slide level2">
<h2>Variables definitions</h2>
<h3 id="random-variables-x-y">Random variables (x, y)</h3>
<p><strong>Response Variable</strong> ( <strong>y</strong> - aka dependent or outcome variable): this variable is predicted or its variation is explained by the explanatory variable. In an experiment, this is the outcome that is measured following manipulation of the explanatory variable.</p>
<p><strong>Explanatory Variable</strong> ( <strong>x</strong> - aka independent or predictor variable): explains variations in the response variable. In an experiment, it is manipulated by the researcher.</p>
<h3 id="quantitative-variables">Quantitative Variables</h3>
<p><strong>Discrete variable</strong>: numeric variables that have a countable number of values between any two values - <code>integer</code> in R (e.g., number of mice, read counts).</p>
<p><strong>Continuous variable</strong>: numeric variables that have an infinite number of values between any two values - <code>numeric</code> in R (e.g., normalized expression values, fluorescent intensity).</p>
<h3 id="categorical-variables">Categorical Variables</h3>
<p><strong>Nominal variable</strong>: (unordered) random variables have categories where order doesn’t matter - <code>factor</code> in R (e.g., country, type of gene, genotype).</p>
<p><strong>Ordinal variable</strong>: (ordered) random variables have ordered categories - order of <code>levels</code> in R ( e.g.&nbsp;grade of tumor).</p>
</section>
<section id="hypothesis-testing-definitions" class="slide level2">
<h2>Hypothesis testing definitions</h2>
<p><strong>Hypothesis testing</strong> is a statistical analysis that uses sample data to assess two mutually exclusive theories about the properties of a population. Statisticians call these theories the null hypothesis and the alternative hypothesis. A hypothesis test assesses your sample statistic and factors in an estimate of the sample error to determine which hypothesis the data support.</p>
<p>When you can reject the null hypothesis, the results are statistically significant, and your data support the theory that an effect exists at the population level.</p>
<p><a href="https://www.graphpad.com/guides/prism/latest/statistics/hypothesis_testing_and_statistical_significance.htm"><strong>A legal analogy: Guilty or not guilty?</strong></a><strong><br>
</strong>The statistical concept of ‘significant’ vs.&nbsp;‘not significant’ can be understood by comparing to the legal concept of ‘guilty’ vs.&nbsp;‘not guilty’.</p>
<p>In the American legal system (and much of the world) a criminal defendant is presumed innocent until proven guilty. If the evidence proves the defendant guilty beyond a reasonable doubt, the verdict is ‘guilty’. Otherwise the verdict is ‘not guilty’. In some countries, this verdict is ‘not proven’, which is a better description. A ‘not guilty’ verdict does not mean the judge or jury concluded that the defendant is innocent -- it just means that the evidence was not strong enough to persuade the judge or jury that the defendant was guilty.</p>
<p>In statistical hypothesis testing, you start with the null hypothesis (usually that there is no difference between groups). If the evidence produces a small enough P value, you reject that null hypothesis, and conclude that the difference is real. If the P value is higher than your threshold (usually 0.05), you don’t reject the null hypothesis. This doesn’t mean the evidence convinced you that the treatment had no effect, only that the evidence was not persuasive enough to convince you that there is an effect.</p>
<p><strong>Effect</strong> — the difference between the population value and the null hypothesis value. The effect is also known as population effect or the difference. Typically, you do not know the size of the actual effect. However, you can use a hypothesis test to help you determine whether an effect exists and to estimate its size.</p>
<p><strong>Null Hypothesis</strong> or <span class="math inline">\(\mathcal{H}_0\)</span> — one of two mutually exclusive theories about the properties of the population in hypothesis testing. Typically, the null hypothesis states that there is no effect (i.e., the effect size equals zero).</p>
<p><strong>Alternative Hypothesis</strong> or <span class="math inline">\(\mathcal{H}_1\)</span> — the other theory about the properties of the population in hypothesis testing. Typically, the alternative hypothesis states that a population parameter does not equal the null hypothesis value. In other words, there is a non-zero effect. If your sample contains sufficient evidence, you can reject the null and favor the alternative hypothesis.</p>
<p><strong>P-values</strong> — the probability of obtaining test results at least as extreme as the results actually observed, under the assumption that the null hypothesis is correct. Lower p-values represent stronger evidence against the null. P-values in conjunction with the significance level determines whether your data favor the null or alternative hypothesis.</p>
<p><a href="https://www.youtube.com/watch?v=5Z9OIYA8He8">StatQuest: P Values, clearly explained</a></p>
<p><a href="https://www.youtube.com/watch?v=JQc3yx0-Q9E">StatQuest: How to calculate p-values</a></p>
<p><strong>Significance Level</strong> or <span class="math inline">\(a\)</span> — an evidentiary standard set before the study. It is the probability that you say there is an effect when there is no effect (the probability of rejecting the null hypothesis given that it is true). Lower significance levels indicate that you require stronger evidence before you will reject the null.It is usually set at or below .05.</p>

<img data-src="https://upload.wikimedia.org/wikipedia/commons/3/35/Guinness_Glass_2010.jpg" style="width:10.0%" class="r-stretch quarto-figure-center"><p class="caption">Guinness</p></section>
<section id="null-hypothesis-testing" class="slide level2">
<h2>Null hypothesis testing</h2>
<ol type="1">
<li>Specify the variables</li>
<li>Declare null hypothesis <span class="math inline">\(\mathcal{H}_0\)</span></li>
<li>Calculate test-statistic, exact p-value</li>
<li><em>Generate and visualize data reflecting null-distribution</em></li>
<li><em>Calculate the p-value from the test statistic and null distribution</em></li>
</ol>
<p>*4-5: For calculating empirical p-value</p>
</section>
<section id="the-simplicity-underlying-common-tests" class="slide level2">
<h2>The simplicity underlying common tests</h2>
<p>Most of the common statistical models (t-test, correlation, ANOVA; chi-square, etc.) are special cases of linear models or a very close approximation. This simplicity means that there is less to learn. In particular, it all comes down to:<br>
<span class="math inline">\(y = a \cdot x + b\)</span></p>
<p>This needless complexity multiplies when students try to rote learn the parametric assumptions underlying each test separately rather than deducing them from the linear model.</p>
</section>
<section id="parametric-vs-non-parametric-tests" class="slide level2">
<h2>Parametric vs Non-Parametric tests</h2>
<p><strong>Parametric tests</strong> are suitable for normally distributed data.</p>
<p><strong>Non-Parametric tests</strong> are suitable for any continuous data. For the sake of simplicity and sticking with a consistent framework, we will consider Non-Parametric tests as the <strong>ranked versions of the corresponding parametric tests</strong>.</p>
<p><a href="https://statisticsbyjim.com/hypothesis-testing/nonparametric-parametric-tests/">More on choosing Parametric vs Non-Parametric</a></p>
<div class="cell">
<div class="cell-output-display">
<table>
<thead>
<tr class="header">
<th style="text-align: left;">Info</th>
<th style="text-align: left;">Parametric</th>
<th style="text-align: left;">Non-Parametric</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">better descriptor</td>
<td style="text-align: left;">mean</td>
<td style="text-align: left;">median</td>
</tr>
<tr class="even">
<td style="text-align: left;"># of samples (N)</td>
<td style="text-align: left;">many</td>
<td style="text-align: left;">few</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="import-and-tidy-data" class="title-slide slide level1 center">
<h1>Import and tidy data</h1>
<p>We will be using mouse data from <a href="https://wp.cs.ucl.ac.uk/outbredmice/">Resources for Outbred Mice</a>. The goal of the study was to establish genotype-phenotype relationships for highly recombinant outbred mouse populations. We will be using the phenotypic data for our exercises.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># we are reading the data directly from the internet</span></span>
<span id="cb1-2"><a href="#cb1-2"></a>biochem <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"http://mtweb.cs.ucl.ac.uk/HSMICE/PHENOTYPES/Biochemistry.txt"</span>, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-3"><a href="#cb1-3"></a>  janitor<span class="sc">::</span><span class="fu">clean_names</span>()</span>
<span id="cb1-4"><a href="#cb1-4"></a></span>
<span id="cb1-5"><a href="#cb1-5"></a><span class="co"># simplify names a bit more</span></span>
<span id="cb1-6"><a href="#cb1-6"></a><span class="fu">colnames</span>(biochem) <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="at">pattern =</span> <span class="st">"biochem_"</span>, <span class="at">replacement =</span> <span class="st">""</span>, <span class="fu">colnames</span>(biochem))</span>
<span id="cb1-7"><a href="#cb1-7"></a></span>
<span id="cb1-8"><a href="#cb1-8"></a><span class="co"># we are going to simplify this a bit and only keep some columns</span></span>
<span id="cb1-9"><a href="#cb1-9"></a>keep <span class="ot">&lt;-</span> <span class="fu">colnames</span>(biochem)[<span class="fu">c</span>(<span class="dv">1</span>,<span class="dv">6</span>,<span class="dv">9</span>,<span class="dv">14</span>,<span class="dv">15</span>,<span class="dv">24</span><span class="sc">:</span><span class="dv">28</span>)]</span>
<span id="cb1-10"><a href="#cb1-10"></a>biochem <span class="ot">&lt;-</span> biochem[,keep]</span>
<span id="cb1-11"><a href="#cb1-11"></a></span>
<span id="cb1-12"><a href="#cb1-12"></a><span class="co"># get weights for each individual mouse</span></span>
<span id="cb1-13"><a href="#cb1-13"></a><span class="co"># careful: did not come with column names</span></span>
<span id="cb1-14"><a href="#cb1-14"></a>weight <span class="ot">&lt;-</span> <span class="fu">read_tsv</span>(<span class="st">"http://mtweb.cs.ucl.ac.uk/HSMICE/PHENOTYPES/weight"</span>, <span class="at">col_names =</span> F, <span class="at">show_col_types =</span> <span class="cn">FALSE</span>)</span>
<span id="cb1-15"><a href="#cb1-15"></a></span>
<span id="cb1-16"><a href="#cb1-16"></a><span class="co"># add column names</span></span>
<span id="cb1-17"><a href="#cb1-17"></a><span class="fu">colnames</span>(weight) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"subject_name"</span>,<span class="st">"weight"</span>)</span>
<span id="cb1-18"><a href="#cb1-18"></a></span>
<span id="cb1-19"><a href="#cb1-19"></a><span class="co"># add weight to biochem table and get rid of NAs</span></span>
<span id="cb1-20"><a href="#cb1-20"></a>biochem <span class="ot">&lt;-</span> <span class="fu">inner_join</span>(biochem, weight, <span class="at">by=</span><span class="st">"subject_name"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-21"><a href="#cb1-21"></a>  <span class="fu">na.omit</span>()</span>
<span id="cb1-22"><a href="#cb1-22"></a></span>
<span id="cb1-23"><a href="#cb1-23"></a></span>
<span id="cb1-24"><a href="#cb1-24"></a><span class="co"># explore the data a bit</span></span>
<span id="cb1-25"><a href="#cb1-25"></a><span class="fu">colnames</span>(biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "subject_name"    "calcium"         "glucose"         "sodium"         
 [5] "tot_cholesterol" "family"          "gender"          "age"            
 [9] "cage_density"    "litter"          "weight"         </code></pre>
</div>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="fu">str</span>(biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>tibble [1,782 × 11] (S3: tbl_df/tbl/data.frame)
 $ subject_name   : chr [1:1782] "A048005080" "A048005112" "A048006555" "A048007096" ...
 $ calcium        : num [1:1782] 1.97 2.11 1.71 2.49 2.14 2.16 2.29 1.69 1.84 2.31 ...
 $ glucose        : num [1:1782] 12.23 10.99 5.97 10.61 11.88 ...
 $ sodium         : num [1:1782] 123 133 119 148 131 134 146 117 121 144 ...
 $ tot_cholesterol: num [1:1782] 3.01 2.46 3.57 2.61 2.04 2.86 3.22 3.47 3.35 2.29 ...
 $ family         : chr [1:1782] "H2.3:C5.2(3) H2.3:G2.2(3)" "H2.2:C3.1(4) H2.2:G3.1(3)" "E1.3:H1.2(3) E1.3:D1.2(3)" "D3.2:G2.1(5) D3.2:C5.1(5)" ...
 $ gender         : chr [1:1782] "F" "F" "M" "M" ...
 $ age            : num [1:1782] 66 70 72 66 63 72 66 72 64 66 ...
 $ cage_density   : num [1:1782] 5 3 4 4 6 7 4 5 6 3 ...
 $ litter         : num [1:1782] 2 4 1 4 1 3 1 1 1 1 ...
 $ weight         : num [1:1782] 20.3 16.1 19.5 22.2 17.3 18.1 25.6 18.6 23.1 17.3 ...
 - attr(*, "na.action")= 'omit' Named int [1:237] 3 8 11 17 18 23 25 35 38 40 ...
  ..- attr(*, "names")= chr [1:237] "3" "8" "11" "17" ...</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="co"># View(biochem)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">

</section></section>
<section>
<section id="relationship-between-two-or-more-continuous-variables" class="title-slide slide level1 center">
<h1>Relationship between two or more continuous variables?</h1>
<p>What is the relationship between weight and cholesterol?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> tot_cholesterol)) <span class="sc">+</span></span>
<span id="cb6-2"><a href="#cb6-2"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb6-3"><a href="#cb6-3"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb6-4"><a href="#cb6-4"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-2-1.png" width="960" class="r-stretch"></section>
<section id="correlation-vs-regression" class="slide level2">
<h2>Correlation vs Regression</h2>
<p><strong>Correlation</strong> is primarily used to quickly and concisely summarize the direction and strength of the relationships between a set of 2 or more numeric variables.&nbsp;</p>
<p><strong>Regression</strong> is primarily used to build models/equations to predict a key response, Y, from a set of predictor (X) variables.</p>
<table>
<colgroup>
<col style="width: 25%">
<col style="width: 37%">
<col style="width: 37%">
</colgroup>
<thead>
<tr class="header">
<th></th>
<th>Correlation</th>
<th>Regression</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td><strong>Description</strong></td>
<td>Association between 2 or more variables</td>
<td>How an independent variable is numerically related to the dependent variable</td>
</tr>
<tr class="even">
<td><strong>Usage</strong></td>
<td>To represent linear relationship between two variables</td>
<td>To fit a best line and estimate one variable on the basis of another variable</td>
</tr>
<tr class="odd">
<td><strong>Dependent vs Independent variables</strong></td>
<td>Doesn’t matter</td>
<td>must define (i.e.&nbsp;order of relationship matters)</td>
</tr>
<tr class="even">
<td><strong>Interpretation</strong></td>
<td>Correlation coefficient indicates the extent to which two variables move together</td>
<td>Regression indicates the impact of a unit change in the known variable (x) on the estimated variable (y)</td>
</tr>
<tr class="odd">
<td><strong>Goal</strong></td>
<td>To find a numerical value expressing the relationship between variables</td>
<td>To estimate values of random variable on the basis of the values of fixed variable</td>
</tr>
</tbody>
</table>
<p>*Borrowed from and more info available <a href="https://www.graphpad.com/support/faq/what-is-the-difference-between-correlation-and-linear-regression/">here</a> and <a href="https://keydifferences.com/difference-between-correlation-and-regression.html">here</a>.</p>
<h3 id="pearson-correlation">Pearson Correlation</h3>
<p>It was developed by <a href="https://en.wikipedia.org/wiki/Karl_Pearson" title="Karl Pearson">Karl Pearson</a> from a related idea introduced by <a href="https://en.wikipedia.org/wiki/Francis_Galton" title="Francis Galton">Francis Galton</a> in the 1880s, and for which the mathematical formula was derived and published by <a href="https://en.wikipedia.org/wiki/Auguste_Bravais" title="Auguste Bravais">Auguste Bravais</a> in 1844.<sup><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-6">[a]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-7">[6]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-8">[7]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-9">[8]</a><a href="https://en.wikipedia.org/wiki/Pearson_correlation_coefficient#cite_note-10">[9]</a></sup> The naming of the coefficient is thus an example of <a href="https://en.wikipedia.org/wiki/Stigler%27s_Law" title="Stigler's Law">Stigler’s Law</a> (see list of examples <a href="https://en.wikipedia.org/wiki/List_of_examples_of_Stigler%27s_law">here</a>).</p>
<p>Interpretation of coefficient:</p>
<p>1 = perfect linear correlation</p>
<p>0 = no correlation</p>
<p>-1 = perfect linear anti-correlation</p>
<p><span class="math inline">\(Corr(x,y) = \displaystyle \frac {\sum_{i=1}^{n} (x_{i} - \overline{x})(y_{i} - \overline{y})}{\sum_{i=1}^{n} \sqrt(x_{i} - \overline{x})^2 \sqrt(y_{i} - \overline{y})^2}\)</span></p>
<p><span class="math inline">\(x_{i}\)</span> = the “i-th” observation of the variable <span class="math inline">\(x\)</span></p>
<p><span class="math inline">\(\overline{x}\)</span> = mean of all observations of <span class="math inline">\(x\)</span></p>
<p><span class="math inline">\(y_{i}\)</span> = the “i-th” observation of the variable <span class="math inline">\(y\)</span></p>
<p><span class="math inline">\(\overline{y}\)</span> = mean of all observations of <span class="math inline">\(y\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a>r_pearson <span class="ot">&lt;-</span> <span class="fu">cor</span>(<span class="at">x =</span> biochem<span class="sc">$</span>tot_cholesterol, <span class="at">y =</span> biochem<span class="sc">$</span>weight)</span>
<span id="cb7-2"><a href="#cb7-2"></a></span>
<span id="cb7-3"><a href="#cb7-3"></a>r_pearson</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3540731</code></pre>
</div>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a><span class="co"># average total cholesterol</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>avg_chol <span class="ot">&lt;-</span> <span class="fu">mean</span>(biochem<span class="sc">$</span>tot_cholesterol)</span>
<span id="cb9-3"><a href="#cb9-3"></a></span>
<span id="cb9-4"><a href="#cb9-4"></a><span class="co"># average weight</span></span>
<span id="cb9-5"><a href="#cb9-5"></a>avg_weight <span class="ot">&lt;-</span> <span class="fu">mean</span>(biochem<span class="sc">$</span>weight)</span>
<span id="cb9-6"><a href="#cb9-6"></a></span>
<span id="cb9-7"><a href="#cb9-7"></a><span class="co"># difference from mean total cholesterol</span></span>
<span id="cb9-8"><a href="#cb9-8"></a>diff_chol <span class="ot">&lt;-</span> biochem<span class="sc">$</span>tot_cholesterol <span class="sc">-</span> avg_chol</span>
<span id="cb9-9"><a href="#cb9-9"></a></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="co"># difference from mean total cholesterol</span></span>
<span id="cb9-11"><a href="#cb9-11"></a>diff_weight <span class="ot">&lt;-</span> biochem<span class="sc">$</span>weight <span class="sc">-</span> avg_weight</span>
<span id="cb9-12"><a href="#cb9-12"></a></span>
<span id="cb9-13"><a href="#cb9-13"></a><span class="co"># follow formula above</span></span>
<span id="cb9-14"><a href="#cb9-14"></a>manual_pearson <span class="ot">&lt;-</span> <span class="fu">sum</span>(diff_chol<span class="sc">*</span>diff_weight)<span class="sc">/</span>(</span>
<span id="cb9-15"><a href="#cb9-15"></a><span class="fu">sqrt</span>(<span class="fu">sum</span>(diff_chol<span class="sc">^</span><span class="dv">2</span>))<span class="sc">*</span><span class="fu">sqrt</span>(<span class="fu">sum</span>(diff_weight<span class="sc">^</span><span class="dv">2</span>)))</span>
<span id="cb9-16"><a href="#cb9-16"></a></span>
<span id="cb9-17"><a href="#cb9-17"></a>manual_pearson</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3540731</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a><span class="fu">identical</span>(manual_pearson, r_pearson)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1"></a><span class="co"># cov(x = biochem$tot_cholesterol, y = biochem$weight)/(sd(biochem$tot_cholesterol)*sd(biochem$weight))</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section class="slide level2">

<h3 id="spearman-correlation-nonparametric">Spearman Correlation (nonparametric)</h3>
<p><a href="https://en.wikipedia.org/wiki/Spearman%27s_rank_correlation_coefficient">Spearman’s rank correlation coefficient</a>or Spearman’s ρ, named after Charles Spearman is a <strong>nonparametric</strong> measure of <strong>rank</strong> correlation (statistical dependence between the rankings of two variables). It assesses how well the relationship between two variables can be described using a monotonic function.</p>
<p>More info <a href="https://towardsdatascience.com/clearly-explained-pearson-v-s-spearman-correlation-coefficient-ada2f473b8#:~:text=The%20fundamental%20difference%20between%20the,with%20monotonic%20relationships%20as%20well.">here</a>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1"></a>x <span class="ot">&lt;-</span> <span class="fu">seq</span>(<span class="dv">1</span>,<span class="dv">30</span>, <span class="dv">1</span>)</span>
<span id="cb14-2"><a href="#cb14-2"></a>y <span class="ot">&lt;-</span> <span class="dv">2</span><span class="sc">^</span>x</span>
<span id="cb14-3"><a href="#cb14-3"></a></span>
<span id="cb14-4"><a href="#cb14-4"></a><span class="fu">plot</span>(x,y, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">las=</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1"></a><span class="fu">cor</span>(x,y, <span class="at">method =</span> <span class="st">"pearson"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5199042</code></pre>
</div>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1"></a><span class="fu">cor</span>(x,y, <span class="at">method =</span> <span class="st">"spearman"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1"></a><span class="fu">cor</span>(<span class="at">x =</span> biochem<span class="sc">$</span>tot_cholesterol, <span class="at">y =</span> biochem<span class="sc">$</span>weight, <span class="at">method =</span> <span class="st">"spearman"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3596281</code></pre>
</div>
</div>
<img data-src="stats-class-03_files/figure-revealjs/spearman-1.png" width="960" class="r-stretch"></section>
<section id="regression" class="slide level2">
<h2>Regression</h2>
</section>
<section id="equation-for-a-line" class="slide level2">
<h2>Equation for a line</h2>
<p>Remember:<br>
<span class="math inline">\(y = a \cdot x + b\)</span><br>
OR<br>
<span class="math inline">\(y = b + a \cdot x\)</span></p>
<p><span class="math inline">\(a\)</span> is the <strong>SLOPE</strong><br>
<span class="math inline">\(b\)</span> is the <strong>y-intercept</strong></p>

<img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-3-1.png" width="192" class="r-stretch"><p><span class="math inline">\(a\)</span> = 2 (the slope)<br>
<span class="math inline">\(b\)</span> = 0.5 (the y-intercept)</p>
</section>
<section id="stats-equation-for-a-line" class="slide level2">
<h2>Stats equation for a line</h2>
<p>Model: the recipe for <span class="math inline">\(y\)</span> is a slope (<span class="math inline">\(\beta_1\)</span>) times <span class="math inline">\(x\)</span> plus an intercept (<span class="math inline">\(\beta_0\)</span>).</p>
<p><span class="math inline">\(y = \beta_0 + \beta_1 x \qquad \qquad \mathcal{H}_0: \beta_1 = 0\)</span></p>
<p>… which is the same has <span class="math inline">\(y = a \cdot x + b\)</span> (here ordered as <span class="math inline">\(y = b + a \cdot x\)</span>). In R we are lazy and write <code>y ~ 1 + x</code> which R reads like <code>y = 1*number + x*othernumber</code> and the task of t-tests, lm, etc., is simply to find the numbers that best predict <span class="math inline">\(y\)</span>.</p>
<p>Either way you write it, it’s an intercept (<span class="math inline">\(\beta_0\)</span>) and a slope (<span class="math inline">\(\beta_1\)</span>) yielding a straight line:</p>

<img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-4-1.png" width="192" class="r-stretch"><p><span class="math inline">\(\beta_0\)</span> = 0.5 (the y-intercept)<br>
<span class="math inline">\(\beta_1\)</span> = 2 (the slope)</p>
<p><span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x\)</span><br>
<span class="math inline">\(y = .5 \cdot 1 + 2 \cdot x\)</span></p>
<p>Our mission: <strong>FIND THE BEST</strong> <span class="math inline">\(\beta\)</span> coefficients</p>
</section>
<section id="linear-regression" class="slide level2">
<h2>Linear Regression</h2>
<ul>
<li>STEP 1: Make a scatter plot visualize the linear relationship between x and y.</li>
<li>STEP 2: Perform the regression</li>
<li>STEP 3: Look at the <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value</li>
<li>STEP 4: Visualize fit and errors</li>
<li>STEP 5: Calculate <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value ourselves</li>
</ul>
</section>
<section class="slide level2">

<h3 id="step-1-can-mouse-cholesterol-levels-help-explain-mouse-weight">STEP 1: Can mouse cholesterol levels help explain mouse weight?</h3>
<p>Plot weight (y, response variable) and cholesterol levels (x, explanatory variable) of the mice.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> tot_cholesterol)) <span class="sc">+</span></span>
<span id="cb21-2"><a href="#cb21-2"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb21-3"><a href="#cb21-3"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb21-4"><a href="#cb21-4"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-5-1.png" width="960" class="r-stretch"></section>
<section class="slide level2">

<h3 id="step-2-do-the-regression">STEP 2: Do the regression</h3>
<p>Let’s fit a line (linear model).</p>
<p>Remember: <span class="math inline">\(y = \beta_0 \cdot 1+ \beta_1 \cdot x\)</span></p>
<p>linear model equation: <span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol\)</span></p>
<p><span class="math inline">\(\mathcal{H}_0:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> does NOT explain <span class="math inline">\(weight\)</span></p>
<p>Null Hypothesis: <span class="math inline">\(\mathcal{H}_0: \beta_1 = 0\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + 0 \cdot cholesterol\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1\)</span></p>
<p>Alternative Hypothesis: <span class="math inline">\(\mathcal{H}_1: \beta_1 \neq 0\)</span><br>
</p>
<p>Full model: <span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x\)</span></p>
<p><span class="math inline">\(\mathcal{H}_1:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> does explain <span class="math inline">\(weight\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol\)</span></p>
<p>The cool thing here is that we can assess and compare our null and alternative hypothesis by learning and examining the model coefficients (intercept and slope). Essentially, we are comparing a complex model (including cholesterol) to a simple model (weight).</p>
<p><a href="https://statisticsbyjim.com/regression/interpret-constant-y-intercept-regression/" class="uri">https://statisticsbyjim.com/regression/interpret-constant-y-intercept-regression/</a></p>
<h3 id="step-4-look-at-the-r2-f-value-and-p-value">STEP 4: Look at the <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a><span class="co"># fitting a line</span></span>
<span id="cb22-2"><a href="#cb22-2"></a>fit_WvC <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">data =</span> biochem,</span>
<span id="cb22-3"><a href="#cb22-3"></a>              <span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> tot_cholesterol)</span>
<span id="cb22-4"><a href="#cb22-4"></a></span>
<span id="cb22-5"><a href="#cb22-5"></a></span>
<span id="cb22-6"><a href="#cb22-6"></a><span class="co"># base R summary of fit</span></span>
<span id="cb22-7"><a href="#cb22-7"></a><span class="fu">summary</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ 1 + tot_cholesterol, data = biochem)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.9624 -2.1349 -0.2627  2.0113 10.2927 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      14.5560     0.3635   40.04   &lt;2e-16 ***
tot_cholesterol   1.8516     0.1159   15.97   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.121 on 1780 degrees of freedom
Multiple R-squared:  0.1254,    Adjusted R-squared:  0.1249 
F-statistic: 255.1 on 1 and 1780 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<p>That’s a lot of info, but how would I access it? Time to meet your new best friend — <a href="&quot;https://cran.r-project.org/web/packages/broom/vignettes/broom.html&quot;">Broom</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1"></a><span class="co"># information about the model fit</span></span>
<span id="cb24-2"><a href="#cb24-2"></a><span class="fu">glance</span>(fit_WvC) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 12
  r.squared adj.r.squared sigma statistic  p.value    df logLik   AIC   BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     0.125         0.125  3.12      255. 8.92e-54     1 -4556. 9117. 9134.
# ℹ 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a><span class="co"># information about the intercept and coefficients</span></span>
<span id="cb26-2"><a href="#cb26-2"></a><span class="fu">tidy</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term            estimate std.error statistic   p.value
  &lt;chr&gt;              &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)        14.6      0.363      40.0 1.48e-250
2 tot_cholesterol     1.85     0.116      16.0 8.92e- 54</code></pre>
</div>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1"></a>chol_intercept <span class="ot">&lt;-</span> <span class="fu">tidy</span>(fit_WvC)[<span class="dv">1</span>,<span class="dv">2</span>]</span>
<span id="cb28-2"><a href="#cb28-2"></a></span>
<span id="cb28-3"><a href="#cb28-3"></a>chol_slope <span class="ot">&lt;-</span> <span class="fu">tidy</span>(fit_WvC)[<span class="dv">2</span>,<span class="dv">2</span>]</span>
<span id="cb28-4"><a href="#cb28-4"></a><span class="co"># for every 1 unit increase in cholesterol there is a 1.85 unit increase weight </span></span>
<span id="cb28-5"><a href="#cb28-5"></a></span>
<span id="cb28-6"><a href="#cb28-6"></a></span>
<span id="cb28-7"><a href="#cb28-7"></a><span class="co"># add residuals and other information</span></span>
<span id="cb28-8"><a href="#cb28-8"></a><span class="fu">augment</span>(fit_WvC) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1,782 × 8
   weight tot_cholesterol .fitted .resid     .hat .sigma     .cooksd .std.resid
    &lt;dbl&gt;           &lt;dbl&gt;   &lt;dbl&gt;  &lt;dbl&gt;    &lt;dbl&gt;  &lt;dbl&gt;       &lt;dbl&gt;      &lt;dbl&gt;
 1   20.3            3.01    20.1  0.171 0.000566   3.12 0.000000846     0.0547
 2   16.1            2.46    19.1 -3.01  0.00107    3.12 0.000501       -0.965 
 3   19.5            3.57    21.2 -1.67  0.000906   3.12 0.000129       -0.534 
 4   22.2            2.61    19.4  2.81  0.000853   3.12 0.000347        0.901 
 5   17.3            2.04    18.3 -1.03  0.00203    3.12 0.000111       -0.331 
 6   18.1            2.86    19.9 -1.75  0.000622   3.12 0.0000981      -0.561 
 7   25.6            3.22    20.5  5.08  0.000592   3.12 0.000786        1.63  
 8   18.6            3.47    21.0 -2.38  0.000782   3.12 0.000228       -0.763 
 9   23.1            3.35    20.8  2.34  0.000669   3.12 0.000189        0.750 
10   17.3            2.29    18.8 -1.50  0.00140    3.12 0.000161       -0.480 
# ℹ 1,772 more rows</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1"></a><span class="co"># add residuals and other information into the biochem object</span></span>
<span id="cb30-2"><a href="#cb30-2"></a>biochem_WvC <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_WvC, <span class="at">data =</span> biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>For every 1 unit increase in <span class="math inline">\(cholesterol\)</span> there is a 1.8516281 increase in <span class="math inline">\(weight\)</span>.</p>
</section>
<section class="slide level2">

<h3 id="step-5-visualize-fit-and-errors">STEP 5: Visualize fit and errors</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> tot_cholesterol)) <span class="sc">+</span></span>
<span id="cb31-2"><a href="#cb31-2"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="at">col=</span><span class="st">"white"</span>) <span class="sc">+</span></span>
<span id="cb31-3"><a href="#cb31-3"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="fu">pull</span>(chol_intercept), <span class="at">slope =</span> <span class="fu">pull</span>(chol_slope), <span class="at">col =</span> <span class="st">"pink"</span>, <span class="at">size=</span><span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb31-4"><a href="#cb31-4"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">se =</span> F, <span class="at">size=</span><span class="dv">1</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>) <span class="sc">+</span></span>
<span id="cb31-5"><a href="#cb31-5"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb31-6"><a href="#cb31-6"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="stats-class-03_files/figure-revealjs/visualize lm-1.png" width="960"></p>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb32-2"><a href="#cb32-2"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> </span>
<span id="cb32-3"><a href="#cb32-3"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="fu">pull</span>(chol_intercept), <span class="at">slope =</span> <span class="fu">pull</span>(chol_slope), <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb32-4"><a href="#cb32-4"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code</span></span>
<span id="cb32-5"><a href="#cb32-5"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">1</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb32-6"><a href="#cb32-6"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="stats-class-03_files/figure-revealjs/visualize lm-2.png" width="960"></p>
</div>
</div>
</section>
<section class="slide level2">

<p>STEP 6: Calculate <span class="math inline">\(R^2\)</span>, <span class="math inline">\(F\)</span>-value and <span class="math inline">\(p\)</span>-value ourselves</p>
<h4 id="definitions">DEFINITIONS</h4>
<p><span class="math inline">\(SS_{mean}\)</span> — sum of squared errors around the mean of <span class="math inline">\(y\)</span></p>
<p><span class="math inline">\(SS_{mean} = \sum_{i=1}^{n} (data - mean)^2\)</span><br>
<span class="math inline">\(SS_{mean} = \sum_{i=1}^{n} (y_{i} - \overline{y})^2\)</span></p>
<p><span class="math inline">\(Var_{mean}\)</span> — think of it like the average of the sum of squared error around the mean of <span class="math inline">\(y\)</span></p>
<p><span class="math inline">\(Var_{mean} = \displaystyle \frac {\sum_{i=1}^{n} (y_{i} - \overline{y})^2}{n}\)</span></p>
<p><span class="math inline">\(SS_{fit}\)</span> — sum of squared errors around the least-squares fit</p>
<p><span class="math inline">\(SS_{fit} = \sum_{i=1}^{n} (data - line)^2\)</span></p>
<p><span class="math inline">\(SS_{fit} = \sum_{i=1}^{n} (y_{i} - (\beta_0 \cdot 1+ \beta_1 \cdot x)^2\)</span></p>
<p><strong>Residuals</strong>, <span class="math inline">\(e\)</span> — the difference between the observed value of the dependent variable <span class="math inline">\(y\)</span> and the predicted value <span class="math inline">\(\widehat{y}\)</span> is called the residual. Each data point has one residual.</p>
<p><span class="math inline">\(e = y_{i} - \widehat{y}\)</span></p>
<p><span class="math inline">\(Var_{fit}\)</span> — think of it like the average of the sum of squared errors around the least-squares fit</p>
<p><span class="math inline">\(Var_{fit} = \displaystyle \frac {\sum_{i=1}^{n} (y_{i} - (\beta_0 \cdot 1+ \beta_1 \cdot x))}{n}\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1"></a>fit_W <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span>, <span class="at">data =</span> biochem)</span>
<span id="cb33-2"><a href="#cb33-2"></a></span>
<span id="cb33-3"><a href="#cb33-3"></a><span class="fu">summary</span>(fit_W)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ 1, data = biochem)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.3409 -2.4409 -0.4409  2.2591  9.9591 

Coefficients:
            Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept) 20.24091    0.07903   256.1   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.336 on 1781 degrees of freedom</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1"></a>biochem_W <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_W, <span class="at">data =</span> biochem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<h4 id="super-nice-way-of-visualizing">Super nice way of <a href="https://drsimonj.svbtle.com/visualising-residuals">visualizing</a></h4>
<p>First, let’s look at <span class="math inline">\(SS_{mean}\)</span> for avg weight:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1"></a>p_W <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem_W, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb36-2"><a href="#cb36-2"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> biochem_W<span class="sc">$</span>.fitted, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span> <span class="co"># plot linear model fit</span></span>
<span id="cb36-3"><a href="#cb36-3"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> <span class="co"># plot height as points and color code by the value of the residual</span></span>
<span id="cb36-4"><a href="#cb36-4"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb36-5"><a href="#cb36-5"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">25</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb36-6"><a href="#cb36-6"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, let’s look at <span class="math inline">\(SS_{fit}\)</span> for the <span class="math inline">\(weight\)</span> ~ <span class="math inline">\(cholesterol\)</span> :</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>p_WvC <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb37-2"><a href="#cb37-2"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="fu">pull</span>(chol_intercept), <span class="at">slope =</span> <span class="fu">pull</span>(chol_slope), <span class="at">col =</span> <span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb37-3"><a href="#cb37-3"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> <span class="co"># plot height as points and color code by the value of the residual</span></span>
<span id="cb37-4"><a href="#cb37-4"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb37-5"><a href="#cb37-5"></a>  <span class="co"># guides(color = FALSE) + # no legend for color code</span></span>
<span id="cb37-6"><a href="#cb37-6"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">1</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb37-7"><a href="#cb37-7"></a>  <span class="fu">theme_minimal</span>()</span>
<span id="cb37-8"><a href="#cb37-8"></a></span>
<span id="cb37-9"><a href="#cb37-9"></a></span>
<span id="cb37-10"><a href="#cb37-10"></a><span class="fu">plot_grid</span>(p_W, p_WvC, <span class="at">ncol =</span> <span class="dv">2</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"weight by intercept"</span>,<span class="st">"weight by cholesterol"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch"></section>
<section class="slide level2">

<h4 id="for-which-mice-does-the-model-fit-perform-the-most-poorly">For which mice does the model fit perform the most poorly?</h4>
<p>Gotta check residuals!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a><span class="co"># make new variable exception = absolute value of resid &gt; 9 then subject id</span></span>
<span id="cb38-2"><a href="#cb38-2"></a></span>
<span id="cb38-3"><a href="#cb38-3"></a></span>
<span id="cb38-4"><a href="#cb38-4"></a></span>
<span id="cb38-5"><a href="#cb38-5"></a>biochem_WvC<span class="sc">$</span>exceptions <span class="ot">&lt;-</span> <span class="fu">if_else</span>(</span>
<span id="cb38-6"><a href="#cb38-6"></a>  <span class="at">condition =</span> <span class="fu">abs</span>(biochem_WvC<span class="sc">$</span>.resid) <span class="sc">&lt;</span> <span class="dv">9</span>,</span>
<span id="cb38-7"><a href="#cb38-7"></a>  <span class="at">true =</span> <span class="st">""</span>,</span>
<span id="cb38-8"><a href="#cb38-8"></a>  <span class="at">false =</span> biochem_WvC<span class="sc">$</span>subject_name</span>
<span id="cb38-9"><a href="#cb38-9"></a>  )</span>
<span id="cb38-10"><a href="#cb38-10"></a></span>
<span id="cb38-11"><a href="#cb38-11"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight, <span class="at">label =</span> exceptions)) <span class="sc">+</span></span>
<span id="cb38-12"><a href="#cb38-12"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="fu">ifelse</span>(biochem_WvC<span class="sc">$</span>exceptions <span class="sc">==</span> <span class="st">""</span>, <span class="st">"grey50"</span>, <span class="st">"red"</span>)) <span class="sc">+</span></span>
<span id="cb38-13"><a href="#cb38-13"></a>  <span class="fu">geom_text_repel</span>() <span class="sc">+</span></span>
<span id="cb38-14"><a href="#cb38-14"></a><span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span> <span class="co"># plot height as points and color code by the value of the residual</span></span>
<span id="cb38-15"><a href="#cb38-15"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span> <span class="co"># color code for plotting residuals</span></span>
<span id="cb38-16"><a href="#cb38-16"></a>  <span class="co"># guides(color = FALSE) + # no legend for color code</span></span>
<span id="cb38-17"><a href="#cb38-17"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">1</span>) <span class="sc">+</span> <span class="co"># plot line representing residuals</span></span>
<span id="cb38-18"><a href="#cb38-18"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-9-1.png" width="960" class="r-stretch"></section>
<section class="slide level2">

<h4 id="big-baby-and-kryptonate">Big Baby and KryptoNate</h4>
<p>Nate Robinson aka “KryptoNate”<br>
5’ 9”<br>
180 lbs</p>
<p><img data-src="img/Baby_Nate.jpg" style="width:15.0%"></p>
<p>Glen Davis aka “Big Baby”<br>
6’ 9”<br>
280 lbs<br>
<img data-src="img/Baby_Nate2.jpg" style="width:15.0%"></p>
</section>
<section class="slide level2">

<p><span class="math inline">\(R^2\)</span> or coefficient of determination — the proportion of the variance in the dependent variable that is predictable from the independent variable(s).</p>
<p><span class="math inline">\(R^2 = \displaystyle \frac {SS_{mean} - SS_{fit}}{SS_{mean}}\)</span></p>
<p><span class="math inline">\(R^2 = \displaystyle \frac {SS_{w} - SS_{wc}}{SS_{w}}\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>ss.fit <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_WvC<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb39-2"><a href="#cb39-2"></a></span>
<span id="cb39-3"><a href="#cb39-3"></a></span>
<span id="cb39-4"><a href="#cb39-4"></a>ss.mean <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_W<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb39-5"><a href="#cb39-5"></a></span>
<span id="cb39-6"><a href="#cb39-6"></a></span>
<span id="cb39-7"><a href="#cb39-7"></a><span class="co"># Calc R^2 value</span></span>
<span id="cb39-8"><a href="#cb39-8"></a>biochem_WvC_rsq <span class="ot">&lt;-</span> (ss.mean <span class="sc">-</span> ss.fit) <span class="sc">/</span> ss.mean </span>
<span id="cb39-9"><a href="#cb39-9"></a>biochem_WvC_rsq</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1253678</code></pre>
</div>
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1"></a><span class="fu">glance</span>(fit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(r.squared)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1253678</code></pre>
</div>
</div>
<h4 id="interpretation-of-r2">Interpretation of <span class="math inline">\(R^2\)</span></h4>
<p>There is a 13% reduction in the variance when we take mouse <span class="math inline">\(cholesterol\)</span> into account<br>
OR<br>
Mouse <span class="math inline">\(cholesterol\)</span> explains 13% in player <span class="math inline">\(weight\)</span></p>
<p>By the way, this is the same <span class="math inline">\(R\)</span> as from the Pearson correlation:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1"></a><span class="co"># Pearson correlation R value</span></span>
<span id="cb43-2"><a href="#cb43-2"></a><span class="fu">cor</span>(biochem<span class="sc">$</span>weight, biochem<span class="sc">$</span>tot_cholesterol, <span class="at">method =</span> <span class="st">"pearson"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.3540731</code></pre>
</div>
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1"></a><span class="co"># Pearson correlation R^2 value</span></span>
<span id="cb45-2"><a href="#cb45-2"></a><span class="fu">cor</span>(biochem<span class="sc">$</span>weight, biochem<span class="sc">$</span>tot_cholesterol, <span class="at">method =</span> <span class="st">"pearson"</span>)<span class="sc">^</span><span class="dv">2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1253678</code></pre>
</div>
</div>
</section>
<section class="slide level2">

<h4 id="the-f-statistic">The F-statistic</h4>
<p><strong>F-statistic</strong> — the proportion of the variance in the dependent variable that is predictable from the independent variable(s).</p>
<p><span class="math inline">\(F = \displaystyle \frac{SS_{fit}/(p_{fit}-p_{mean})} {SS_{mean}/(n-p_{fit})}\)</span></p>
<p><span class="math inline">\(p_{fit}\)</span> — number of parameters in the fit line<br>
<span class="math inline">\(p_{mean}\)</span> — number of parameters in the mean line<br>
<span class="math inline">\(n\)</span> — number of data points</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1"></a><span class="co"># F-value</span></span>
<span id="cb47-2"><a href="#cb47-2"></a>biochem_WvC_F <span class="ot">&lt;-</span> ((ss.mean <span class="sc">-</span> ss.fit) <span class="sc">/</span> (<span class="dv">2</span> <span class="sc">-</span> <span class="dv">1</span>)) <span class="sc">/</span> </span>
<span id="cb47-3"><a href="#cb47-3"></a>  (ss.fit <span class="sc">/</span> (<span class="fu">nrow</span>(biochem_WvC) <span class="sc">-</span> <span class="dv">2</span>))</span>
<span id="cb47-4"><a href="#cb47-4"></a></span>
<span id="cb47-5"><a href="#cb47-5"></a>biochem_WvC_F </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 255.1411</code></pre>
</div>
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1"></a><span class="fu">glance</span>(fit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(statistic)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   value 
255.1411 </code></pre>
</div>
</div>
<h4 id="p-value-from-the-f-statistic">P-value from the F-statistic</h4>
<p>We need to generate a null distribution of <span class="math inline">\(F-statistic\)</span> values to compare to our observed <span class="math inline">\(F-statistic\)</span>.</p>
<p>Therefore, we will randomize the player Height and Weight and then calculate the <span class="math inline">\(F-statistic\)</span>.</p>
<p><img data-src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" style="width:10.0%"></p>
<p><br></p>
<p>We will do this many many times to generate a null distribution of <span class="math inline">\(F-statistic\)</span>s.<br>
<br></p>
<p><img data-src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" style="width:5.0%"> <img data-src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" style="width:5.0%"><br>
<img data-src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" style="width:5.0%"> <img data-src="https://media.giphy.com/media/3o7TKDedZiHXRJNveU/giphy.gif" style="width:5.0%"> <br></p>
<p>The p-value will be the probability of obtaining an <span class="math inline">\(F-statistic\)</span> in the null distribution at least as extreme as our observed <span class="math inline">\(F-statistic\)</span>.</p>
<p>Another beautiful Statquest explaining <a href="https://www.youtube.com/watch?v=nk2CQITm_eo">how to go from F-statistic to p-value</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1"></a><span class="co"># set up an empty tibble to hold our null distribution</span></span>
<span id="cb51-2"><a href="#cb51-2"></a>fake_biochem <span class="ot">&lt;-</span> <span class="fu">tribble</span>()</span>
<span id="cb51-3"><a href="#cb51-3"></a></span>
<span id="cb51-4"><a href="#cb51-4"></a></span>
<span id="cb51-5"><a href="#cb51-5"></a><span class="co"># sample function to randomize/permute data</span></span>
<span id="cb51-6"><a href="#cb51-6"></a><span class="fu">sample</span>(<span class="at">x =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  3  1  7  4  6  5  2  8 10  9</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1"></a><span class="co"># we will perform 100 permutations</span></span>
<span id="cb53-2"><a href="#cb53-2"></a>myPerms <span class="ot">&lt;-</span> <span class="dv">100</span></span>
<span id="cb53-3"><a href="#cb53-3"></a></span>
<span id="cb53-4"><a href="#cb53-4"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>myPerms) {</span>
<span id="cb53-5"><a href="#cb53-5"></a></span>
<span id="cb53-6"><a href="#cb53-6"></a>  tmp <span class="ot">&lt;-</span> <span class="fu">bind_cols</span>(</span>
<span id="cb53-7"><a href="#cb53-7"></a>    biochem_WvC[<span class="fu">sample</span>(<span class="fu">nrow</span>(biochem_WvC)), <span class="st">"weight"</span>],</span>
<span id="cb53-8"><a href="#cb53-8"></a>    biochem_WvC[<span class="fu">sample</span>(<span class="fu">nrow</span>(biochem_WvC)), <span class="st">"tot_cholesterol"</span>],</span>
<span id="cb53-9"><a href="#cb53-9"></a>    <span class="st">"perm"</span><span class="ot">=</span><span class="fu">factor</span>(<span class="fu">rep</span>(i,<span class="fu">nrow</span>(biochem_WvC)))</span>
<span id="cb53-10"><a href="#cb53-10"></a>    )</span>
<span id="cb53-11"><a href="#cb53-11"></a></span>
<span id="cb53-12"><a href="#cb53-12"></a>  fake_biochem <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(fake_biochem,tmp)</span>
<span id="cb53-13"><a href="#cb53-13"></a>  <span class="fu">rm</span>(tmp)</span>
<span id="cb53-14"><a href="#cb53-14"></a></span>
<span id="cb53-15"><a href="#cb53-15"></a>}</span>
<span id="cb53-16"><a href="#cb53-16"></a></span>
<span id="cb53-17"><a href="#cb53-17"></a></span>
<span id="cb53-18"><a href="#cb53-18"></a><span class="co"># let's look at permutations 1 and 2</span></span>
<span id="cb53-19"><a href="#cb53-19"></a><span class="fu">ggplot</span>(fake_biochem <span class="sc">%&gt;%</span> <span class="fu">filter</span>(perm <span class="sc">%in%</span> <span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)), <span class="fu">aes</span>(<span class="at">x=</span>weight, <span class="at">y=</span>tot_cholesterol, <span class="at">color=</span>perm)) <span class="sc">+</span></span>
<span id="cb53-20"><a href="#cb53-20"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb53-21"><a href="#cb53-21"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-13-1.png" width="960"></p>
</div>
</div>
<h4 id="remember-your-best-friend">Remember your best friend</h4>
<p><a href="https://cran.r-project.org/web/packages/broom/vignettes/broom_and_dplyr.html">BROOM</a></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1"></a><span class="co"># here we will calculate and extract linear model results for each permutation individualy using nest, mutate, and map functions</span></span>
<span id="cb54-2"><a href="#cb54-2"></a></span>
<span id="cb54-3"><a href="#cb54-3"></a>fake_biochem_lms <span class="ot">&lt;-</span> fake_biochem <span class="sc">%&gt;%</span></span>
<span id="cb54-4"><a href="#cb54-4"></a>  <span class="fu">nest</span>(<span class="at">data =</span> <span class="sc">-</span>perm) <span class="sc">%&gt;%</span></span>
<span id="cb54-5"><a href="#cb54-5"></a>  <span class="fu">mutate</span>(</span>
<span id="cb54-6"><a href="#cb54-6"></a>    <span class="at">fit =</span> <span class="fu">map</span>(data, <span class="sc">~</span> <span class="fu">lm</span>(weight <span class="sc">~</span> tot_cholesterol, <span class="at">data =</span> .x)),</span>
<span id="cb54-7"><a href="#cb54-7"></a>    <span class="at">glanced =</span> <span class="fu">map</span>(fit, glance)</span>
<span id="cb54-8"><a href="#cb54-8"></a>  ) <span class="sc">%&gt;%</span></span>
<span id="cb54-9"><a href="#cb54-9"></a>  <span class="fu">unnest</span>(glanced)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<h4 id="lets-take-a-look-at-the-null-distribution-of-f-statistics-from-the-randomized-values">Let’s take a look at the null distribution of F-statistics from the randomized values</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1"></a>fake_biochem_lms <span class="sc">%&gt;%</span></span>
<span id="cb55-2"><a href="#cb55-2"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(<span class="at">x =</span> statistic)) <span class="sc">+</span></span>
<span id="cb55-3"><a href="#cb55-3"></a>  <span class="fu">geom_density</span>(<span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb55-4"><a href="#cb55-4"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-15-1.png" width="960"></p>
</div>
</div>
<p>remember that the <span class="math inline">\(F-statistic\)</span> we observed was ~255!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1"></a>fake_biochem_lms <span class="sc">%&gt;%</span></span>
<span id="cb56-2"><a href="#cb56-2"></a>  <span class="fu">ggplot</span>(., <span class="fu">aes</span>(<span class="at">x =</span> statistic)) <span class="sc">+</span></span>
<span id="cb56-3"><a href="#cb56-3"></a>  <span class="fu">xlim</span>(<span class="dv">0</span>,biochem_WvC_F<span class="sc">*</span><span class="fl">1.1</span>) <span class="sc">+</span></span>
<span id="cb56-4"><a href="#cb56-4"></a>  <span class="fu">geom_density</span>(<span class="at">color=</span><span class="st">"red"</span>) <span class="sc">+</span></span>
<span id="cb56-5"><a href="#cb56-5"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> biochem_WvC_F, <span class="at">color =</span> <span class="st">"blue"</span>) <span class="sc">+</span></span>
<span id="cb56-6"><a href="#cb56-6"></a><span class="co">#  scale_x_log10() +</span></span>
<span id="cb56-7"><a href="#cb56-7"></a>  <span class="fu">theme_minimal</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-16-1.png" width="960"></p>
</div>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1"></a><span class="fu">glance</span>(fit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(p.value)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       value 
8.924155e-54 </code></pre>
</div>
</div>
<h4 id="interpretation-of-p-value">Interpretation of <span class="math inline">\(p-value\)</span></h4>
<p>There is no value more extreme than our observed <span class="math inline">\(F-statistic\)</span>.<br>
Therefore, the empirical <span class="math inline">\(p-value &lt; 0.001\)</span> — empirical what we calculated by randomizing our data.</p>
<p>The exact <span class="math inline">\(p-value\)</span> is 8.9241551^{-54} AKA <em>“a ridiculously small number”</em>.</p>
<p><strong>Correct</strong> — Assuming that <span class="math inline">\(cholesterol\)</span> has zero effect on <span class="math inline">\(weight\)</span> in the population, you’d obtain the sample effect, or larger, in 8.9241551^{-54} AKA <em>“a ridiculously small number”</em> of studies because of random sample error.</p>
<p><strong>Incorrect</strong> —There’s a 8.9241551^{-54} AKA <em>“a ridiculously small number”</em> chance of making a mistake by rejecting the null hypothesis.</p>
<p>More on <a href="https://statisticsbyjim.com/hypothesis-testing/interpreting-p-values/">p-value (mis)interpretation</a></p>
</section>
<section class="slide level2">

<h3 id="how-to-find-the-best-least-squares-fit">How to find the best (least squares) fit?</h3>
<ol type="1">
<li>Rotate the line of fit<br>
</li>
<li>Find the fit that minimizes the Sum of Squared Residuals or <span class="math inline">\(SS_{fit}\)</span><br>
</li>
<li>This is the derivative (slope of tangent at best point = 0) of the function describing the <span class="math inline">\(SS_{fit}\)</span> and the next rotation is 0.</li>
</ol>
<p><a href="https://youtu.be/PaFPbb66DxQ">StatQuest: Fitting a line to data, aka least squares, aka linear regression.</a></p>
<p><a href="https://youtu.be/sDv4f4s2SB8">StatQuest: Gradient Descent, Step-by-Step</a></p>
</section>
<section class="slide level2">

<h3 id="non-parametric-version">Non-parametric version</h3>
<h4 id="rank">Theory: rank-transformation</h4>
<p><code>rank</code> simply takes a list of numbers and “replace” them with the integers of their rank (1st smallest, 2nd smallest, 3rd smallest, etc.). So the result of the rank-transformation <code>rank(c(3.6, 3.4, -5.0, 8.2))</code> is <code>3, 2, 1, 4</code>. See that in the figure above?</p>
<p>A <em>signed</em> rank is the same, just where we rank according to absolute size first and then add in the sign second. So the signed rank here would be <code>2, 1, -3, 4</code>. Or in code:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1"></a>npfit_WvC <span class="ot">&lt;-</span> <span class="fu">lm</span>(<span class="at">formula =</span> <span class="fu">rank</span>(weight) <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> <span class="fu">rank</span>(tot_cholesterol), <span class="at">data =</span> biochem)</span>
<span id="cb59-2"><a href="#cb59-2"></a></span>
<span id="cb59-3"><a href="#cb59-3"></a><span class="fu">tidy</span>(npfit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 2 × 5
  term                  estimate std.error statistic   p.value
  &lt;chr&gt;                    &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
1 (Intercept)            571.      22.8         25.1 4.01e-119
2 rank(tot_cholesterol)    0.360    0.0221      16.3 1.54e- 55</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1"></a><span class="fu">glance</span>(npfit_WvC) <span class="sc">%&gt;%</span> <span class="fu">pull</span>(r.squared)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.1293324</code></pre>
</div>
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1"></a><span class="co"># cor(biochem$weight, biochem$tot_cholesterol, method = "spearman")^2</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="multiple-regression" class="slide level2">
<h2>Multiple regression</h2>
<p>Remember:</p>
<p><span class="math inline">\(y = \beta_0 \cdot 1+ \beta_1 \cdot x\)</span></p>
<p>Let’s add an explanatory variable:</p>
<p><span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x + \beta_2 \cdot z\)</span></p>
<p>linear model equation: <span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol + \beta_2 \cdot sodium\)</span></p>
<p><span class="math inline">\(\mathcal{H}_0:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> and <span class="math inline">\(sodium\)</span> does NOT explain <span class="math inline">\(weight\)</span></p>
<p>Null Hypothesis: <span class="math inline">\(\mathcal{H}_0: \beta_1, \beta_2 = 0\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + 0 \cdot cholesterol + 0 \cdot sodium\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1\)</span></p>
<p>Alternative Hypothesis: <span class="math inline">\(\mathcal{H}_1: \beta_1,\beta_2 \neq 0\)</span><br>
</p>
<p>Full model: <span class="math inline">\(y = \beta_0 \cdot 1 + \beta_1 \cdot x + \beta_2 \cdot z\)</span><br>
<span class="math inline">\(\mathcal{H}_1:\)</span> Mouse <span class="math inline">\(cholesterol\)</span> and <span class="math inline">\(sodium\)</span> does explain <span class="math inline">\(weight\)</span></p>
<p><span class="math inline">\(weight = \beta_0 \cdot 1 + \beta_1 \cdot cholesterol + \beta_2 \cdot sodium\)</span></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1"></a><span class="co"># how do different variables correlate with weight?</span></span>
<span id="cb64-2"><a href="#cb64-2"></a>biochem <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3"></a>  <span class="fu">select_if</span>(is.numeric) <span class="sc">%&gt;%</span></span>
<span id="cb64-4"><a href="#cb64-4"></a>  <span class="fu">cor</span>() <span class="sc">%&gt;%</span></span>
<span id="cb64-5"><a href="#cb64-5"></a>  as.data.frame <span class="sc">%&gt;%</span></span>
<span id="cb64-6"><a href="#cb64-6"></a>  <span class="fu">select</span>(weight) <span class="sc">%&gt;%</span></span>
<span id="cb64-7"><a href="#cb64-7"></a>  <span class="fu">arrange</span>(<span class="sc">-</span>weight) <span class="sc">%&gt;%</span> <span class="fu">rownames</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "weight"          "tot_cholesterol" "age"             "sodium"         
[5] "glucose"         "calcium"         "litter"          "cage_density"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1"></a><span class="co"># does sodium + cholesterol predict weight better than cholesterol alone?</span></span>
<span id="cb66-2"><a href="#cb66-2"></a>fit_WvC_S <span class="ot">&lt;-</span>  <span class="fu">lm</span>(<span class="at">data =</span> biochem, <span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> tot_cholesterol <span class="sc">+</span> sodium)  </span>
<span id="cb66-3"><a href="#cb66-3"></a></span>
<span id="cb66-4"><a href="#cb66-4"></a></span>
<span id="cb66-5"><a href="#cb66-5"></a>fit_WvC_S <span class="sc">%&gt;%</span> <span class="fu">glance</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 12
  r.squared adj.r.squared sigma statistic  p.value    df logLik   AIC   BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     0.171         0.170  3.04      183. 3.72e-73     2 -4508. 9024. 9046.
# ℹ 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a>biochem_WvC_S <span class="ot">&lt;-</span> <span class="fu">augment</span>(fit_WvC_S, <span class="at">data =</span> biochem)</span>
<span id="cb68-2"><a href="#cb68-2"></a></span>
<span id="cb68-3"><a href="#cb68-3"></a>ss.fit_C_S <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_WvC_S<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb68-4"><a href="#cb68-4"></a></span>
<span id="cb68-5"><a href="#cb68-5"></a>var.fit_C_S <span class="ot">&lt;-</span> ss.fit_C_S<span class="sc">/</span><span class="fu">nrow</span>(biochem_WvC_S)</span>
<span id="cb68-6"><a href="#cb68-6"></a> </span>
<span id="cb68-7"><a href="#cb68-7"></a></span>
<span id="cb68-8"><a href="#cb68-8"></a>ss.fit_C <span class="ot">&lt;-</span> <span class="fu">sum</span>(biochem_WvC<span class="sc">$</span>.resid<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb68-9"><a href="#cb68-9"></a></span>
<span id="cb68-10"><a href="#cb68-10"></a>var.fit_C <span class="ot">&lt;-</span> ss.fit_C<span class="sc">/</span><span class="fu">nrow</span>(biochem_WvC)</span>
<span id="cb68-11"><a href="#cb68-11"></a></span>
<span id="cb68-12"><a href="#cb68-12"></a></span>
<span id="cb68-13"><a href="#cb68-13"></a></span>
<span id="cb68-14"><a href="#cb68-14"></a>fit_Wvall <span class="ot">&lt;-</span>  <span class="fu">lm</span>(<span class="at">data =</span> biochem, <span class="at">formula =</span> weight <span class="sc">~</span> <span class="dv">1</span> <span class="sc">+</span> tot_cholesterol <span class="sc">+</span> age <span class="sc">+</span> sodium <span class="sc">+</span> glucose <span class="sc">+</span> calcium <span class="sc">+</span> litter <span class="sc">+</span> cage_density)  </span>
<span id="cb68-15"><a href="#cb68-15"></a></span>
<span id="cb68-16"><a href="#cb68-16"></a>fit_Wvall <span class="sc">%&gt;%</span> <span class="fu">glance</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 12
  r.squared adj.r.squared sigma statistic   p.value    df logLik   AIC   BIC
      &lt;dbl&gt;         &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
1     0.253         0.250  2.89      85.7 1.33e-107     7 -4415. 8849. 8898.
# ℹ 3 more variables: deviance &lt;dbl&gt;, df.residual &lt;int&gt;, nobs &lt;int&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1"></a>fit_Wvall <span class="sc">%&gt;%</span> <span class="fu">tidy</span>() </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 8 × 5
  term            estimate std.error statistic  p.value
  &lt;chr&gt;              &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;    &lt;dbl&gt;
1 (Intercept)      -6.53      1.68       -3.90 9.98e- 5
2 tot_cholesterol   1.97      0.109      18.2  9.08e-68
3 age               0.133     0.0161      8.28 2.49e-16
4 sodium            0.179     0.0145     12.4  1.12e-33
5 glucose           0.159     0.0284      5.59 2.59e- 8
6 calcium          -5.65      0.692      -8.16 6.13e-16
7 litter           -0.0557    0.0523     -1.06 2.88e- 1
8 cage_density     -0.316     0.0648     -4.87 1.22e- 6</code></pre>
</div>
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> biochem, <span class="fu">aes</span>(<span class="at">y =</span> weight, <span class="at">x =</span> calcium)) <span class="sc">+</span></span>
<span id="cb72-2"><a href="#cb72-2"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb72-3"><a href="#cb72-3"></a>  <span class="fu">scale_color_manual</span>() <span class="sc">+</span></span>
<span id="cb72-4"><a href="#cb72-4"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="stats-class-03_files/figure-revealjs/multiple regression-1.png" width="960" class="r-stretch"></section>
<section id="non-linear-regression" class="slide level2">
<h2>Non-linear regression</h2>
<p><strong>Non-linear regression</strong> — observational data are modeled by a function which is a nonlinear combination of the model parameters and depends on one or more independent variables. The data are fitted by a method of successive approximations.</p>
<p><strong>DANGER OVERFITTING</strong></p>
<h4 id="perform-loess-regression">Perform loess regression</h4>
<p><strong>Loess regression</strong> — a non-parametric technique that uses local weighted regression to fit a smooth curve through points in a scatter plot. LOESS combines much of the simplicity of linear least squares regression with the flexibility of nonlinear regression. It does this by fitting simple models to localized subsets of the data to build up a function that describes the deterministic part of the variation in the data, point by point. In fact, one of the chief attractions of this method is that the data analyst is not required to specify a global function of any form to fit a model to the data, only to fit segments of the data.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1"></a><span class="co"># local-weighted regression fit</span></span>
<span id="cb73-2"><a href="#cb73-2"></a>loessfit_WvC <span class="ot">&lt;-</span> <span class="fu">loess</span>(<span class="at">formula =</span> weight <span class="sc">~</span>  tot_cholesterol, <span class="at">data =</span> biochem)</span>
<span id="cb73-3"><a href="#cb73-3"></a></span>
<span id="cb73-4"><a href="#cb73-4"></a>loess_biochem_WvC <span class="ot">&lt;-</span> <span class="fu">augment</span>(loessfit_WvC, biochem) </span>
<span id="cb73-5"><a href="#cb73-5"></a></span>
<span id="cb73-6"><a href="#cb73-6"></a></span>
<span id="cb73-7"><a href="#cb73-7"></a><span class="fu">summary</span>(loessfit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Call:
loess(formula = weight ~ tot_cholesterol, data = biochem)

Number of Observations: 1782 
Equivalent Number of Parameters: 5.24 
Residual Standard Error: 3.112 
Trace of smoother matrix: 5.73  (exact)

Control settings:
  span     :  0.75 
  degree   :  2 
  family   :  gaussian
  surface  :  interpolate     cell = 0.2
  normalize:  TRUE
 parametric:  FALSE
drop.square:  FALSE </code></pre>
</div>
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1"></a><span class="fu">summary</span>(fit_WvC)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Call:
lm(formula = weight ~ 1 + tot_cholesterol, data = biochem)

Residuals:
    Min      1Q  Median      3Q     Max 
-8.9624 -2.1349 -0.2627  2.0113 10.2927 

Coefficients:
                Estimate Std. Error t value Pr(&gt;|t|)    
(Intercept)      14.5560     0.3635   40.04   &lt;2e-16 ***
tot_cholesterol   1.8516     0.1159   15.97   &lt;2e-16 ***
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

Residual standard error: 3.121 on 1780 degrees of freedom
Multiple R-squared:  0.1254,    Adjusted R-squared:  0.1249 
F-statistic: 255.1 on 1 and 1780 DF,  p-value: &lt; 2.2e-16</code></pre>
</div>
</div>
<h4 id="plot-loess-fit-depicting-residuals">Plot loess fit depicting residuals</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> loess_biochem_WvC, <span class="fu">aes</span>(<span class="at">x =</span> tot_cholesterol, <span class="at">y =</span> weight)) <span class="sc">+</span></span>
<span id="cb77-2"><a href="#cb77-2"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>lm, <span class="at">col =</span> <span class="st">"black"</span>, <span class="at">se =</span> F, <span class="at">size=</span>.<span class="dv">25</span>, <span class="at">linetype=</span><span class="st">"dashed"</span>) <span class="sc">+</span> <span class="co"># linear fit black dashed line</span></span>
<span id="cb77-3"><a href="#cb77-3"></a>  <span class="fu">geom_smooth</span>(<span class="at">method=</span>loess, <span class="at">col =</span> <span class="st">"red"</span>, <span class="at">se =</span> F, <span class="at">size=</span>.<span class="dv">25</span>) <span class="sc">+</span> <span class="co"># loess fit red line</span></span>
<span id="cb77-4"><a href="#cb77-4"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span>.<span class="dv">5</span>, <span class="fu">aes</span>(<span class="at">color =</span> .resid)) <span class="sc">+</span></span>
<span id="cb77-5"><a href="#cb77-5"></a>  <span class="fu">scale_color_gradient2</span>(<span class="at">low =</span> <span class="st">"blue"</span>, <span class="at">mid =</span> <span class="st">"black"</span>, <span class="at">high =</span> <span class="st">"yellow"</span>) <span class="sc">+</span></span>
<span id="cb77-6"><a href="#cb77-6"></a>  <span class="co"># guides(color = FALSE) +</span></span>
<span id="cb77-7"><a href="#cb77-7"></a>  <span class="fu">geom_segment</span>(<span class="fu">aes</span>(<span class="at">xend =</span> tot_cholesterol, <span class="at">yend =</span> .fitted), <span class="at">alpha =</span> .<span class="dv">25</span>) <span class="sc">+</span></span>
<span id="cb77-8"><a href="#cb77-8"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="stats-class-03_files/figure-revealjs/unnamed-chunk-19-1.png" width="960" class="r-stretch"></section>
<section class="slide level2">

</section></section>
<section id="references-and-resources" class="title-slide slide level1 center">
<h1>References and resources</h1>
<p>I have borrowed heavily from, directly taken, and/or highly recommend the following fantastic resources:</p>
<ol type="1">
<li><a href="https://lindeloev.github.io/tests-as-linear/">Common statistical tests are linear models</a> from Jonas Kristoffer Lindeløv<br>
</li>
<li><a href="https://www.youtube.com/channel/UCtYLUTtgS3k1Fg4y5tAhLbw">Statquest</a></li>
<li><a href="https://www.rapidtables.com/math/symbols/Statistical_Symbols.html">Stats gobbledygook</a></li>
<li><a href="http://www.sthda.com/english/articles/39-regression-model-diagnostics/161-linear-regression-assumptions-and-diagnostics-in-r-essentials/">Linear Regression Assumptions and Diagnostics in R: Essentials</a><br>
</li>
<li><a href="https://www.graphpad.com/guides/prism/latest/statistics/stat_---_principles_of_statistics_-.htm">PRINCIPLES OF STATISTICS</a> from GraphPad/SAS.</li>
</ol>

<div class="footer footer-default">

</div>
</section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../../../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../../../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../../../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../../../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../../../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../../../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../../../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../../../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../../../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../../../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: false,

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        text: function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>