<!DOCTYPE html>
<html lang="en"><head>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-html/tabby.min.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/light-border.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.5.57">

  <meta name="author" content="Kristen Wells-Wrasman">
  <meta name="dcterms.date" content="2024-09-24">
  <title>MOLB 7950 – Single cell RNA-Seq</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="../site_libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    div.columns{display: flex; gap: min(4vw, 1.5em);}
    div.column{flex: auto; overflow-x: auto;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    ul.task-list li input[type="checkbox"] {
      width: 0.8em;
      margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
      vertical-align: middle;
    }
    /* CSS for syntax highlighting */
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; font-weight: bold; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; font-weight: bold; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="../site_libs/revealjs/dist/theme/quarto.css">
  <link href="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/font-awesome/css/all.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-chalkboard/style.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-drop/drop-runtime.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/reveal-pointer/pointer.css" rel="stylesheet">
  <link href="../site_libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-titled.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-titled) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-titled.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-titled .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-titled .callout-title  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-titled.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-titled.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-title {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-title {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-title {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-titled .callout-body > .callout-content > :last-child {
    padding-bottom: 0.5rem;
    margin-bottom: 0;
  }

  .callout.callout-titled .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-titled) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-title {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-title {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-title {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-title {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-title {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
  <link href="../site_libs/htmltools-fill-0.5.8.1/fill.css" rel="stylesheet">
  <script src="../site_libs/htmlwidgets-1.6.4/htmlwidgets.js"></script>
  <script src="../site_libs/viz-1.8.2/viz.js"></script>
  <link href="../site_libs/DiagrammeR-styles-0.2/styles.css" rel="stylesheet">
  <script src="../site_libs/grViz-binding-1.0.11/grViz.js"></script>
<meta property="og:title" content="Single cell RNA-Seq – MOLB 7950">
<meta property="og:description" content="Introduction and Quality Control">
<meta property="og:site_name" content="MOLB 7950">
<meta name="twitter:title" content="Single cell RNA-Seq – MOLB 7950">
<meta name="twitter:description" content="Introduction and Quality Control">
<meta name="twitter:image" content="https://rnabioco.github.io/molb-7950/slides/images/twitter-card.png">
<meta name="twitter:creator" content="@rnabioco">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="quarto-title-block center">
  <h1 class="title">Single cell RNA-Seq</h1>
  <p class="subtitle">Introduction and Quality Control</p>

<div class="quarto-title-authors">
<div class="quarto-title-author">
<div class="quarto-title-author-name">
Kristen Wells-Wrasman 
</div>
        <p class="quarto-title-affiliation">
            RNA Bioscience Initiative | CU Anschutz
          </p>
    </div>
</div>

  <p class="date">2024-09-24</p>
</section>
<section id="contact-info" class="slide level2">
<h2>Contact Info</h2>
<p>Greetings experimentalist humans 👋</p>
<p><i class="fa fa-envelope"></i> &nbsp; <a href="mailto:kristen.wells-wrasman@cuanschutz.edu">kristen.wells-wrasman@cuanschutz.edu</a> <br></p>
<p>RBI Informatics Fellows <a href="https://medschool.cuanschutz.edu/rbi/training-and-education/rbi-office-hours">Office hours</a></p>
<p><i class="fa fa-envelope"></i> &nbsp; <a href="mailto:rbi.fellows@cuanschutz.edu">rbi.fellows@cuanschutz.edu</a> <br></p>
</section>
<section id="learning-key" class="slide level2">
<h2>Learning key</h2>
<ul>
<li class="fragment">We will switch between lecture and your exercise <code>qmd</code>.</li>
<li class="fragment">To denote a slide that corresponds to your exercise, I will include ⌨️ in the slide title</li>
</ul>
</section>
<section id="learning-objectives" class="slide level2">
<h2>Learning Objectives</h2>
<div class="columns">
<div class="column">
<h3 id="lecture-1">Lecture 1</h3>
<ul>
<li>Identify key quality control issues with single cell RNA-seq data and perform filtering to exclude poor quality cells</li>
<li>Interact with single cell data using Bioconductor tools</li>
</ul>
</div><div class="column">
<h3 id="lecture-2">Lecture 2</h3>
<ul>
<li>Perform analysis to identify cell types in single cell data by using unsupervised clustering methods and comparing to public datasets</li>
<li>Describe the importance and reasoning for conducting each step of the analysis</li>
</ul>
</div></div>
</section>
<section id="single-cell-or-bulk" class="slide level2 smaller">
<h2>Single cell or bulk?</h2>
<div class="columns">
<div class="column">
<h3 id="single-cell">Single cell</h3>
<ul>
<li class="fragment">High level overview of general transcriptomic landscape of genes that are expressed highly at the single cell level
<ul>
<li class="fragment">Sequencing depth is low per cell so we only have confident detection of highly expressed genes</li>
<li class="fragment">Differential expression is less well developed to compare different conditions</li>
</ul></li>
<li class="fragment">Most techniques only capture the 5’ or 3’ end</li>
<li class="fragment">Good for identifying subpopulation of cells that change between conditions
<ul>
<li class="fragment">Doesn’t require prior knowledge of surface proteins to sort out a population</li>
<li class="fragment">Doesn’t average across all cells in the experiment</li>
</ul></li>
</ul>
</div><div class="column">
<h3 id="bulk">Bulk</h3>
<ul>
<li class="fragment">Global overview of transcriptomic landscape of an entire sample using high to low expressed genes.
<ul>
<li class="fragment">Sequencing depth is much deeper so there is higher confidence to detect mid to low range expression</li>
<li class="fragment">Good for novel transcript identification and assessing how global transcriptome changes between conditions</li>
</ul></li>
<li class="fragment">Captures full RNA molecule so can be used for RNA-splicing analysis</li>
<li class="fragment">Doesn’t work for subpopulation analysis
<ul>
<li class="fragment">All cells are averaged so determining what is happening to one cell type is challengin</li>
<li class="fragment">Are your results because of a transcriptomic change or a change in cell type frequencies?</li>
</ul></li>
</ul>
</div></div>
</section>
<section id="single-cell-sequencing" class="slide level2 smaller">
<h2>Single cell sequencing</h2>
<div class="columns">
<div class="column">
<p><img data-src="../img/block-rna/10x_bead_example.jpg"></p>
<p><img data-src="../img/block-rna/cb-umi-scheme.jpg"></p>
</div><div class="column">
<p><img data-src="../img/block-rna/drop-generation.png"></p>
</div></div>
</section>
<section id="where-is-the-barcode-and-umi" class="slide level2">
<h2>Where is the barcode and UMI?</h2>
<ul>
<li class="fragment"><p>Below is the design for the 3’ 10x genomics assay <img data-src="../img/block-rna/cb-umi-scheme.jpg"></p></li>
<li class="fragment"><p>Other assays may have different locations of there cell barcode and UMI.</p></li>
<li class="fragment"><p>These may also be different lengths.</p></li>
<li class="fragment"><p>Or even on read 2 (ex Parse bioscience).</p></li>
<li class="fragment"><p>Be sure to check what kit was used to prep your data and always perform sanity checks throughout the analysis!</p></li>
</ul>
</section>
<section id="library-prep" class="slide level2">
<h2>Library prep</h2>

<img data-src="../img/block-rna/10x-library-scheme.jpg" class="r-stretch"></section>
<section id="other-single-cell-methods" class="slide level2">
<h2>Other single cell methods</h2>
<p>droplet-based scRNA-seq: e.g.&nbsp;10x Genomics or Drop-Seq</p>
<p>Smart-seq based scRNA-seq: (bulk-RNA-seq on single cells in individual wells/tubes)</p>
<p><a href="http://cite-seq.com/">CITE-Seq</a>: gene expression + cell surface protein abundance</p>
<p><a href="https://www.10xgenomics.com/products/single-cell-immune-profiling">VDJ-Seq</a>: Gene expression + targeted sequencing of T-Cell and B-Cell receptors</p>
<p>Many others: ATAC, spatial transcriptomics, DNA sequencing, etc. (see <a href="https://www.nature.com/articles/s41576-019-0093-7">Integrative Single cell analysis</a>)</p>
</section>
<section id="overview-of-analysis-steps" class="slide level2">
<h2>Overview of analysis steps</h2>
<div class="cell">
<div class="cell-output-display">
<div class="grViz html-widget html-fill-item" id="htmlwidget-cdd05d50dfc926828cc1" style="width:60%;height:50%;"></div>
<script type="application/json" data-for="htmlwidget-cdd05d50dfc926828cc1">{"x":{"diagram":"\ndigraph workflow {\n  graph [layout = dot,\n         rankdir = TB]\n\n  node [shape = cicle,\n        style = filled,\n        fontcolor = black,\n        fontname = \"Helvetica\"]\n\n  # green\n  node [fillcolor = \"#009E73\"]\n  cr [label = \"Process FASTQ to UMI count matrix\n(Cellranger, Alevin, or STARsolo)\"]\n\n  # blue\n  node [fillcolor = \"#56B4E9\"]\n  cell_qc [label = \"QC cells\n (% mitochondrial UMIs,\n# of UMIs/Genes,\nremove empty droplets)\"]\n  norm [label = \"Normalize UMI counts\n(Normalize by deconvolution)\"]\n\n  # yellow\n  node [fillcolor = \"#F0E442\"]\n  feature [label = \"Identify variable genes\n(a.k.a Feature selection)\"]\n  dim_red [label = \"Dimensionality reduction (PCA)\"]\n  cluster [label = \"Clustering\n (using Shared Nearest Neighbors)\"]\n  viz [label = \"Make 2D-Visualization\n(PCA, UMAP, tSNE\nForce-directed graph)\"]\n  traj [label = \"Trajectory Inference\n(Slingshot,\nPAGA, scVelo)\"]\n\n  # blue\n  node [fillcolor = \"#56B4E9\"]\n\n  markers [label = \"Discover cell type markers\"]\n  annot [label = \"Annotate cell types\"]\n\n  # orange\n  node [fillcolor = \"#E69F00\"]\n  downstream  [label = \"Downstream analysis\n(Differential expression,\nShifts in cell type composition,\n Find new cell-types/states)\"]\n\n  edge [color = black\n        fontname = \"Helvetica\"]\n\n  cr -> cell_qc [label = \"Load into \nSingleCellExperiment\"]\n  cell_qc -> norm\n  norm -> feature\n  norm -> markers\n  feature -> dim_red\n  dim_red -> cluster\n  dim_red -> traj\n  dim_red -> viz\n  cluster -> markers\n  markers -> annot\n\n  edge [color = \"grey\"\n        style = \"dashed\"]\n  annot -> cell_qc [label = \"Repeat\n as needed\"]\n  annot -> feature\n  annot -> dim_red\n  annot -> cluster\n  annot -> downstream\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
<section id="from-raw-reads-to-a-umi-count-matrix" class="slide level2">
<h2>From raw reads to a UMI count matrix</h2>
<div class="columns">
<div class="column">
<p><img data-src="../img/block-rna/count-matrix.png"></p>
</div><div class="column">
<ul>
<li><a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger"><code>cellranger</code></a> from 10x Genomics (STAR)</li>
<li><a href="https://salmon.readthedocs.io/en/latest/alevin.html"><code>alevin</code></a> (Salmon)</li>
<li><a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md"><code>STAR-solo</code></a> (From the STAR developers)</li>
</ul>
</div></div>
</section>
<section id="alevin" class="slide level2">
<h2>Alevin</h2>
<div class="sourceCode" id="cb1"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a></a><span class="ex">$</span> salmon <span class="at">-h</span></span>
<span id="cb1-2"><a></a><span class="ex">salmon</span> v1.3.0</span>
<span id="cb1-3"><a></a></span>
<span id="cb1-4"><a></a><span class="ex">Usage:</span>  salmon <span class="at">-h</span><span class="kw">|</span><span class="ex">--help</span> or</span>
<span id="cb1-5"><a></a>        <span class="ex">salmon</span> <span class="at">-v</span><span class="kw">|</span><span class="ex">--version</span> or</span>
<span id="cb1-6"><a></a>        <span class="ex">salmon</span> <span class="at">-c</span><span class="kw">|</span><span class="ex">--cite</span> or</span>
<span id="cb1-7"><a></a>        <span class="ex">salmon</span> <span class="pp">[--</span><span class="ss">no</span><span class="pp">-</span><span class="ss">version</span><span class="pp">-</span><span class="ss">check</span><span class="pp">]</span> <span class="op">&lt;</span>COMMAND<span class="op">&gt;</span> [-h <span class="kw">|</span> <span class="ex">options]</span></span>
<span id="cb1-8"><a></a></span>
<span id="cb1-9"><a></a><span class="ex">Commands:</span></span>
<span id="cb1-10"><a></a>     <span class="ex">index</span>      : create a salmon index</span>
<span id="cb1-11"><a></a>     <span class="ex">quant</span>      : quantify a sample</span>
<span id="cb1-12"><a></a>     <span class="ex">alevin</span>     : single cell analysis <span class="co"># &lt;------</span></span>
<span id="cb1-13"><a></a>     <span class="ex">swim</span>       : perform super-secret operation</span>
<span id="cb1-14"><a></a>     <span class="ex">quantmerge</span> : merge multiple quantifications into a single file  </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="alevin-1" class="slide level2">
<h2>Alevin</h2>
<div class="sourceCode" id="cb2"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a></a><span class="ex">$</span> salmon index ...</span>
<span id="cb2-2"><a></a><span class="ex">$</span> salmon alevin </span>
<span id="cb2-3"><a></a>   <span class="ex">-l</span> ISR                    <span class="co"># library type</span></span>
<span id="cb2-4"><a></a>   <span class="ex">-1</span> read1.fastq.gz         <span class="co"># reads </span></span>
<span id="cb2-5"><a></a>   <span class="ex">-2</span> read2.fastq.gz         <span class="co"># reads</span></span>
<span id="cb2-6"><a></a>   <span class="ex">--chromiumv3</span>              <span class="co"># chemistry</span></span>
<span id="cb2-7"><a></a>   <span class="ex">-i</span> /path/to/salmon/index  <span class="co"># index path</span></span>
<span id="cb2-8"><a></a>   <span class="ex">-o</span> /path/to/output        <span class="co"># output</span></span>
<span id="cb2-9"><a></a>   <span class="ex">--tgMap</span> transcript_to_gene.tsv </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="alevin-output-files" class="slide level2">
<h2>Alevin output files</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a></a><span class="va">$ls</span> alevin/</span>
<span id="cb3-2"><a></a><span class="ex">alevin.log</span>          <span class="co"># run info</span></span>
<span id="cb3-3"><a></a><span class="ex">featureDump.txt</span>     <span class="co"># info on each cell barcode</span></span>
<span id="cb3-4"><a></a><span class="ex">quants_mat.gz</span>       <span class="co"># binary file with UMI counts </span></span>
<span id="cb3-5"><a></a><span class="ex">quants_mat_cols.txt</span> <span class="co"># genes in count matrix</span></span>
<span id="cb3-6"><a></a><span class="ex">quants_mat_rows.txt</span> <span class="co"># cell barcodes in count matrix</span></span>
<span id="cb3-7"><a></a><span class="ex">quants_tier_mat.gz</span>  <span class="co"># info about mapping for each gene</span></span>
<span id="cb3-8"><a></a><span class="ex">whitelist.txt</span>       <span class="co"># valid barcodes discovered by alevin</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="alevinqc" class="slide level2">
<h2>AlevinQC</h2>
<p>Can generate interactive QC reports using <a href="https://csoneson.github.io/alevinQC/">alevinQC</a></p>
<p><a href="../data/block-rna/scrna/qc/pbmc/alevinReport.html">PBMC QC report</a></p>
</section>
<section id="qc-cell-or-empty-droplet" class="slide level2 smaller">
<h2>QC: cell or empty droplet?</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><p>In a typical droplet scRNA-seq experiment 100k - 1M cell barcodes are detected, but only 1-10k cells are loaded</p></li>
<li><p>Most of these droplets are “empty” and contain very few reads.</p></li>
<li><p>What is the source of these reads in the “empty” droplets?</p></li>
<li><p>How do we determine if the data from a particular cell barcode is derived from a single cell?</p></li>
</ul>
</div><div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="slides-31_files/figure-revealjs/cb-umi-plot-1.png" width="864"></p>
</figure>
</div>
</div>
</div>
</div></div>
</section>
<section id="qc-cell-or-empty-droplet-1" class="slide level2 smaller">
<h2>QC: cell or empty droplet?</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="slides-31_files/figure-revealjs/unnamed-chunk-2-1.png" width="864"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<ul>
<li class="fragment"><p>Fit a curve to the observed data and identify point where first derivative is minimized.</p></li>
<li class="fragment"><p>Any barcodes less than the “knee”, test sequences for off-by-one errors against the barcodes above the knee.</p></li>
<li class="fragment"><p>Take top half of cells above the knee and train a classifier using multiple criteria (% mapping, % mitochondrial and rRNA reads, duplicate rate, …)</p></li>
<li class="fragment"><p>Classify ambiguous cells in lower half into likely cells or not.</p></li>
</ul>
</div></div>
</section>
<section id="qc-cell-or-empty-droplet-2" class="slide level2">
<h2>QC: Cell or empty droplet?</h2>
<div class="columns">
<div class="column">
<h4 id="good-data">Good data</h4>
<div class="cell">
<div class="cell-output-display">
<div>
<figure>
<p><img data-src="slides-31_files/figure-revealjs/plot-label-1.png" width="864"></p>
</figure>
</div>
</div>
</div>
</div><div class="column">
<h4 id="bad-data">Bad data</h4>
<p><img data-src="../img/block-rna/bad_data.png"></p>
</div></div>
</section>
<section id="cell-calling-sanity-check" class="slide level2 incremental">
<h2>Cell calling sanity check</h2>
<ul>
<li class="fragment">It’s always a good think to sanity check your data</li>
<li class="fragment">After calling cells, how can we perform a sanity check?</li>
<li class="fragment">I once reanalyzed data from a <em>published</em> paper and found that they had treated two separate sequencing runs as two separate captures. If this happens to you, what is a quick and easy way to make sure you treated your data correctly at the cell calling step?</li>
</ul>
</section>
<section id="doublets-and-multiplets" class="slide level2 smaller">
<h2>Doublets and Multiplets</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img data-src="../img/block-rna/10x-barnyard.png"></p>
</div><div class="column" style="width:50%;">
<ul>
<li class="fragment"><p>Doublets are not clearly identifiable using simple QC metrics, so cannot be reliably removed with filtering with # of UMIs or genes detected.</p></li>
<li class="fragment"><p><code>scran::doublet_cluster</code> : Compare each cluster to an in silico mix of two other clusters. Get per cluster score of likelihood of being a doublet.</p></li>
<li class="fragment"><p><code>scran::doublet_cell</code> : Compare each cell to a mix of two other randomly selected cells. Get per cell score of likelihood of being a doublet.</p></li>
<li class="fragment"><p>Doublets can also arise due to sample prep, e.g.&nbsp;incomplete generations of a single cell suspension. These doublets are difficult to exclude from the data</p></li>
</ul>
</div></div>
</section>
<section id="turning-to-our-exercise" class="slide level2">
<h2>Turning to our exercise ⌨️</h2>
<p>Before jumping into the analysis, let’s step back and start running through the exercise for today</p>
<p>Start by loading in the packages</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a></a><span class="co">#| label: load-libs</span></span>
<span id="cb4-2"><a></a><span class="co">#| message: false</span></span>
<span id="cb4-3"><a></a><span class="co">#| warning: false</span></span>
<span id="cb4-4"><a></a><span class="co">#| echo: false</span></span>
<span id="cb4-5"><a></a><span class="fu">library</span>(here)</span>
<span id="cb4-6"><a></a><span class="fu">library</span>(scran)</span>
<span id="cb4-7"><a></a><span class="fu">library</span>(scater)</span>
<span id="cb4-8"><a></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb4-9"><a></a><span class="fu">library</span>(DropletUtils)</span>
<span id="cb4-10"><a></a><span class="fu">library</span>(tximport)</span>
<span id="cb4-11"><a></a><span class="fu">library</span>(Matrix)</span>
<span id="cb4-12"><a></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb4-13"><a></a><span class="fu">library</span>(eds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="raw-data-the-umi-count-matrix" class="slide level2">
<h2>Raw data: the UMI count matrix ⌨️</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li><p>scRNA-seq libraries produce reads from 100,000 - 1,000,000 cell barcodes</p></li>
<li><p>A matrix of 20,000 genes x 1,000,000 barcodes is 20 billion values (!).</p></li>
<li><p>&gt;95% are zeros due to empty droplets and the low efficiency of the library prep (&lt; 10-20% of RNA captured).</p></li>
</ul>
</div><div class="column">
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a></a>vals <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb5-2"><a></a>  <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>,</span>
<span id="cb5-3"><a></a>  <span class="dv">2</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">0</span>,</span>
<span id="cb5-4"><a></a>  <span class="dv">0</span>, <span class="dv">0</span>, <span class="dv">1</span>, <span class="dv">0</span>, <span class="dv">1</span></span>
<span id="cb5-5"><a></a>)</span>
<span id="cb5-6"><a></a>m <span class="ot">&lt;-</span> <span class="fu">Matrix</span>(vals, <span class="at">ncol =</span> <span class="dv">5</span>)</span>
<span id="cb5-7"><a></a>m</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>3 x 5 sparse Matrix of class "dgCMatrix"
              
[1,] . . . . 1
[2,] . . . . .
[3,] . 2 . . 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a></a><span class="fu">summary</span>(m)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>3 x 5 sparse Matrix of class "dgCMatrix", with 3 entries
  i j x
1 3 2 2
2 1 5 1
3 3 5 1</code></pre>
</div>
</div>
</div></div>
</section>
<section id="sparse-matricies" class="slide level2">
<h2>Sparse matricies ⌨️</h2>
<ul>
<li class="fragment">Use the <code>as</code> function to convert to a sparse matrix</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>3 x 5 sparse Matrix of class "dgCMatrix"
              
[1,] . . . . 1
[2,] . . . . .
[3,] . 2 . . 1</code></pre>
</div>
</div>
<ul>
<li class="fragment">This only stores non-zero values</li>
<li class="fragment">Internally, values are stored as a row column value triplet</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>3 x 5 sparse Matrix of class "dgCMatrix", with 3 entries
  i j x
1 3 2 2
2 1 5 1
3 3 5 1</code></pre>
</div>
</div>
</section>
<section id="sparse-matricies-1" class="slide level2">
<h2>Sparse matricies ⌨️</h2>
<ul>
<li class="fragment">Functions at manipulate matricies (<code>rowMeans</code>, <code>colSums</code>, <code>apply</code>, <code>[</code>) can be used on sparseMatricies as long as the <code>Matrix</code> package is loaded.</li>
</ul>
<p>How can we extract the first 2 rows and first 3 columns of the sparse matrix <code>sm</code> that we generated above?</p>
</section>
<section id="sparse-matricies-2" class="slide level2">
<h2>Sparse matricies ⌨️</h2>
<ul>
<li class="fragment">Functions at manipulate matricies (<code>rowMeans</code>, <code>colSums</code>, <code>apply</code>, <code>[</code>) can be used on sparseMatricies as long as the <code>Matrix</code> package is loaded.</li>
</ul>
<p>How can we extract the first 2 rows and first 3 columns of the sparse matrix <code>sm</code> that we generated above?</p>
</section>
<section id="sparse-matricies-3" class="slide level2">
<h2>Sparse matricies ⌨️</h2>
<p>How can we calculate the sum of the columns of <code>sm</code>?</p>
</section>
<section id="sparse-matricies-4" class="slide level2">
<h2>Sparse matricies ⌨️</h2>
<p>How can we calculate the sum of the columns of <code>sm</code>?</p>
</section>
<section id="base-r-subsetting" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<ul>
<li class="fragment">Basic R concepts for subsetting and referencing columns are important in single cell analysis</li>
<li class="fragment">Vectors can be subset by index (position), logical vector (<code>c(TRUE, FALSE)</code>) or name (if vector is named)</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code> [1] "a" "b" "c" "d" "e" "f" "g" "h" "i" "j" "k" "l" "m" "n" "o" "p" "q" "r" "s"
[20] "t" "u" "v" "w" "x" "y" "z"</code></pre>
</div>
</div>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "b" "d" "f"</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-1" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "a" "e" "i" "o" "u"</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-2" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>  A   Z 
"a" "z" </code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-3" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<ul>
<li class="fragment">Matrices are 2 dimensional vectors and have similar subsetting rules except there are two dimensions, rows and columns.</li>
</ul>
<p><code>matrix[rows_to_subset, columns_to_subset]</code></p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3] [,4]
[1,]    2    8   14   20
[2,]    4   10   16   22
[3,]    6   12   18   24</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-4" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
[1,]    7   19
[2,]    8   20
[3,]    9   21
[4,]   10   22
[5,]   11   23
[6,]   12   24</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-5" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
[1,]    7   19
[2,]    8   20
[3,]    9   21</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-6" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2] [,3] [,4]
[1,]    4   10   16   22
[2,]    5   11   17   23
[3,]    6   12   18   24</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-7" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>     [,1] [,2]
[1,]    1    7
[2,]    2    8
[3,]    3    9
[4,]    4   10
[5,]    5   11
[6,]    6   12</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-8" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<p>The base R <code>data.frame</code> and Bioconductor <code>DataFrame</code> can also be subset with the <code>[</code> and we can reference individual vectors in a data.frame using <code>$</code>.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>               mpg cyl disp
Mazda RX4     21.0   6  160
Mazda RX4 Wag 21.0   6  160
Datsun 710    22.8   4  108</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-9" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code> [1] 21.0 21.0 22.8 21.4 18.7 18.1 14.3 24.4 22.8 19.2 17.8 16.4 17.3 15.2 10.4
[16] 10.4 14.7 32.4 30.4 33.9 21.5 15.5 15.2 13.3 19.2 27.3 26.0 30.4 15.8 19.7
[31] 15.0 21.4</code></pre>
</div>
</div>
</section>
<section id="base-r-subsetting-10" class="slide level2">
<h2>Base R subsetting ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>                mpg cyl  disp  hp drat   wt  qsec vs am gear carb
Mazda RX4      21.0   6 160.0 110 3.90 2620 16.46  0  1    4    4
Mazda RX4 Wag  21.0   6 160.0 110 3.90 2875 17.02  0  1    4    4
Datsun 710     22.8   4 108.0  93 3.85 2320 18.61  1  1    4    1
Hornet 4 Drive 21.4   6 258.0 110 3.08 3215 19.44  1  0    3    1
Merc 240D      24.4   4 146.7  62 3.69 3190 20.00  1  0    4    2
Merc 230       22.8   4 140.8  95 3.92 3150 22.90  1  0    4    2
Fiat 128       32.4   4  78.7  66 4.08 2200 19.47  1  1    4    1
Honda Civic    30.4   4  75.7  52 4.93 1615 18.52  1  1    4    2
Toyota Corolla 33.9   4  71.1  65 4.22 1835 19.90  1  1    4    1
Toyota Corona  21.5   4 120.1  97 3.70 2465 20.01  1  0    3    1
Fiat X1-9      27.3   4  79.0  66 4.08 1935 18.90  1  1    4    1
Porsche 914-2  26.0   4 120.3  91 4.43 2140 16.70  0  1    5    2
Lotus Europa   30.4   4  95.1 113 3.77 1513 16.90  1  1    5    2
Volvo 142E     21.4   4 121.0 109 4.11 2780 18.60  1  1    4    2
               new_column_name
Mazda RX4               Hello!
Mazda RX4 Wag           Hello!
Datsun 710              Hello!
Hornet 4 Drive          Hello!
Merc 240D               Hello!
Merc 230                Hello!
Fiat 128                Hello!
Honda Civic             Hello!
Toyota Corolla          Hello!
Toyota Corona           Hello!
Fiat X1-9               Hello!
Porsche 914-2           Hello!
Lotus Europa            Hello!
Volvo 142E              Hello!</code></pre>
</div>
</div>
</section>
<section id="read-in-alevin-output-with-tximport" class="slide level2">
<h2>Read in Alevin output with tximport</h2>
<ul>
<li class="fragment">Now that we know how to work with a sparse matrix, let’s read in our data</li>
<li class="fragment"><code>tximport</code> has methods for importing the binary data from alevin</li>
<li class="fragment">We need to supply a path to the <code>quants_mat.gz</code> file.</li>
<li class="fragment">If you want to load multiple samples use iteration approaches (e.g.&nbsp;<code>lapply</code>, <code>purrr::map</code>, a <code>for</code> loop).</li>
<li class="fragment">Also note that the <a href="https://bioconductor.org/packages/release/bioc/html/eds.html"><code>eds</code></a> package was installed which greatly speeds up the loading of the matrix.</li>
</ul>
<p>We will load in data from a 10x Genomics scRNA-seq library generated from human periperhal blood mononuclear cells (PMBCS).</p>
</section>
<section id="read-in-alevin-output-with-tximport-1" class="slide level2">
<h2>Read in Alevin output with tximport ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "abundance"           "counts"              "countsFromAbundance"</code></pre>
</div>
</div>
<p><code>tx</code> is a list with 3 elements, <code>abundance</code>, <code>counts</code>, and <code>countsFromAbundance</code>. Let’s look at the counts element</p>
</section>
<section id="read-in-alevin-output-with-tximport-2" class="slide level2">
<h2>Read in Alevin output with tximport ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>6 x 3 sparse Matrix of class "dgCMatrix"
                GCTGCAGTCCGATCTC ACTATGGAGGTCCCTG ATTTCTGTCTCTATGT
ENSG00000243485                .                .                .
ENSG00000284332                .                .                .
ENSG00000237613                .                .                .
ENSG00000268020                .                .                .
ENSG00000290826                .                .                .
ENSG00000240361                .                .                .</code></pre>
</div>
</div>
<p>Here you can see that <code>tx$counts</code> is a sparse matrix that is genes (rows) by cells (columns).</p>
<p>How many barcodes are in <code>tx$counts</code>? How many genes?</p>
</section>
<section id="read-in-alevin-output-with-tximport-3" class="slide level2">
<h2>Read in Alevin output with tximport ⌨️</h2>
<p>What fraction of the matrix is non-zero? We can use the <code>nnzero</code>function from the <code>Matrix</code> package check</p>
</section>
<section id="read-in-alevin-output-with-tximport-4" class="slide level2">
<h2>Read in Alevin output with tximport ⌨️</h2>
</section>
<section id="single-cell-analysis-packages" class="slide level2">
<h2>Single cell analysis packages</h2>
<p>Key resource for single cell analysis in Bioconductor: <a href="https://bioconductor.org/books/release/OSCA/">Orchestrating Single Cell Analysis</a></p>
<p><a href="https://bioconductor.org/packages/release/bioc/html/SingleCellExperiment.html"><code>SingleCellExperiment</code></a> is the core datastructure for storing single cell data.</p>
<p><a href="https://bioconductor.org/packages/release/bioc/html/scran.html"><code>scran</code></a> provides algorithms for low-level processing of single cell data.</p>
<p><a href="https://bioconductor.org/packages/release/bioc/html/scater.html"><code>scater</code></a> provides plotting, data transformation, and quality control functionality</p>
</section>
<section id="singlecellexperiment" class="slide level2">
<h2>SingleCellExperiment</h2>

<img data-src="../img/block-rna/sce.png" class="r-stretch"></section>
<section id="create-a-singlecellexperiment-object" class="slide level2">
<h2>Create a SingleCellExperiment object ⌨️</h2>
<p>A <code>SingleCellExperiment</code> object can be created from our sparse matrix using the <code>SingleCellExperiment()</code> function.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 62266 6075 
metadata(0):
assays(1): counts
rownames(62266): ENSG00000290825 ENSG00000223972 ... ENSG00000210195
  ENSG00000210196
rowData names(0):
colnames(6075): GCTGCAGTCCGATCTC ACTATGGAGGTCCCTG ... ACGTAGGGTGACAGCA
  TCTCAGCTCGCCGAAC
colData names(0):
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</section>
<section id="exploring-the-singlecellexperiment-object" class="slide level2">
<h2>Exploring the SingleCellExperiment object ⌨️</h2>
<p>The <code>SingleCellExperiment</code> object stores the gene x cell count matrix within <code>assays()</code>.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>List of length 1
names(1): counts</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>4 x 4 sparse Matrix of class "dgCMatrix"
                GCTGCAGTCCGATCTC ACTATGGAGGTCCCTG ATTTCTGTCTCTATGT
ENSG00000290825                .                .                .
ENSG00000223972                .                .                .
ENSG00000227232                .                .                .
ENSG00000278267                .                .                .
                TATCTGTAGGTGATAT
ENSG00000290825                .
ENSG00000223972                .
ENSG00000227232                .
ENSG00000278267                .</code></pre>
</div>
</div>
</section>
<section id="exploring-the-singlecellexperiment-object-1" class="slide level2">
<h2>Exploring the SingleCellExperiment object ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>4 x 4 sparse Matrix of class "dgCMatrix"
                GCTGCAGTCCGATCTC ACTATGGAGGTCCCTG ATTTCTGTCTCTATGT
ENSG00000290825                .                .                .
ENSG00000223972                .                .                .
ENSG00000227232                .                .                .
ENSG00000278267                .                .                .
                TATCTGTAGGTGATAT
ENSG00000290825                .
ENSG00000223972                .
ENSG00000227232                .
ENSG00000278267                .</code></pre>
</div>
</div>
</section>
<section id="exploring-the-singlecellexperiment-object-2" class="slide level2">
<h2>Exploring the SingleCellExperiment object ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>4 x 4 sparse Matrix of class "dgCMatrix"
                GCTGCAGTCCGATCTC ACTATGGAGGTCCCTG ATTTCTGTCTCTATGT
ENSG00000290825                .                .                .
ENSG00000223972                .                .                .
ENSG00000227232                .                .                .
ENSG00000278267                .                .                .
                TATCTGTAGGTGATAT
ENSG00000290825                .
ENSG00000223972                .
ENSG00000227232                .
ENSG00000278267                .</code></pre>
</div>
</div>
</section>
<section id="accessing-cell-metadata" class="slide level2">
<h2>Accessing cell metadata ⌨️</h2>
<ul>
<li class="fragment">We generate cell-level information</li>
<li class="fragment">such as quality control metrics, clustering results, and celltype assignments.</li>
<li class="fragment">This data is stored within a data frame called <code>colData</code></li>
<li class="fragment">access using <code>colData()</code></li>
<li class="fragment">Specialized Bioconductor specific data.frame class (<code>DataFrame</code>) which has similar semantics and functionality to a base R data.frame.</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6075 rows and 0 columns</code></pre>
</div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6075 rows and 1 column
                 cell_source
                 &lt;character&gt;
GCTGCAGTCCGATCTC        PBMC
ACTATGGAGGTCCCTG        PBMC
ATTTCTGTCTCTATGT        PBMC
TATCTGTAGGTGATAT        PBMC
AGCCAGCCAAAGCACG        PBMC
...                      ...
CATCCCAAGTACTCGT        PBMC
CTCCTCCCATGAAGCG        PBMC
AGTTCCCCATGTCAGT        PBMC
ACGTAGGGTGACAGCA        PBMC
TCTCAGCTCGCCGAAC        PBMC</code></pre>
</div>
</div>
</section>
<section id="accessing-gene-metadata" class="slide level2">
<h2>Accessing gene metadata ⌨️</h2>
<ul>
<li class="fragment">Gene-level metadata is stored in a data.frame called <code>rowData()</code>.</li>
<li class="fragment">Use the rowData to store gene ids, symbols, and other information about genes.</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 62266 rows and 0 columns</code></pre>
</div>
</div>
</section>
<section id="manipulating-a-singlecellexperiment" class="slide level2">
<h2>Manipulating a SingleCellExperiment ⌨️</h2>
<ul>
<li class="fragment">Calculate the total number of counts in each cell and store these counts in the `colData().*</li>
</ul>
</section>
<section id="manipulating-a-singlecellexperiment-1" class="slide level2">
<h2>Manipulating a SingleCellExperiment ⌨️</h2>
<ul>
<li class="fragment">Calculate the total number of counts for each gene, summed across cells</li>
<li class="fragment">Calculate the number of cells with &gt; 0 counts per gene</li>
<li class="fragment">store both of these values in the rowData().</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 62266 rows and 2 columns
                total_gene_counts n_cells_expr
                        &lt;numeric&gt;    &lt;integer&gt;
ENSG00000290825         65.594607           43
ENSG00000223972          0.000000            0
ENSG00000227232          5.506349           24
ENSG00000278267          0.000000            0
ENSG00000243485          0.333333            1
...                           ...          ...
ENSG00000198695              2054         1509
ENSG00000210194                 0            0
ENSG00000198727            274421         5704
ENSG00000210195                 0            0
ENSG00000210196                 0            0</code></pre>
</div>
</div>
</section>
<section id="manipulating-a-singlecellexperiment-2" class="slide level2">
<h2>Manipulating a SingleCellExperiment ⌨️</h2>
<ul>
<li class="fragment">We can subset the SingleCellExperiment using the same techniques as base R data.</li>
<li class="fragment">Note that <code>dplyr</code> verbs do not work with <code>SingleCellExperiment</code></li>
</ul>
<div class="sourceCode" id="cb34"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a></a>data.frame[rows, columns]</span>
<span id="cb34-2"><a></a></span>
<span id="cb34-3"><a></a>sce[genes, cells]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="manipulating-a-singlecellexperiment-3" class="slide level2">
<h2>Manipulating a SingleCellExperiment ⌨️</h2>
<ul>
<li class="fragment">We can subset the SingleCellExperiment using the same techniques as base R data.</li>
<li class="fragment">Note that <code>dplyr</code> verbs do not work with <code>SingleCellExperiment</code></li>
</ul>
<div class="sourceCode" id="cb35"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a></a><span class="co"># subset to data from first 4 genes and cells</span></span>
<span id="cb35-2"><a></a>sce[<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>]</span>
<span id="cb35-3"><a></a></span>
<span id="cb35-4"><a></a><span class="co"># subset to cells from PBMC cells</span></span>
<span id="cb35-5"><a></a>sce[, sce<span class="sc">$</span>cell_source <span class="sc">==</span> <span class="st">"PBMC"</span>]</span>
<span id="cb35-6"><a></a></span>
<span id="cb35-7"><a></a></span>
<span id="cb35-8"><a></a>genes_to_keep <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ENSG00000223972"</span>, <span class="st">"ENSG00000210195"</span>, <span class="st">"ENSG00000210196"</span>)</span>
<span id="cb35-9"><a></a>sce[genes_to_keep, ]</span>
<span id="cb35-10"><a></a></span>
<span id="cb35-11"><a></a>cells_to_keep <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"ACTATGGAGGTCCCTG"</span>, <span class="st">"GCTGCAGTCCGATCTC"</span>, <span class="st">"TCTCAGCTCGCCGAAC"</span>)</span>
<span id="cb35-12"><a></a>sce[, cells_to_keep]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="manipulating-a-singlecellexperiment-4" class="slide level2">
<h2>Manipulating a SingleCellExperiment</h2>
<p><code>ncol()</code>: # of cells<br>
<code>nrow()</code>: # of gene<br>
<code>dims()</code>: # of genes and cells<br>
<code>rownames()</code>: rownames in matrices (e.g.&nbsp;genes)<br>
<code>colnames()</code>: colnames in matrices (e.g.&nbsp;cells)<br>
<code>cbind()</code>: combine multiple SingleCellExperiments by column<br>
<code>rbind()</code>: combine multiple SingleCellExperiments by row</p>
</section>
<section id="storing-gene-identifiers" class="slide level2">
<h2>Storing gene identifiers</h2>
<p>Ensembl gene ids are the rownames of our matrix (e.g.&nbsp;ENSG00000289576, ENSG00000221539). * These identifiers are guaranteed to be unique and are more stable and reliable than gene symbols (e.g.&nbsp;ACTB, GAPDH). * This becomes important if you want to compare to external datasets or ensure that your data can be easily used by others in the future. * But they aren’t easy to interpret</p>
</section>
<section id="storing-gene-identifiers-1" class="slide level2">
<h2>Storing gene identifiers ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 62266 rows and 4 columns
                total_gene_counts n_cells_expr        gene         gene_id
                        &lt;numeric&gt;    &lt;integer&gt; &lt;character&gt;     &lt;character&gt;
ENSG00000290825         65.594607           43     DDX11L2 ENSG00000290825
ENSG00000223972          0.000000            0     DDX11L1 ENSG00000223972
ENSG00000227232          5.506349           24      WASH7P ENSG00000227232
ENSG00000278267          0.000000            0   MIR6859-1 ENSG00000278267
ENSG00000243485          0.333333            1 MIR1302-2HG ENSG00000243485
...                           ...          ...         ...             ...
ENSG00000198695              2054         1509      MT-ND6 ENSG00000198695
ENSG00000210194                 0            0       MT-TE ENSG00000210194
ENSG00000198727            274421         5704      MT-CYB ENSG00000198727
ENSG00000210195                 0            0       MT-TT ENSG00000210195
ENSG00000210196                 0            0       MT-TP ENSG00000210196</code></pre>
</div>
</div>
</section>
<section id="updating-our-rownames" class="slide level2">
<h2>Updating our rownames ⌨️</h2>
<ul>
<li class="fragment">Goal rename rownames to symbols</li>
<li class="fragment">Problem, some are <code></code>, <code>NA</code>, or duplicated</li>
<li class="fragment"><code>uniquifyFeatureNames()</code> is a convenience function that will rename gene symbols that are <code>NA</code> or duplicated values to the ensembl ID or a combination of gene symbol and ensembl ID</li>
</ul>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] "DDX11L2_ENSG00000290825" "DDX11L1"                
[3] "WASH7P"                  "MIR6859-1"              
[5] "MIR1302-2HG"             "MIR1302-2"              </code></pre>
</div>
</div>
</section>
<section>
<section id="filtering-low-quality-cells" class="title-slide slide level1 center">
<h1>Filtering low quality cells ⌨️</h1>
<p>Next we perform some filtering and quality control to remove low expression genes and poor quality cells.</p>
<p>Our SingleCellExperiment has 62266 genes in the matrix. Most of these are not expressed. We want to exclude these genes as they won’t provide any useful data for the analysis.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 20858 6075 
metadata(0):
assays(1): counts
rownames(20858): DDX11L2_ENSG00000290825 WASH7P ... MT-ND6 MT-CYB
rowData names(5): total_gene_counts n_cells_expr gene gene_id n_cells
colnames(6075): GCTGCAGTCCGATCTC ACTATGGAGGTCCCTG ... ACGTAGGGTGACAGCA
  TCTCAGCTCGCCGAAC
colData names(2): cell_source total_counts
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</section>
<section id="filtering-low-quality-cells-1" class="slide level2">
<h2>Filtering low quality cells</h2>
<p>To exclude low-quality cells we will use the following metrics:</p>
<ul>
<li class="fragment">Number of counts per cell barcode</li>
<li class="fragment">Number of genes detected per barcode</li>
<li class="fragment">The percentage of counts from mitochondrial genes per barcode</li>
</ul>
<p>A low number of counts, a low number of detected genes, and a high percentage of mitochondrial counts suggests that the cell had a broken membrane and the cytoplasmic mRNA leaked out.</p>
</section>
<section id="filtering-low-quality-cells-2" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<p>To calculate these metrics we can use <code>addPerCellQCMetrics</code> from scater. Mitochondrial genes are named with a common “MT-” prefix (e.g.&nbsp;MT-CO2, MT-ATP6, MR-RNR2), which we can use to identify them.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6075 rows and 8 columns
                 cell_source total_counts       sum  detected subsets_Mito_sum
                 &lt;character&gt;    &lt;numeric&gt; &lt;numeric&gt; &lt;integer&gt;        &lt;numeric&gt;
GCTGCAGTCCGATCTC        PBMC        31924   31908.0      5718          2668.49
ACTATGGAGGTCCCTG        PBMC        35845   35801.5      6073          3963.23
ATTTCTGTCTCTATGT        PBMC        31788   31758.3      6377          2496.33
TATCTGTAGGTGATAT        PBMC        32025   31998.0      6260          2961.15
AGCCAGCCAAAGCACG        PBMC        29882   29856.0      5746          2660.44
...                      ...          ...       ...       ...              ...
CATCCCAAGTACTCGT        PBMC          151   151.000       132               12
CTCCTCCCATGAAGCG        PBMC          133   132.500       131                6
AGTTCCCCATGTCAGT        PBMC          173   172.667       166                8
ACGTAGGGTGACAGCA        PBMC          255   253.000       217               13
TCTCAGCTCGCCGAAC        PBMC          144   144.000       149                4
                 subsets_Mito_detected subsets_Mito_percent     total
                             &lt;integer&gt;            &lt;numeric&gt; &lt;numeric&gt;
GCTGCAGTCCGATCTC                    13              8.36307   31908.0
ACTATGGAGGTCCCTG                    15             11.07002   35801.5
ATTTCTGTCTCTATGT                    15              7.86038   31758.3
TATCTGTAGGTGATAT                    15              9.25417   31998.0
AGCCAGCCAAAGCACG                    15              8.91092   29856.0
...                                ...                  ...       ...
CATCCCAAGTACTCGT                     5              7.94702   151.000
CTCCTCCCATGAAGCG                     5              4.52830   132.500
AGTTCCCCATGTCAGT                     6              4.63320   172.667
ACGTAGGGTGACAGCA                     6              5.13834   253.000
TCTCAGCTCGCCGAAC                     2              2.77778   144.000</code></pre>
</div>
</div>
</section>
<section id="filtering-low-quality-cells-3" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<p>We can use the <code>plotColData()</code> function from scater to plot various metrics (as a ggplot2 object).</p>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-36-1.png" width="960" class="r-stretch"></section>
<section id="filtering-low-quality-cells-4" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<p>We can use the <code>plotColData()</code> function from scater to plot various metrics (as a ggplot2 object).</p>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-37-1.png" width="960" class="r-stretch"></section>
<section id="filtering-low-quality-cells-5" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<p>We can use the <code>plotColData()</code> function from scater to plot various metrics (as a ggplot2 object).</p>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-38-1.png" width="960" class="r-stretch"></section>
<section id="filtering-low-quality-cells-6" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<p>We can use the <code>plotColData()</code> function from scater to plot various metrics (as a ggplot2 object).</p>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-39-1.png" width="960" class="r-stretch"></section>
<section id="filtering-low-quality-cells-7" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<p>We can also extract <code>colData</code> as a dataframe to make custom <code>ggplot2</code> plots</p>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-40-1.png" width="960" class="r-stretch"></section>
<section id="filtering-low-quality-cells-8" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<ul>
<li class="fragment">Selecting an appropriate cutoff can be somewhat arbitrary</li>
<li class="fragment">Risk of excluding meaningful cell populations.</li>
<li class="fragment">Start with lenient cutoffs, then later increasing the stringency after examining the clustering and cell types.</li>
</ul>
<p><em>How many cells pass these criteria?</em></p>
<p><em>How many cells pass these criteria?</em></p>
</section>
<section id="filtering-low-quality-cells-9" class="slide level2">
<h2>Filtering low quality cells ⌨️</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>[1] 4565</code></pre>
</div>
</div>
</section>
<section id="visualizing-qc-failed-cells" class="slide level2">
<h2>Visualizing qc failed cells ⌨️</h2>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-43-1.png" width="960" class="r-stretch"></section>
<section id="remove-low-quality-cells" class="slide level2">
<h2>Remove low quality cells ⌨️</h2>
<p>Lastly we can subset the <code>SingleCellExperiment</code> to exclude the low-quality cells.</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>class: SingleCellExperiment 
dim: 20858 4565 
metadata(0):
assays(1): counts
rownames(20858): DDX11L2_ENSG00000290825 WASH7P ... MT-ND6 MT-CYB
rowData names(5): total_gene_counts n_cells_expr gene gene_id n_cells
colnames(4565): GCTGCAGTCCGATCTC ACTATGGAGGTCCCTG ... CTGGATCCACCGTACG
  TACAACGGTCTCGCGC
colData names(9): cell_source total_counts ... total pass_qc
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):</code></pre>
</div>
</div>
</section>
<section id="analysis-workflow" class="slide level2">
<h2>Analysis workflow</h2>
<div class="cell">
<div class="cell-output-display">
<div class="grViz html-widget html-fill-item" id="htmlwidget-6de7588810d36595ae33" style="width:100%;height:100%;"></div>
<script type="application/json" data-for="htmlwidget-6de7588810d36595ae33">{"x":{"diagram":"\ndigraph workflow {\n  graph [layout = dot,\n         rankdir = LR]\n\n  node [shape = cicle,\n        style = filled,\n        fontcolor = black,\n        fontname = \"Helvetica\"]\n\n  # green\n  node [fillcolor = \"#009E73\"]\n  load [label= \"Import data\ntximport::tximport()\nSingleCellExperiment()\ncounts()\"]\n\n  # blue\n  node [fillcolor = \"#56B4E9\"]\n  cell_qc [label = \"QC cells\n addPerCellQCMetrics()\n plotColData()\"]\n  norm [label = \"Normalize UMI counts\nquickCluster()\n computeSumFactors()\n logNormCounts()\"]\n\n  # yellow\n  node [fillcolor = \"#F0E442\"]\n  feature [label = \"Identify variable genes\nmodelGeneVarByPoisson()\n getTopHVGs()\"]\n  dim_red [label = \"Dimensionality reduction via PCA \n runPCA()\"]\n  cluster [label = \"Clustering\n clusterCells()\"]\n  viz [label = \"Make 2D-Visualization\nrunUMAP()\"]\n\n  # blue\n  node [fillcolor = \"#56B4E9\"]\n\n  markers [label = \"Discover cell type markers \nscoreMarkers()\"]\n  annot [label = \"Annotate cell types\nclustifyr and SingleR\"]\n\n  edge [color = black\n        fontname = \"Helvetica\"]\n\n  load -> cell_qc\n  cell_qc -> norm\n  norm -> feature\n  norm -> markers\n  feature -> dim_red\n  dim_red -> cluster\n  dim_red -> viz\n  cluster -> markers\n  markers -> annot\n\n  edge [color = \"grey\"\n        style = \"dashed\"]\n  annot -> cell_qc [label = \"Repeat\n as needed\"]\n  annot -> feature\n  annot -> dim_red\n  annot -> cluster\n}","config":{"engine":"dot","options":null}},"evals":[],"jsHooks":[]}</script>
</div>
</div>
</section>
<section id="normalization" class="slide level2">
<h2>Normalization ⌨️</h2>
<ul>
<li class="fragment">Normalization attempts to correct for technical biases that will distort biological signal in the data.</li>
<li class="fragment">A large source of variation arises due to differences in sequencing depth between cells.</li>
<li class="fragment">This can be seen by performing PCA on the unnormalized counts.</li>
<li class="fragment">We will use <code>runPCA</code> from scater to perform PCA.</li>
</ul>
</section>
<section id="normalization-1" class="slide level2">
<h2>Normalization ⌨️</h2>
<ul>
<li class="fragment">We can now visualize this PCA</li>
</ul>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-47-1.png" width="960" class="r-stretch"><p>Note that PC1 is correlated with the total UMI counts (<code>sum</code>), meaning that the largest source of variation is related to differences in sequencing depth rather than biological differences between cells.</p>
</section></section>
<section>
<section id="normalization-2" class="title-slide slide level1 center">
<h1>Normalization ⌨️</h1>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-48-1.png" width="960" class="r-stretch"></section>
<section id="normalization-3" class="slide level2">
<h2>Normalization ⌨️</h2>
<ul>
<li class="fragment">Normalize with <code>scran</code></li>
</ul>
<ol type="1">
<li class="fragment">Crude clustering to group related cells</li>
<li class="fragment">Identifying a cell-specific normalization factor (size factor)</li>
<li class="fragment">Scaling the counts by this factor</li>
<li class="fragment">log transforming the data (base 2 with a pseudocount).</li>
</ol>
<div class="footer">
<p>Description of algorithim: Lun AT, Bach K, Marioni JC. Pooling across cells to normalize single-cell RNA sequencing data with many zero counts. Genome Biol. 2016 Apr 27;17:75. doi: 10.1186/s13059-016-0947-7. PMID: 27122128; PMCID: PMC4848819.</p>
</div>
</section>
<section id="normalization-4" class="slide level2">
<h2>Normalization ⌨️</h2>
<p>Plot the PCA with normalization</p>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-51-1.png" width="960" class="r-stretch"></section>
<section id="normalization-5" class="slide level2">
<h2>Normalization ⌨️</h2>

<img data-src="slides-31_files/figure-revealjs/unnamed-chunk-52-1.png" width="960" class="r-stretch"><p>PC1 no longer correlates with UMI counts.</p>
</section>
<section id="further-reading" class="slide level2">
<h2>Further Reading</h2>
<p>eBook: Orchestrating single cell analyses with Bioconductor <a href="https://osca.bioconductor.org/">eBook</a> <a href="https://doi.org/10.1038/s41592-019-0654-x">Publication</a></p>
<p>Review: <a href="https://www.embopress.org/doi/full/10.15252/msb.20188746">Current best practices for analysis</a></p>
<p>Blog: <a href="https://ltla.github.io/SingleCellThoughts/">Single Cell Thoughts</a></p>
<p>Blog: <a href="http://www.nxn.se/">What do you mean “heterogeneity”</a></p>

<div class="quarto-auto-generated-content">
<div class="footer footer-default">
<p>Course website: <a href="https://rnabioco.github.io/molb-7950" class="uri">https://rnabioco.github.io/molb-7950</a></p>
</div>
</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="../site_libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="../site_libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="../site_libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-chalkboard/plugin.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-drop/drop-runtime.js"></script>
  <script src="../site_libs/revealjs/plugin/reveal-pointer/pointer.js"></script>
  <script src="../site_libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="../site_libs/revealjs/plugin/notes/notes.js"></script>
  <script src="../site_libs/revealjs/plugin/search/search.js"></script>
  <script src="../site_libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="../site_libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.togglePdfExport(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleChalkboard(event)\"><kbd>b</kbd> Toggle Chalkboard</a></li>\n<li class=\"slide-tool-item\" data-item=\"5\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.toggleNotesCanvas(event)\"><kbd>c</kbd> Toggle Notes Canvas</a></li>\n<li class=\"slide-tool-item\" data-item=\"6\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.downloadDrawings(event)\"><kbd>d</kbd> Download Drawings</a></li>\n<li class=\"slide-tool-item\" data-item=\"7\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'chalkboard': {"buttons":true},
'drop': {"button":true,"shortcut":"`","engine":"webr","webr":{"packages":[]},"pyodide":{"packages":[]}},
'pointer': {"key":"q","color":"red","pointerSize":16,"alwaysVisible":false},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, RevealChalkboard, RevealDrop, RevealPointer, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        // dispatch for htmlwidgets
        function fireSlideEnter() {
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);
        }

        function fireSlideChanged(previousSlide, currentSlide) {
          fireSlideEnter();

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const isCodeAnnotation = (el) => {
        for (const clz of el.classList) {
          if (clz.startsWith('code-annotation-')) {                     
            return true;
          }
        }
        return false;
      }
      const onCopySuccess = function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        let tooltip;
        if (window.bootstrap) {
          button.setAttribute("data-bs-toggle", "tooltip");
          button.setAttribute("data-bs-placement", "left");
          button.setAttribute("data-bs-title", "Copied!");
          tooltip = new bootstrap.Tooltip(button, 
            { trigger: "manual", 
              customClass: "code-copy-button-tooltip",
              offset: [0, -8]});
          tooltip.show();    
        }
        setTimeout(function() {
          if (tooltip) {
            tooltip.hide();
            button.removeAttribute("data-bs-title");
            button.removeAttribute("data-bs-toggle");
            button.removeAttribute("data-bs-placement");
          }
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      }
      const getTextToCopy = function(trigger) {
          const codeEl = trigger.previousElementSibling.cloneNode(true);
          for (const childEl of codeEl.children) {
            if (isCodeAnnotation(childEl)) {
              childEl.remove();
            }
          }
          return codeEl.innerText;
      }
      const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
        text: getTextToCopy
      });
      clipboard.on('success', onCopySuccess);
      if (window.document.getElementById('quarto-embedded-source-code-modal')) {
        // For code content inside modals, clipBoardJS needs to be initialized with a container option
        // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
        const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
          text: getTextToCopy,
          container: window.document.getElementById('quarto-embedded-source-code-modal')
        });
        clipboardModal.on('success', onCopySuccess);
      }
        var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
        var mailtoRegex = new RegExp(/^mailto:/);
          var filterRegex = new RegExp("https:\/\/rnabioco\.github\.io\/molb-7950\/");
        var isInternal = (href) => {
            return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
        }
        // Inspect non-navigation links and adorn them if external
     	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
        for (var i=0; i<links.length; i++) {
          const link = links[i];
          if (!isInternal(link.href)) {
            // undo the damage that might have been done by quarto-nav.js in the case of
            // links that we want to consider external
            if (link.dataset.originalHref !== undefined) {
              link.href = link.dataset.originalHref;
            }
          }
        }
      function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
        const config = {
          allowHTML: true,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'light-border',
          placement: 'bottom-start',
        };
        if (contentFn) {
          config.content = contentFn;
        }
        if (onTriggerFn) {
          config.onTrigger = onTriggerFn;
        }
        if (onUntriggerFn) {
          config.onUntrigger = onUntriggerFn;
        }
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          // use id or data attribute instead here
          let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note) {
            return note.innerHTML;
          } else {
            return "";
          }
        });
      }
      const findCites = (el) => {
        const parentEl = el.parentElement;
        if (parentEl) {
          const cites = parentEl.dataset.cites;
          if (cites) {
            return {
              el,
              cites: cites.split(' ')
            };
          } else {
            return findCites(el.parentElement)
          }
        } else {
          return undefined;
        }
      };
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const citeInfo = findCites(ref);
        if (citeInfo) {
          tippyHover(citeInfo.el, function() {
            var popup = window.document.createElement('div');
            citeInfo.cites.forEach(function(cite) {
              var citeDiv = window.document.createElement('div');
              citeDiv.classList.add('hanging-indent');
              citeDiv.classList.add('csl-entry');
              var biblioDiv = window.document.getElementById('ref-' + cite);
              if (biblioDiv) {
                citeDiv.innerHTML = biblioDiv.innerHTML;
              }
              popup.appendChild(citeDiv);
            });
            return popup.innerHTML;
          });
        }
      }
    });
    </script>
    

</body></html>