---
title: "Where do proteins bind in the genome?"
author: "Jay Hesselberth"
---

## What to map and how to map it?

:::: {.columns}
::: {.column width="50%"}

**Targets**

- Transcription factors
- Histone modifications
- Chromatin remodelers
- RNA polymerases
- Other factors that bind chromatin

:::
::: {.column width="50%"}

**Methods**

- ChIP-seq
- MNase-ChIP-seq
- CUT&RUN
- CUT&TAG

:::
::::

## 

![](../img/block-dna/ChIPseq_2.png){fig-align="center"}

## 

![](../img/block-dna/ChIP_Data.png){fig-align="center"}

## 

![](../img/block-dna/cut-and-run.png){fig-align="center"}

## 

![](../img/block-dna/chip-resolution-comparison.png) {fig-align="center"}

## Comparison of factor-centric methods {.smaller}

:::: {.columns}
::: {.column width="50%"}
![](../img/block-dna/chip-comparison-overview.png)
:::
::: {.column width="50%"}
+----------------+----------------+---------------------+
| Method         | Resolution     | Sequencing cost     |
+:==============:+:==============:+:===================:+
| ChIP-seq       | Low            | High                |
+----------------+----------------+---------------------+
| MNase-ChIP-seq | High           | High                |
+----------------+----------------+---------------------+
| CUT&RUN        | High           | Low                 |
+----------------+----------------+---------------------+

:::
::::

## Workflow

## Example: CTCF CUT&RUN in K562 cells

