---
title: "Resources for the DNA block"
---

## Foundational work

These studies below laid the methodological groundwork for high throughput chromatin characterization by DNA sequencing. We're not going to use them directly in class, but if you are going to do similar experiments, they're worth a read. Some of the authors are here at CU Anschutz!

::: {.callout-tip}
Data sets from these studies would be excellent choices as starting points for your final projects.
:::

#### DNase-seq

Hesselberth JR, Chen X, Zhang Z, Sabo PJ, Sandstrom R, Reynolds AP, Thurman RE, Neph S, Kuehn MS, Noble WS, Fields S, Stamatoyannopoulos JA. Global mapping of protein-DNA interactions in vivo by digital genomic footprinting. Nat Methods. 2009 19305407; PMCID: PMC2668528. [\[Link\]](https://www.ncbi.nlm.nih.gov/pmc/articles/pmid/19305407/)

#### ATAC-seq

Buenrostro JD, Giresi PG, Zaba LC, Chang HY, Greenleaf WJ. Transposition of native chromatin for fast and sensitive epigenomic profiling of open chromatin, DNA-binding proteins and nucleosome position. Nat Methods. 2013 PMID: 24097267; PMCID: PMC3959825. [[Link]](https://pubmed.ncbi.nlm.nih.gov/24097267/)

#### MNase-seq

Ramachandran S, Ahmad K, Henikoff S. Transcription and Remodeling Produce Asymmetrically Unwrapped Nucleosomal Intermediates. Mol Cell. 2017 PMID: 29225036; PMCID: PMC6421108. [\[Link\]](https://linkinghub.elsevier.com/retrieve/pii/S1097-2765(17)30874-2)

#### CUT&RUN / CUT&TAG

Skene PJ, Henikoff S. An efficient targeted nuclease strategy for high-resolution mapping of DNA binding sites. Elife. 2017 PMID: 28079019; PMCID: PMC5310842. [\[Link\]](https://pubmed.ncbi.nlm.nih.gov/28079019/)

Kaya-Okur HS, Wu SJ, Codomo CA, Pledger ES, Bryson TD, Henikoff JG, Ahmad K, Henikoff S. CUT&Tag for efficient epigenomic profiling of small samples and single cells. Nat Commun. 2019 PMID: 31036827; PMCID: PMC6488672. [[Link]](https://pubmed.ncbi.nlm.nih.gov/31036827/)

## Software we'll use in class

- Read over the [GViz vignette](https://bioconductor.org/packages/devel/bioc/vignettes/Gviz/inst/doc/Gviz.html) to understand how we'll use it to vissualize genome-scale data on a reference sequence.

- Read over the [valr vignette](https://rnabioco.github.io/valr/articles/valr.html) to understand how we'll do BEDtools-like (see below) analysis within RStudio.

- Look over the [ComplexHeatmap]() and [EnrichedHeatmap]() documentation, especially XXX. These tools will help us make "meta-plots": figures that plot genomic signals relative to features.

## Other important tools

These are other tools I'll mention in class. We're not going to use them directly, but they are important tools in upstream data processing and analysis.

### Alignment software

[Bowtie2]() and [BWA]() are popular choices for short read alignment. They are fast, free, and well-maintained. 

### Peak calling

[MACS](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2592715/) is the gold-standard in peak calling. It models read coverage as a Poisson process, straightforward identification of regions of higher than expected coverage (i.e., peaks) to be identified using a single parmaeter (lambda) that captures the mean and variance of read coverage.

### Interval analysis

- [BEDtools](https://bedtools.readthedocs.io/en/latest/) is the "Swiss Army knife" of genome interval analysis. It provides a host of command-line tools that can be linked together for powerful genome signal manipulation.
