---
title: "Chromatin accessibility - Class 1"
author: "Jay Hesselberth"
---

# DNA Accessibility

Today we will work with two datas xets in budding yeast that map genome-wide accessibility. We will go back to our old friends:

## Libraries

1. rtracklayer
1. Gviz - this is a way to generate pretty images of genome browser snapshots within Rstudio.

These tools are useful for us to look at the data directly, and are great use to you when you write papers and want to create nice plots. 

```{r}
#| label: load-libs
#| message: false
library(tidyverse)
library(cowplot)
library(here)
library(patchwork)

library(valr)
library(Gviz)
library(rtracklayer)
```

# Patterns of DNA-protein binding

## Load the data 

Today we will analyze ATAC-seq and MNase-seq datasets from budding yeast. Here are the references for the datasets:

1.  ATAC-seq - <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4617971/> (GSM1621323)
2.  MNase-seq - <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3889552/> (GSM1199785)

In case of MNase-seq, usually, DNA around 150 bp is extracted, so as to look only at nucleosome occupancy & positioning. However, the study we are analyzing today did not perform size selection. This is very important as now we can look at both TFs and nucleosomes within the MNase-seq data.

First, we will determine the fragment size distributions obtained from the two experiments. These sizes are the fingerprints of particles that were protecting the DNA inside the cells.

I have performed the alignment of paired-end reads and converted all reads into a bed file where each line of the bed file denotes a single fragment from start to end.

First, the ATAC-seq reads:

```{r}
atac_tbl <- valr::read_bed(
  here("data/block-dna/yeast_atac_chrII.bed.gz"),
) 
atac_tbl
```

Second, the MNase reads.

```{r}
mnase_tbl <- valr::read_bed(
  here("data/block-dna/yeast_mnase_chrII.bed.gz")
)
mnase_tbl
```

You might have guessed by by the file names that we are looking at chromosome II of the yeast genome.

You will see that working with a small genome is a huge advantage -- we can study the whole chromosome in this class. For *Drosophila* and human data we dealt with before, we were looking at a subsection of the chromosome.

For the MNase BED file, you see that there are only three columns: `chrom`, `start`, and `end`. Calculating fragment length is very simple:

```{r}
#| echo: true
mutate(mnase_tbl, frag_len = end - start)
```

## Length distributions of chromatin-derived DNA fragments

Let's examine the fragment length distribution. 

```{r}
#| echo: true
#| warning: false
#| output-location: column
acc_tbl <-
  bind_rows(
    mutate(mnase_tbl, type = "mnase"),
    mutate(atac_tbl, type = "atac")
  )

ggplot(
  acc_tbl,
  aes(x = end - start)
  ) +
  geom_histogram(
    # single base-pair resolution
    binwidth = 1
  ) + 
  facet_grid(
    rows = vars(type),
    scales = "free_y"
    ) +
  xlim(30, 500) +
  labs(x = "fragment length") +
  theme_cowplot()
```

### Interpreations

1.  How would you describe the two distributions? Are they similar?

2.  Can you make any biological conclusions based on the length distributions?

## Periodicity in the fragment lengths

The ATAC data seems to be periodic. How can we test that hypothesis? We can calculate the autocorrelation of the length distribution. Can someone explain what [autocorrelation](https://en.wikipedia.org/wiki/Autocorrelation) means?

We'll use the base `hist` function to calculate the densities of the above histogram. Let's write a function we can use to analyze fragment lengths.

```{r}
#| label: fun-fragment-len
#| echo: true
fragment_len_hist <- function(tbl) {
  frag_lens <- 
    mutate(
      tbl,
      frag_len = end - start
    ) |>
    filter(
      frag_len >= 30 &
        frag_len <= 500
    ) |>
    pull(frag_len)
  
  hist(
    frag_lens,
    breaks = seq(30, 500, 1),
    plot = FALSE
  )
}
```


```{r}
atac_frag_lens <- fragment_len_hist(atac_tbl)
```

The `density` slot contains a vector of densities at base-pair resolution. We will use `acf()` to calculate the autocorrelation of these values, and will store the tidied result. 

```{r}
#| echo: true
atac_acf_tbl <-
  acf(
    atac_hist$density,
    lag.max = 40,
    plot = FALSE
  ) |>
    broom::tidy()

atac_acf_tbl
```

Now let's plot the autocorrelation:

```{r}
plot_len_acf <- function(tbl, title) {
  ggplot(
    tbl,
    aes(lag, acf)
  ) +
    geom_point(size = 2) +
    geom_line() +
    theme_minimal_grid() +
    geom_vline(xintercept = c(10, 21), color = 'red') +
    labs(title = title)
}
```

```{r}
plot_atac_acf <- plot_len_acf(atac_acf_tbl, "ATAC ACF")
plot_atac_acf
```

So, it looks like there significant bumps in autocorrelation at 10 and 21 bp positions, indicating that ATAC length distribution is periodic.

How do we confirm these bumps are interesting? Let's calculate the acf of a negative control -- the length distribution of the MNase data.

```{r}
#| label: plot-acfs
#| echo: true
#| output-location: default
mnase_frag_lens <-
  fragment_len_hist(mnase_tbl)

mnase_acf_tbl <-
  acf(
    mnase_frag_lens$density,
    lag.max = 40,
    plot = FALSE
  ) |>
  broom::tidy()

plot_mnase_acf <- plot_len_acf(mnase_acf_tbl, "MNase ACF")

# patchwork plot
plot_atac_acf / plot_mnase_acf
```

We see a monotonic decrease here, which confirms that the bumps we see are distinctive features of ATAC-seq data.

## Visualize the raw read density in a given region of the genome

We will use Gviz to visualize read densities in any given region.

Gviz can plot annotation tracks and data tracks.

## Load tracks

First, we will load the gene annotations from ENSEMBL. What does the function BiomartGeneRegionTrack do?

```{r}
sgd_genes <-
  UcscTrack(
    genome = "sacCer3",
    track = "sgdGene",
    chromosome = "chrII",
    start = 530811,
    end = 540885,
    trackType = "GeneRegionTrack",
    gene = "name",
    rstarts = "exonStarts",
    rends = "exonEnds",
    symbol = "name",
    transcript = "name",
    strand = "strand"
  )

plotTracks(sgd_genes)
```

We will import the bigwig file containing yeast nucleosomal sized fragments from MNase-seq using import.bw function from rtracklayer. What is "GRanges"?

```{r}
mnase_
mnase_nuc <- import.bw("data_acc/yeast_mnase_134_160.bw", as = "GRanges")
mnase_nuc
```

We will then load the GRanges object as a track for Gviz to plot:

```{r}
mnase_nuc_trk <- DataTrack(mnase_nuc, name = "MNase_nuc")
mnase_nuc_trk
```

Now, we can make a plot for this particular region of chrII:

```{r fig.height = 3, fig.width = 5}
plotTracks(c(yeast_genome, mnase_nuc_trk), from = 530811, to = 540885, chromosome = "chrII", type = "l", transcriptAnnotation = "gene", shape = "arrow", type = "h")
```

That looks great! Let us load all the other datasets we have.

1.  We will first load the bigwigs as a GRanges onject
2.  Then we will convert that into a data track that Gviz can plot

```{r}
mnase_tf <- import.bw("data_acc/yeast_mnase_lt50.bw", as = "GRanges")
atac_short <- import.bw("data_acc/yeast_atac_lt120.bw", as = "GRanges")
atac_long <- import.bw("data_acc/yeast_atac_gt120.bw", as = "GRanges")
mnase_tf_trk <- DataTrack(mnase_tf, name = "MNase_TF")
atac_short_trk <- DataTrack(atac_short, name = "ATAC_Short")
atac_long_trk <- DataTrack(atac_long, name = "ATAC_Long")
```

Now, we just have to make a list of tracks to plot and Gviz takes care of teh rest. Some questions to think about as you look at the tracks:

1.  What is each dataset reporting on?
2.  What can you infer about genome function based on these tracks?
3.  What are the major differences between MNase-seq and ATAC-seq based on these tracks?

```{r fig.height = 3, fig.width = 5}
plotTracks(c(yeast_genome, mnase_nuc_trk, mnase_tf_trk, atac_long_trk, atac_short_trk), from = 530811, to = 540885, chromosome = "chrII", type = "h", transcriptAnnotation = "gene", shape = "arrow")
```

## Metaplots and heatmap

So a small region of chrII is telling us a lot about chromatin structure and the different methods to profile accessibility.

Next, we want to make sure these conclusions are valid throughout the genome. Well since we want to keep the file sizes small, we can ask if it is valid throughout chrII. We will go to our old pal, deeptools to do this analysis.

First we will plot the profiles of all our datasets relative to the transcription start site (TSS), where all the action seems to be happening:

```{bash eval=FALSE}
computeMatrix reference-point --referencePoint center -a 750 -b 750 \
-S data_acc/yeast_atac_gt120.bw  data_acc/yeast_mnase_134_160.bw \
data_acc/yeast_atac_lt120.bw  data_acc/yeast_mnase_lt50.bw \
--regionsFileName data_acc/yeast_tss_chrII.bed \
--outFileName data_acc/tss.mat.gz --missingDataAsZero

plotHeatmap -m data_acc/tss.mat.gz -o data_acc/heatmap_tss.png \
--zMin 0 50 0 0 \
--zMax 500 225 850 800 \
--colorList '#e5f5f9,#2ca25f' '#efedf5,#756bb1' '#e5f5f9,#2ca25f' '#efedf5,#756bb1' \
--whatToShow 'heatmap and colorbar'

plotProfile -m data_acc/tss.mat.gz -o data_acc/profile_tss.png \
--yMin 75 50 50 50 \
--yMax 500 225 850 800 \
--colors '#2ca25f' '#756bb1' '#2ca25f' '#756bb1' \
--plotType=fill
```

```{r echo=FALSE, out.width = '100%'}
knitr::include_graphics("data_acc/profile_tss.png")
```

```{r echo=FALSE, out.width = '100%'}
knitr::include_graphics("data_acc/heatmap_tss.png")
```

1.  How do heatmaps and metaplots work?
2.  What can you conclude from these plots?
3.  What are the disting features of MNase-seq and ATAC-seq that become clearer when looking at many genes at the same time?
4.  What are some hypotheses you can generate based on these plots?

## Perspective matters!

We centered our analysis on TSSs and observed many interesting features. We can change our perspective slightly and see what happens. We can measure the position of the first nucleosome downstream of the TSS highly accurately in yeast (at bp resolution!). I have put the list of +1 positions in a bed file. Let us see how the plots differ when we shift the center to +1 nucleosomes instead of TSSs:

```{bash eval=FALSE}

computeMatrix reference-point --referencePoint center -a 750 -b 750 \
-S data_acc/yeast_atac_gt120.bw  data_acc/yeast_mnase_134_160.bw \
data_acc/yeast_atac_lt120.bw  data_acc/yeast_mnase_lt50.bw \
--regionsFileName data_acc/yeast_p1_chrII.bed \
--outFileName data_acc/p1.mat.gz --missingDataAsZero

plotHeatmap -m data_acc/p1.mat.gz -o data_acc/heatmap_p1.png \
--zMin 0 50 0 0 --zMax 500 225 850 800 \
--colorList '#e5f5f9,#2ca25f' '#efedf5,#756bb1' '#e5f5f9,#2ca25f' '#efedf5,#756bb1' \
--whatToShow 'heatmap and colorbar'

plotProfile -m data_acc/p1.mat.gz -o data_acc/profile_p1.png \
--yMin 75 50 25 50 \
--yMax 500 225 850 800 \
--colors '#2ca25f' '#756bb1' '#2ca25f' '#756bb1' \
--plotType=fill
```

```{r echo=FALSE, out.width = '100%'}
knitr::include_graphics("data_acc/profile_p1.png")
```

```{r echo=FALSE, out.width = '100%'}
knitr::include_graphics("data_acc/heatmap_p1.png")
```

