---
title: "Exercises 8"
author: "Jay Hesselberth"
date: last-modified
format:
  html:
    toc: true
    toc-depth: 2
    code-fold: true
---

# Putting it all together

For the next two classes we'll combine everything we've learned to process and visualize data from some some biological experiments. These exercises will illustrate a complete analysis pipeline -- from data tidying to manipulation and visualization -- using tools from the tidyverse.

![](../img/tidy-workflow.png)

### Libraries

Load the libraries you need for analysis below.

```{r}

```

# A quantitative PCR experiment

Here is the experimental setup:

-   Two cell lines (`wt` and `mut`) were treated with a drug that induces interferon expression

-   After specific time points, cells were harvested and actin and interferon mRNA were analyzed by quantitative PCR (with 3 technical replicates), with a control containing no reverse transcriptase.

### Load the data

These data are in two TSV files:

-   `data/qpcr_names.tsv.gz`
-   `data/qpcr_data.tsv.gz`

Load these data sets and inspect.

```{r}

```

Note the shape of the data and the names of the rows and columns. Do they remind you of anything?

![](../img/96-well-pcr-plate.png)

### Tidy the data

Given the experimental setup and the shape of the tibbles, you should be able to answer: **Are these data tidy?**

-   What are the variables in the data?
-   Are the variables the column names?

```{r}

```

### Merge the data

Note the structure of the tidied data. What columns (variables) are shared by both tibbles?

How we can *join* the data from these two tibbles, linking the sample identifiers with their gene expression values?

```{r}

```

### Summarize the data

Calculate the mean and standard deviation across replicates.

Do this two ways:

1.  Calculate the statistics for each gene separately.

2.  Calculate a ratio of interferon to actin levels for each sample *before* calculating the mean and standard deviation of the ratios.

```{r}

```

### Plot the data

Now we can plot the summary statistics. We'll use [`ggplot2::geom_pointrange()`](https://ggplot2.tidyverse.org/reference/geom_linerange.html) to represent the mean and standard deviation.

```{r}

```

### Interpret the plot

-   What can you say about the expression of ACTIN and IFN?

-   What can you say about the mutant and wild-type cells?


# A yeast gene expression experiment

Next, we'll examine some gene expression data from the budding yeast *S. cerevisiae*. We'll roughly follow the analysis strategy taken by David Robinson in his blog [*Variance Explained*](http://varianceexplained.org/r/tidy-genomics-broom/).

The [data](http://growthrate.princeton.edu/transcriptome/download.shtml) come from:

> Brauer MJ, Huttenhower C, Airoldi EM, Rosenstein R, Matese JC, Gresham D, Boer VM, Troyanskaya OG, Botstein D. **Coordination of growth rate, cell cycle, stress response, and metabolic activity in yeast.** *Mol Biol Cell.* 2008 [**\[Link**\]](https://pubmed.ncbi.nlm.nih.gov/17959824/)

They used [chemostats](https://dunham.gs.washington.edu/chemostatv2.htm) to control the growth rate of cells under different nutrient-limited conditions.

In this experiment, cells were grown in different media limited for (one of) glucose, uracil, leucine, sulfate, phosphate, or ammonia. Over a series of fixed, equilibrium growth rates (established by the dilution rate of fresh media), cells were harvested and gene expression was measured by genome-wide microarrays.

### Load the data

A raw version of the gene expression data are in:

-   `data/brauer_gene_exp_wide.tsv.gz`

In addition, two tibbles contain related information:

-   `data/yeast_go_terms.tsv.gz`
-   `data/yeast_prot_prop.tsv.gz`

Load each of the above files and inspect.

```{r}

```

### Tidy the data

**Are these data tidy?**

```{r}

```

### Heatmap of gene expression values

Heatmaps are a useful approach to visualize thousands of data points, orgnaized by experimental variables to show patterns in the data.

We'll use the [ComplexHeatmap](https://jokergoo.github.io/ComplexHeatmap-reference/book/) package from Bioconductor, which provides a flexible framework for generating heatmaps.

```{r}

```

### Expression of select genes associated with nutrient metabolism

Examine genes (`common_name`) that start with LEU, URA, PHO, and SUL.

```{r}

```

## Homework

Your homework tonight is to review everything we did today. You won't have anything due tomorrow. There will be a more substantial problem set due on Friday at 5pm.

