<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">
<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">
<meta name="author" content="Neelanjan Mukherjee">
<meta name="dcterms.date" content="2024-10-09">
<title>RBP-RNA part 1 – MOLB 7950</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>

<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark"><script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<meta property="og:title" content="RBP-RNA part 1 – MOLB 7950">
<meta property="og:description" content="Informatics and Statistics for Molecular Biology, Fall 2023.">
<meta property="og:site_name" content="MOLB 7950">
<meta name="twitter:title" content="RBP-RNA part 1 – MOLB 7950">
<meta name="twitter:description" content="Informatics and Statistics for Molecular Biology, Fall 2023.">
<meta name="twitter:image" content="https://rnabioco.github.io/molb-7950/exercises/images/twitter-card.png">
<meta name="twitter:creator" content="@rnabioco">
<meta name="twitter:card" content="summary_large_image">
</head>
<body class="nav-sidebar docked">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top"><nav class="quarto-secondary-nav"><div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item">RBP-RNA part 1</li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav></header><!-- content --><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation docked overflow-auto"><div class="pt-lg-2 mt-2 text-center sidebar-header">
      <a href="../index.html" class="sidebar-logo-link">
      <img src="../img/logo.png" alt="" class="sidebar-logo py-0 d-lg-inline d-none"></a>
      <div class="sidebar-tools-main tools-wide">
    <a href="https://github.com/rnabioco/molb-7950" title="Course GitHub page" class="quarto-navigation-tool px-1" aria-label="Course GitHub page"><i class="bi bi-github"></i></a>
    <a href="https://sso.posit.cloud/cu-anschutz" title="Posit Cloud SSO login" class="quarto-navigation-tool px-1" aria-label="Posit Cloud SSO login"><i class="bi bi-cloud"></i></a>
    <a href="https://calendly.com/molb7950" title="Office hours scheduling" class="quarto-navigation-tool px-1" aria-label="Office hours scheduling"><i class="bi bi-calendar"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
<li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Course information</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Schedule</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course-info/syllabus.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Syllabus</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course-info/support.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Support</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course-info/team.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Teaching team</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Resources</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../resources/bootcamp-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bootcamp resources</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../resources/block-dna-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">DNA Block resources</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../resources/block-rna-resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">RNA Block resources</span></a>
  </div>
</li>
      </ul>
</li>
        <li class="px-0"><hr class="sidebar-divider hi "></li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Assignments</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">
<li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course-info/problem-sets.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Problet Set Overview</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../course-info/final-projects.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Final Project Overview</span></a>
  </div>
</li>
      </ul>
</li>
    </ul>
</div>
</nav><div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active"><h2 id="toc-title">On this page</h2>
   
  <ul>
<li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives</a></li>
  <li><a href="#background" id="toc-background" class="nav-link" data-scroll-target="#background">Background</a></li>
  <li><a href="#why-do-we-need-to-know-binding-sites" id="toc-why-do-we-need-to-know-binding-sites" class="nav-link" data-scroll-target="#why-do-we-need-to-know-binding-sites">Why do we need to know binding sites?</a></li>
  <li><a href="#mapping-of-rbp-binding-sites-clip-seq" id="toc-mapping-of-rbp-binding-sites-clip-seq" class="nav-link" data-scroll-target="#mapping-of-rbp-binding-sites-clip-seq">Mapping of RBP binding sites: CLIP-seq</a></li>
  <li><a href="#mapping-of-rbp-binding-sites-par-clip" id="toc-mapping-of-rbp-binding-sites-par-clip" class="nav-link" data-scroll-target="#mapping-of-rbp-binding-sites-par-clip">Mapping of RBP binding sites: PAR-CLIP</a></li>
  <li><a href="#mapping-of-rbp-binding-sites-analysis" id="toc-mapping-of-rbp-binding-sites-analysis" class="nav-link" data-scroll-target="#mapping-of-rbp-binding-sites-analysis">Mapping of RBP binding sites: analysis</a></li>
  <li><a href="#analysis-overview" id="toc-analysis-overview" class="nav-link" data-scroll-target="#analysis-overview">Analysis overview</a></li>
  <li><a href="#pre-processing" id="toc-pre-processing" class="nav-link" data-scroll-target="#pre-processing">Pre-processing</a></li>
  <li><a href="#calling-binding-sites-paralyzer" id="toc-calling-binding-sites-paralyzer" class="nav-link" data-scroll-target="#calling-binding-sites-paralyzer">Calling binding sites: PARalyzer</a></li>
  <li><a href="#todays-menu" id="toc-todays-menu" class="nav-link" data-scroll-target="#todays-menu">Today’s menu</a></li>
  <li><a href="#annotation-of-binding-sites" id="toc-annotation-of-binding-sites" class="nav-link" data-scroll-target="#annotation-of-binding-sites">Annotation of binding sites</a></li>
  <li><a href="#set-up-annotation-database" id="toc-set-up-annotation-database" class="nav-link" data-scroll-target="#set-up-annotation-database">Set up annotation database</a></li>
  <li><a href="#extract-annotation-categories" id="toc-extract-annotation-categories" class="nav-link" data-scroll-target="#extract-annotation-categories">Extract annotation categories</a></li>
  <li><a href="#compare-introns-and-cds-length" id="toc-compare-introns-and-cds-length" class="nav-link" data-scroll-target="#compare-introns-and-cds-length">Compare introns and cds length</a></li>
  <li><a href="#elavl1hur" id="toc-elavl1hur" class="nav-link" data-scroll-target="#elavl1hur">ELAVL1/HuR</a></li>
  <li><a href="#par-clip-data" id="toc-par-clip-data" class="nav-link" data-scroll-target="#par-clip-data">PAR-CLIP data</a></li>
  <li><a href="#annotate-par-clip-data" id="toc-annotate-par-clip-data" class="nav-link" data-scroll-target="#annotate-par-clip-data">Annotate PAR-CLIP data</a></li>
  <li><a href="#hur-binding-region-preference" id="toc-hur-binding-region-preference" class="nav-link" data-scroll-target="#hur-binding-region-preference">HuR binding region preference</a></li>
  <li><a href="#binding-region-length-biases" id="toc-binding-region-length-biases" class="nav-link" data-scroll-target="#binding-region-length-biases">binding region length biases</a></li>
  <li><a href="#control-for-clip-binding-sites" id="toc-control-for-clip-binding-sites" class="nav-link" data-scroll-target="#control-for-clip-binding-sites">Control for CLIP-binding sites</a></li>
  <li><a href="#compare-hur-vs-random" id="toc-compare-hur-vs-random" class="nav-link" data-scroll-target="#compare-hur-vs-random">Compare HuR vs Random</a></li>
  <li><a href="#minute-break" id="toc-minute-break" class="nav-link" data-scroll-target="#minute-break">5 MINUTE BREAK</a></li>
  <li><a href="#what-sequence-does-hur-bind-to" id="toc-what-sequence-does-hur-bind-to" class="nav-link" data-scroll-target="#what-sequence-does-hur-bind-to">What sequence does HuR bind to?</a></li>
  <li><a href="#counting-k-mers" id="toc-counting-k-mers" class="nav-link" data-scroll-target="#counting-k-mers">Counting k-mers</a></li>
  <li><a href="#calculate-6mers-in-hur-sites" id="toc-calculate-6mers-in-hur-sites" class="nav-link" data-scroll-target="#calculate-6mers-in-hur-sites">Calculate 6mers in HuR sites</a></li>
  <li><a href="#calculate-6mers-in-3utrs" id="toc-calculate-6mers-in-3utrs" class="nav-link" data-scroll-target="#calculate-6mers-in-3utrs">Calculate 6mers in 3utrs</a></li>
  <li><a href="#sequences-enriched-in-hur-sites-vs-3utr" id="toc-sequences-enriched-in-hur-sites-vs-3utr" class="nav-link" data-scroll-target="#sequences-enriched-in-hur-sites-vs-3utr">Sequences enriched in hur sites vs 3utr</a></li>
  <li><a href="#sequences-enriched-in-hur-sites-vs-3utr-1" id="toc-sequences-enriched-in-hur-sites-vs-3utr-1" class="nav-link" data-scroll-target="#sequences-enriched-in-hur-sites-vs-3utr-1">Sequences enriched in hur sites vs 3utr</a></li>
  </ul><div class="toc-actions"><ul><li><a href="https://github.com/rnabioco/molb-7950/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/rnabioco/molb-7950/edit/main/exercises/ex-27.qmd" class="toc-action"><i class="bi empty"></i>Edit this page</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content"><header id="title-block-header" class="quarto-title-block default"><div class="quarto-title">
<h1 class="title">RBP-RNA part 1</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Neelanjan Mukherjee </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">October 9, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header><section id="learning-objectives" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="learning-objectives">Learning Objectives</h2>
<ul>
<li><p>Describe the steps involved in performing CLIP-seq experiment to identify RBP binding sites transcriptome-wide.</p></li>
<li><p>Understand how CLIP-seq data is processed from reads to generate binding sites.</p></li>
<li><p>Assign the binding sites to genes and transcipt regions - 3’ UTR, CDS, intron, etc.</p></li>
<li><p>Determine the sequences bound by the RBP using proper background sequence models.</p></li>
</ul></section><section id="background" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="background">Background</h2>
<p><img src="../img/block-rna/RNA_life_cycle.png" class="img-fluid"></p>
<p>RNA-binding proteins (RBPs) interact with RNAs to control every regulatory step from transcription to translation.</p>
<p>RBPs bind to specific regulatory elements with RNAs. These regulatory elements are specific sequence and/or structures recognized by the RBPs.</p>
<p>For example, an RBP might bind to specific sequences in the 3’ UTR of an mRNA and recruit enzymes to promote deadenylation (removal of polyA tails) and subsequent target RNA decay.</p>
</section><section id="why-do-we-need-to-know-binding-sites" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="why-do-we-need-to-know-binding-sites">Why do we need to know binding sites?</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure"><p><img src="../img/block-rna/motivation.jpg" class="img-fluid figure-img"></p>
<figcaption>Genome browser shot of Ago1 CLIP-seq data for the Drosophila gene mbt 3’ UTR.</figcaption></figure>
</div>
<ul>
<li>RBPs and microRNAs recognize short sequences: 4-7 nucleotides.</li>
<li>These sequences appear quite frequently (purple) in a 3’ UTR.</li>
<li>Conservation is often used to focus sites more likely to be functional/important.</li>
<li>However, the binding evidence allows one to know the functional regulatory element.</li>
</ul></section><section id="mapping-of-rbp-binding-sites-clip-seq" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="mapping-of-rbp-binding-sites-clip-seq">Mapping of RBP binding sites: CLIP-seq</h2>
<div class="columns">
<div class="column" style="width:50%;">
<ul>
<li>Covalent cross-linking of RBPs to RNA using 254 nm UV light.</li>
<li>Lyse cells</li>
<li>Mild digestion with RNase (leaving protein protected RNA fragments )</li>
<li>Immunoprecipitate your RBP of interest (better have a good antibody)</li>
<li>Ligate labeled 3’ adapter RNA</li>
<li>Cut out region corresponding to crosslinked RBP (and matched input)</li>
<li>Proteinase K treat and isolate RNA</li>
<li>Reverse transcription</li>
<li>Make library from cDNA fragments and SEQUENCE!</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img src="../img/block-rna/clipseq.jpg" class="img-fluid"></p>
</div>
</div>
</section><section id="mapping-of-rbp-binding-sites-par-clip" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="mapping-of-rbp-binding-sites-par-clip">Mapping of RBP binding sites: PAR-CLIP</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>PAR-CLIP</strong>: photoactivatable ribonucleoside enhanced clip</p>
<ul>
<li>4-thiouridine (4sU) incorporation increases cross-linking efficiency</li>
<li>Cross-linking ~310 nm UV (blacklight)</li>
<li>T -&gt; C transitions provide nucleotide resolution binding</li>
<li>During cDNA generation, preferential base pairing of the 4SU crosslink product to a guanine instead of an adenine results in a thymine (T) to cytosine (C) transition in the PCR-amplified sequence, serving as a diagnostic mutation at the site of contact.</li>
</ul>
</div><div class="column" style="width:50%;">
<p><img src="../img/block-rna/nihms832576f1.jpg" class="img-fluid" width="389"></p>
</div>
</div>
</section><section id="mapping-of-rbp-binding-sites-analysis" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="mapping-of-rbp-binding-sites-analysis">Mapping of RBP binding sites: analysis</h2>
<div class="columns">
<div class="column nonincremental" style="width:50%;">
<p>Most CLIP-seq approaches have single-nucleotide resolution information. However, they vary in the frequency of that information and the efficiency of the procedure.</p>
<p>The basic concept to <strong>call a peak/binding sites</strong> from CLIP-seq:</p>
<ul>
<li>Map the reads to genome/transcriptome</li>
<li>Group overlapping reads</li>
<li>Use nucleotide level information to de/refine position of RBP-binding sites</li>
</ul>
<p>In this class we will be working with PAR-CLIP data. Regardless, I will show you how to access ENCODE eCLIP data. You would easily be able to apply what you learn on those data.</p>
</div><div class="column" style="width:50%;">
<p><img src="../img/block-rna/reads_cliptype.jpg" class="img-fluid"></p>
</div>
</div>
</section><section id="analysis-overview" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="analysis-overview">Analysis overview</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><img src="../img/block-rna/workflow.jpg" class="img-fluid"></p>
</div>
<div class="column nonincremental" style="width:50%;">
<ol type="1">
<li>Filter out low quality or short reads (&lt;18 for larger genomes)<br>
</li>
<li>Trim adapters<br>
</li>
<li>Collapse PCR duplicate reads (molecular indexes)<br>
</li>
<li>Align to genome/transcriptome<br>
</li>
<li>Call peaks</li>
<li>Downstream analysis</li>
</ol>
</div>
</div>
</section><section id="pre-processing" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="pre-processing">Pre-processing</h2>
<p><a href="https://cutadapt.readthedocs.io/en/stable/">Cutadapt</a></p>
<p><img src="../img/block-rna/adapters.jpg" class="img-fluid"></p>
</section><section id="calling-binding-sites-paralyzer" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="calling-binding-sites-paralyzer">Calling binding sites: PARalyzer</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p>The pattern of T = &gt; C conversions, coupled with read density, can thus provide a strong signal to generate a high-resolution map of confident RNA-protein interaction sites.</p>
<p>A non-parametric kernel-density estimate used to identify the RNA-protein interaction sites from a combination of T = &gt; C conversions and read density.</p>
<p>See <a href="https://pubmed.ncbi.nlm.nih.gov/21851591/">PARalyzer</a> for more information.</p>
</div><div class="column" style="width:50%;">
<p><img src="../img/block-rna/FMRreads.png" class="img-fluid"></p>
</div>
</div>
</section><section id="todays-menu" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="todays-menu">Today’s menu</h2>
<p>We will be starting with position of the binding sites in the genome (the output of PARalyzer) and we will:</p>
<section id="a.-annotated-the-genes-and-regions-of-the-binding-sites." class="level4"><h4 class="anchored" data-anchor-id="a.-annotated-the-genes-and-regions-of-the-binding-sites.">1a. Annotated the genes and regions of the binding sites.</h4>
</section><section id="b.-calculate-a-summary-table-for-binding-sites-per-gene-per-region." class="level4"><h4 class="anchored" data-anchor-id="b.-calculate-a-summary-table-for-binding-sites-per-gene-per-region.">1b. Calculate a summary table for binding sites per gene per region.</h4>
</section><section id="perform-motif-analysis-accounting-for-the-background-sequence-regions." class="level4"><h4 class="anchored" data-anchor-id="perform-motif-analysis-accounting-for-the-background-sequence-regions.">2. Perform motif analysis accounting for the background sequence regions.</h4>
<p><img src="../img/block-rna/workflow_2hallf.jpg" class="img-fluid"></p>
</section></section><section id="annotation-of-binding-sites" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="annotation-of-binding-sites">Annotation of binding sites</h2>
<p>Where are the binding sites?<br>
- Which genes?<br>
- What region of those genes?<br>
- How many binding sites per region?<br>
- How many binding sites per gene?<br>
- How many binding sites per gene by region?</p>
<p>We will use <code>annotatr</code> and <code>Granges</code> to answer these questions.</p>
<p><img src="../img/block-rna/annotation.jpeg" class="img-fluid"></p>
</section><section id="set-up-annotation-database" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="set-up-annotation-database">Set up annotation database</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>possible_annotations <span class="ot">&lt;-</span> <span class="fu">builtin_annotations</span>()</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># grep to keep those containing "hg19"</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>hg19_annots <span class="ot">&lt;-</span> <span class="fu">grep</span>(</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="st">"??_genes"</span>,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  possible_annotations,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">value =</span> T)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="co"># WHY DID WE PICK hg19?</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># let's keep 5' utr, cds, intron, 3' utr and intergenic</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>my_hg19_annots <span class="ot">&lt;-</span> hg19_annots[<span class="fu">c</span>(<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">7</span>,<span class="dv">10</span>,<span class="dv">11</span>)]</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co"># build the annotation database</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>annotations <span class="ot">&lt;-</span> <span class="fu">build_annotations</span>(</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">genome =</span> <span class="st">'hg19'</span>,</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">annotations =</span> ??)</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>annotations</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="extract-annotation-categories" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="extract-annotation-categories">Extract annotation categories</h2>
<p>What information is contained within the <code>annotations</code> object?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>my_hg19_annots[<span class="dv">3</span>]</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="co"># get introns</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>annotation_introns <span class="ot">&lt;-</span> annotations[annotations<span class="sc">$</span>type <span class="sc">==</span> ??]</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># collapsing introns</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>annotation_introns <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>(annotation_introns)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a><span class="co"># length of each intron interval</span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>length_introns <span class="ot">&lt;-</span> <span class="fu">width</span>(??)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co">#get cds</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>annotation_cds <span class="ot">&lt;-</span> annotations[annotations<span class="sc">$</span>type <span class="sc">==</span> my_hg19_annots[<span class="dv">1</span>]]</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="co"># collapsing cds</span></span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>annotation_cds <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>(annotation_cds)</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="co"># length of each cds interval</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>length_cds <span class="ot">&lt;-</span> <span class="fu">width</span>(annotation_cds)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="compare-introns-and-cds-length" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="compare-introns-and-cds-length">Compare introns and cds length</h2>
<div class="cell" data-output-location="column-fragment">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>all_length <span class="ot">&lt;-</span> <span class="fu">bind_rows</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">nt=</span>??,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>       <span class="at">cat=</span><span class="fu">rep</span>(<span class="st">"??"</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>               <span class="fu">length</span>(??))),</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">tibble</span>(<span class="at">nt=</span>length_cds,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">cat=</span><span class="fu">rep</span>(<span class="st">"cds"</span>,</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>               <span class="fu">length</span>(length_cds)))</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(all_length,</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">x=</span>??, <span class="at">color=</span>??)) <span class="sc">+</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_density</span>() <span class="sc">+</span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_log10</span>(<span class="at">breaks =</span> <span class="fu">trans_breaks</span>(<span class="st">"log10"</span>,</span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>                                      <span class="cf">function</span>(x) <span class="dv">10</span><span class="sc">^</span>x),</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>                <span class="at">labels =</span> <span class="fu">trans_format</span>(<span class="st">"log10"</span>,</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>                                      <span class="fu">math_format</span>(<span class="dv">10</span><span class="sc">^</span>.x))) <span class="sc">+</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_cowplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The typical* human intron is way longer than a CDS exon.</p>
</section><section id="elavl1hur" class="level2 smaller nonincremental"><h2 class="smaller nonincremental anchored" data-anchor-id="elavl1hur">ELAVL1/HuR</h2>
<p>ELAVL1/HuR is an essential RNA-binding protein (RBP).</p>
<p>HuR binds to AU-rich elements (ARE) in 3’ UTRs of mRNAs to promote mRNA stability.</p>
<p><img src="../img/block-rna/hur_mechanism.png" class="img-fluid"></p>
<p>This model makes a few specific prediction:</p>
<ol type="1">
<li>HuR binds to the 3’ UTR.<br>
</li>
<li>HuR binds to AU-rich sequences (AUUUA).<br>
</li>
<li>HuR binding promotes target RNA stabilization (and binding by the other RBPs to the ARE promotes destabilization).</li>
</ol>
<p>We will explore these predictions during the next couple classes.</p>
</section><section id="par-clip-data" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="par-clip-data">PAR-CLIP data</h2>
<p>Reminder that we will be using this resource: <a href="https://github.com/BIMSBbioinfo/RCAS_meta-analysis">rag-tag ENCODE</a></p>
<p>We are looking for an ELAVL1 PAR-CLIP corresponding to this SRA (short-read archive) ID: <strong>SRR248532</strong></p>
<div class="cell" data-output-location="column-fragment">
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">hur_regions</span> <span class="op">&lt;-</span> <span class="fu">read_regions</span><span class="op">(</span>con <span class="op">=</span> <span class="st">"https://raw.githubusercontent.com/BIMSBbioinfo/RCAS_meta-analysis/master/rbp-binding-sites/??.clusters.bed"</span>,</span>
<span>                            genome <span class="op">=</span> <span class="st">'hg19'</span>,</span>
<span>                            format <span class="op">=</span> <span class="st">'??'</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co"># looks like there are 111236 binding sites!!!</span></span>
<span><span class="co"># we are only going to keep those w/a score &gt; 1</span></span>
<span><span class="co"># this will reduce it down to ~20K sites</span></span>
<span><span class="co"># the score reflects the frequency of T-&gt;C conversions vs no conversion and non-T-&gt;C conversions per binding site</span></span>
<span><span class="va">hur_regions</span> <span class="op">&lt;-</span> <span class="va">hur_regions</span><span class="op">[</span><span class="va">hur_regions</span><span class="op">$</span><span class="va">score</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="va">hur_regions</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="annotate-par-clip-data" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="annotate-par-clip-data">Annotate PAR-CLIP data</h2>
<div class="cell" data-output-location="column-fragment">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># let's annotate</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>hur_annot <span class="ot">&lt;-</span> <span class="fu">annotate_regions</span>(</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    <span class="at">regions =</span> ??,</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">annotations =</span> annotations,</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    <span class="at">ignore.strand =</span> <span class="cn">FALSE</span>,</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>    <span class="at">quiet =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># we put in ~20K and got ~50K??</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a><span class="co"># redundant info due to transcript_id</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a><span class="co"># keep only columns we need</span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>myInfo <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"seqnames"</span>,<span class="st">"start"</span>,<span class="st">"end"</span>,<span class="st">"width"</span>,<span class="st">"strand"</span>,<span class="st">"annot.symbol"</span>,<span class="st">"annot.type"</span>)</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>hur_annot <span class="ot">&lt;-</span> hur_annot[,??] <span class="sc">%&gt;%</span></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>() </span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a><span class="co"># getting rid of the "hg19_genes_" string to simplify `annot.type`</span></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>hur_annot<span class="sc">$</span>annot.type <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"hg19_genes_"</span>,<span class="st">""</span>,hur_annot<span class="sc">$</span>annot.type)</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="hur-binding-region-preference" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="hur-binding-region-preference">HuR binding region preference</h2>
<p>It looks like HuR prefers binding to 3’ UTRs and introns. That is a bit of a surprise given the model above indicating 3’ UTR binding. Well let’s take a step back and frame our expectation using what we know about the genome.</p>
<p>For example, how much of the genome is coding?</p>
<p>In this case, how many basepairs are introns and 3’ UTRs in the genome?</p>
</section><section id="binding-region-length-biases" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="binding-region-length-biases">binding region length biases</h2>
<div class="cell" data-output-location="column-fragment">
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="co"># total intron length</span></span>
<span><span class="co"># sum(</span></span>
<span><span class="co">#   width(</span></span>
<span><span class="co">#     GenomicRanges::reduce(</span></span>
<span><span class="co">#       annotations[annotations$type %in% my_hg19_annots[3]]</span></span>
<span><span class="co">#       )</span></span>
<span><span class="co">#     )</span></span>
<span><span class="co">#   )</span></span>
<span></span>
<span><span class="va">mylengths</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html">vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">my_hg19_annots</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">mylengths</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu">width</span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/inter-range-methods.html">reduce</a></span><span class="op">(</span><span class="va">annotations</span><span class="op">[</span><span class="va">annotations</span><span class="op">$</span><span class="va">type</span> <span class="op"><a href="https://rdrr.io/r/base/match.html">%in%</a></span> <span class="va">my_hg19_annots</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  </span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/names.html">names</a></span><span class="op">(</span><span class="va">mylengths</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"hg19_genes_"</span>,<span class="st">""</span>,<span class="va">my_hg19_annots</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>  </span>
<span>  </span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/barplot.html">barplot</a></span><span class="op">(</span><span class="va">mylengths</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">4</span><span class="op">]</span>, las<span class="op">=</span><span class="fl">2</span>, main <span class="op">=</span> <span class="st">"total bases per category"</span>, log<span class="op">=</span><span class="st">"y"</span><span class="op">)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="control-for-clip-binding-sites" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="control-for-clip-binding-sites">Control for CLIP-binding sites</h2>
<p>We need a way to figure out a null model OR background expectation.</p>
<p>What if we were to take our HuR binding and randomize their position and then repeat the annotation on the randomized binding sites?</p>
<div class="cell" data-output-location="column-fragment">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># randomize regions based on par-clip regions</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>random_sites <span class="ot">&lt;-</span> ??(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">regions =</span> hur_regions,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">allow.overlaps =</span> <span class="cn">TRUE</span>,</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">per.chromosome =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># annotate  randomize regions</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>random_sites_annot <span class="ot">&lt;-</span> <span class="fu">annotate_regions</span>(</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">regions =</span> ??,</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">annotations =</span> annotations,</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">ignore.strand =</span> <span class="cn">FALSE</span>,</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">quiet =</span> <span class="cn">FALSE</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.frame</span>()</span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a><span class="co"># select cols, keep unique</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>random_sites_annot <span class="ot">&lt;-</span> random_sites_annot[,myInfo] <span class="sc">%&gt;%</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">unique</span>() </span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a><span class="co"># clean names</span></span>
<span id="cb7-20"><a href="#cb7-20" aria-hidden="true" tabindex="-1"></a>random_sites_annot<span class="sc">$</span>annot.type <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">"hg19_genes_"</span>,<span class="st">""</span>,random_sites_annot<span class="sc">$</span>annot.type)</span>
<span id="cb7-21"><a href="#cb7-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-22"><a href="#cb7-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-23"><a href="#cb7-23" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(random_sites_annot<span class="sc">$</span>annot.type)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="compare-hur-vs-random" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="compare-hur-vs-random">Compare HuR vs Random</h2>
<div class="cell" data-output-location="slide">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a new tibble with the annotation counts for hur (observed) and random (expected) binding sties</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>site_dist <span class="ot">&lt;-</span> <span class="fu">bind_cols</span>(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">region=</span><span class="fu">names</span>(<span class="fu">table</span>(hur_annot<span class="sc">$</span>annot.type)),</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">observed=</span><span class="fu">table</span>(??<span class="sc">$</span>annot.type),</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">expected=</span><span class="fu">table</span>(??<span class="sc">$</span>annot.type)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>          ) <span class="sc">%&gt;%</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">enrichment =</span> observed<span class="sc">/</span>expected) <span class="co"># calculate enrichment</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>site_dist_long <span class="ot">&lt;-</span> <span class="fu">pivot_longer</span>(site_dist,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                               <span class="at">cols =</span> <span class="fu">c</span>(??,??,??)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>                               )</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(site_dist_long) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"region"</span>,<span class="st">"type"</span>,<span class="st">"value"</span>)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># plot counts for parclip vs random barplot</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a><span class="co"># ggplot(site_dist_long %&gt;% filter(type!="enrichment"),</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co">#        aes(y = value, x = region, fill = type)) +</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="co">#   geom_bar(stat="identity", position=position_dodge()) +</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="co">#   scale_fill_manual(values=c("black","red")) +</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a><span class="co">#   theme_cowplot()</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="co"># plot enrichment barplot</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(site_dist_long <span class="sc">%&gt;%</span> <span class="fu">filter</span>(type<span class="sc">==</span><span class="st">"??"</span>),</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">y =</span> ??,</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> ??)) <span class="sc">+</span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"Observed vs Expected"</span>) <span class="sc">+</span></span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_cowplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="minute-break" class="level2"><h2 class="anchored" data-anchor-id="minute-break">5 MINUTE BREAK</h2>
</section><section id="what-sequence-does-hur-bind-to" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="what-sequence-does-hur-bind-to">What sequence does HuR bind to?</h2>
<p>Is it just <em>AUUUA</em>?</p>
<p><strong>Different transcript regions have different nucleotide composition.</strong></p>
<ul>
<li><p>5’ UTRs are more GC-rich</p></li>
<li><p>3’ UTRs are more AU-rich</p></li>
</ul>
<p><img src="../img/block-rna/gc.jpg" class="img-fluid"></p>
<p>We need to account for these differences when asking what the binding preferences are for sites in a given transcript region.</p>
</section><section id="counting-k-mers" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="counting-k-mers">Counting k-mers</h2>
<p>Steps to determine k-mer composition (we use 6mers) for any set of intervals</p>
<p>We’ll do it for both HuR binding sites and then compare it to background seqs.</p>
<ol type="1">
<li>Create a <code>Granges</code> object for a given annotation category.<br>
</li>
<li>Remove duplicated intervals (from diff transcript ids) with <code>reduce</code>.<br>
</li>
<li>Retrieve seqeunces using <code>getSeqs</code>
</li>
<li>Create a dataframe containing the count and frequency of each 6mer.</li>
</ol></section><section id="calculate-6mers-in-hur-sites" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="calculate-6mers-in-hur-sites">Calculate 6mers in HuR sites</h2>
<p>Since HuR preferentially binds to 3’ UTRs, that is the region we will focus on.</p>
<div class="cell" data-output-location="slide">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a Grange for 3' UTRs </span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>hur_3utr <span class="ot">&lt;-</span> <span class="fu">makeGRangesFromDataFrame</span>(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">df =</span> hur_annot <span class="sc">%&gt;%</span> <span class="fu">filter</span>(??<span class="sc">==</span><span class="st">"??"</span>),</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">ignore.strand =</span> F,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">seqnames.field =</span> <span class="st">"seqnames"</span>,</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">start.field =</span> <span class="st">"start"</span>,</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">end.field =</span> <span class="st">"end"</span>,</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">strand.field =</span> <span class="st">"strand"</span>,</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">keep.extra.columns =</span> T</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co"># get  sequencees for those coordinates</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>hur_3utr_seqs <span class="ot">&lt;-</span> <span class="fu">getSeq</span>(BSgenome.Hsapiens.UCSC.hg19, ??)</span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a><span class="co"># count  all 6 mer instances per sequence, add all instances, and turn into a dataframe with column name counts</span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a>hur_utr3_6mer <span class="ot">&lt;-</span> ??(</span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> ??,</span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">6</span>, <span class="co"># k</span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">as.prob =</span> F,</span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">simplify.as=</span><span class="st">"matrix"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colSums</span>(.) <span class="sc">%&gt;%</span></span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(hur_utr3_6mer) <span class="ot">&lt;-</span> <span class="st">"hur_utr_count"</span></span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a><span class="co"># calc freq </span></span>
<span id="cb9-27"><a href="#cb9-27" aria-hidden="true" tabindex="-1"></a>hur_utr3_6mer<span class="sc">$</span>hur_utr_freq <span class="ot">&lt;-</span> hur_utr3_6mer<span class="sc">$</span>hur_utr_count<span class="sc">/</span><span class="fu">sum</span>(hur_utr3_6mer<span class="sc">$</span>hur_utr_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="calculate-6mers-in-3utrs" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="calculate-6mers-in-3utrs">Calculate 6mers in 3utrs</h2>
<p>Next, we will calculate 6mer frequencies in 3’ UTRs. This will serve as a null model or background that we can compare with the HuR binding site 6mers.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create a Grange for 3' UTRs </span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>threeUTR <span class="ot">&lt;-</span> GenomicRanges<span class="sc">::</span><span class="fu">reduce</span>(annotations[annotations<span class="sc">$</span>type <span class="sc">%in%</span> my_hg19_annots[??]])</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># get sequences for those coordinates</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>threeUTR_seqs <span class="ot">&lt;-</span> <span class="fu">getSeq</span>(BSgenome.Hsapiens.UCSC.hg19, threeUTR)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># count  all 6 mer instances per sequence, add all instances, and turn into a dataframe with column name counts</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>utr3_6mer <span class="ot">&lt;-</span> ??(</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> ??,</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">width =</span> <span class="dv">6</span>,</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">as.prob =</span> F,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">simplify.as=</span><span class="st">"matrix"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colSums</span>(.) <span class="sc">%&gt;%</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(utr3_6mer) <span class="ot">&lt;-</span> <span class="st">"utr_count"</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a><span class="co"># calc freq</span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>utr3_6mer<span class="sc">$</span>utr_freq <span class="ot">&lt;-</span> utr3_6mer<span class="sc">$</span>utr_count<span class="sc">/</span><span class="fu">sum</span>(utr3_6mer<span class="sc">$</span>utr_count)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sequences-enriched-in-hur-sites-vs-3utr" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="sequences-enriched-in-hur-sites-vs-3utr">Sequences enriched in hur sites vs 3utr</h2>
<div class="cell" data-output-location="slide">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check if rownames are identical and stop if they are  not</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">stopifnot</span>(<span class="fu">identical</span>(<span class="fu">rownames</span>(hur_utr3_6mer),<span class="fu">rownames</span>(utr3_6mer)))</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># bind_cols</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>utr3_df <span class="ot">&lt;-</span> <span class="fu">bind_cols</span>(hur_utr3_6mer,</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>                     utr3_6mer)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Calc enrichment</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>utr3_df<span class="sc">$</span>hur_enrich <span class="ot">&lt;-</span> utr3_df<span class="sc">$</span>hur_utr_freq<span class="sc">/</span>utr3_df<span class="sc">$</span>utr_freq</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a><span class="co"># plot coloring and labeling all &gt; 8 fold enriched</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(<span class="at">data =</span> utr3_df,</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>       <span class="fu">aes</span>(<span class="at">y =</span> ??, <span class="co"># hur</span></span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>           <span class="at">x =</span> ?? <span class="co"># utrs</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>             )</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>       ) <span class="sc">+</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">color =</span> <span class="fu">ifelse</span>(utr3_df<span class="sc">$</span>hur_enrich <span class="sc">&gt;</span> <span class="dv">8</span>, <span class="st">"red"</span>, <span class="st">"black"</span>)) <span class="sc">+</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">"kmers: HuR 3'UTR sites"</span>) <span class="sc">+</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">"kmers: 3'UTR"</span>) <span class="sc">+</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_abline</span>(<span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">slope =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text_repel</span>(<span class="fu">aes</span>(<span class="at">label=</span><span class="fu">ifelse</span>(hur_enrich <span class="sc">&gt;</span> <span class="dv">8</span>,</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true" tabindex="-1"></a>                                   <span class="fu">rownames</span>(utr3_df),</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true" tabindex="-1"></a>                                   <span class="st">""</span>))) <span class="sc">+</span></span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_cowplot</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section><section id="sequences-enriched-in-hur-sites-vs-3utr-1" class="level2 smaller"><h2 class="smaller anchored" data-anchor-id="sequences-enriched-in-hur-sites-vs-3utr-1">Sequences enriched in hur sites vs 3utr</h2>
<p>Not just AU-rich…</p>
<div class="cell">
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r code-with-copy"><code class="sourceCode R"><span><span class="va">utr3_df</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rownames_to_column</span><span class="op">(</span>var <span class="op">=</span> <span class="st">"6mer"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">-</span><span class="va">hur_enrich</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">gt</span><span class="op">(</span><span class="op">)</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>


</section></main><!-- /main --><script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/rnabioco\.github\.io\/molb-7950\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer"><div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      &nbsp;
    <div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/rnabioco/molb-7950/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li><li><a href="https://github.com/rnabioco/molb-7950/edit/main/exercises/ex-27.qmd" class="toc-action"><i class="bi empty"></i>Edit this page</a></li></ul></div></div>
    <div class="nav-footer-right">
<p>This page is built with <a href="https://quarto.org/">Quarto</a>.</p>
</div>
  </div>
</footer>


</body></html>