---
title: "Single-cell RNA-seq Problem Set I"
author: "Yor name here"
---

Grade (out of 20):

For this problem set we will be reanalyzing some public single cell RNA-seq data ([publication](https://doi.org/10.1038/s41591-018-0233-1)). The dataset contains PBMCs from a patient with Acute Myeloid Leukemia (AML). The data we will be analyzing consists of two samples, one taken 2 days (day2) after treatment with a chemotherapeutic (Venetoclax and Azacitidine) or one taken prior to treatment (day0). 

The datasets have been processed with `alevin`, and the output is here `data/aml`.


Q1A 2 points) Load the matrices into R separately using `tximport`. Call cells with `emptydrops()` and make a plot with the total UMI count per barcode versus barcode rank plots for each sample. Color the barcodes with whether they were called cells or not. 

```{r}
library(tximport)
library(tidyverse)
library(DropletUtils)

# Load in the data for both days
d0 <- tximport("data/aml/alevin/day0/alevin/quants_mat.gz", type = "alevin")
d2 <- tximport("data/aml/alevin/day2/alevin/quants_mat.gz", type = "alevin")

# TODO
# Run empty drops on both datasets using the counts slots in both d0 and d2 objects
# Make a Knee plot with colored cells like in the "Quality Control: how to
# distinguish an empty droplet from a cell-containing droplet?" section of the
# lecture
```

1B 2 points) How many cells are called for each sample?

```{r}
# Hint - find the number of cells in d0$count and d2$count thathave an FDR < 0.05
```

1C 2 points) Filter the matrices to only retain cells. Use a cutoff of `FDR < 0.05`.

```{r}
# code here
```

1D 2 points) What proportion of the UMIs are within cell containing droplets for each sample?

```{r}
# code here
```


1E 2 points) Based on these plots and metrics what do you think of the quality of these samples?

    Short answer here...


1F 2 points) The day2 sample and to some extent the day0 sample show a third knee in the plot at high UMI counts. Speculate what could cause this additional knee in the distribution. 

    Short answer here...

Q2A 2 points) Next we will generate a Seurat object for theses sample. In class we discussed loading a single matrix into a Seurat object. In this exercise we will generate a seurat object with both samples.

Use the following pseudocode to combine the two matrices into 1. We will rename the cell barcodes to ensure that they are unique across the two samples. 

```{r}
# assume that d0_mat means the day 0 matrix only containing cells
# assume that d2_mat means the day 2 matrix only containing cells
colnames(d0_mat) <- paste0("day0_", colnames(d0_mat))
colnames(d2_mat) <- paste0("day2_", colnames(d2_mat))
head(colnames(d0_mat))
head(colnames(d2_mat))

d0_d2_mat <- cbind(d0_mat, d2_mat)
```

Use `CreateSeuratObject` to generate a Seurat object from this new matrix. Set the `Ident` of the Seurat object to indicate the sample each cell came from (day0 or day2). (Hint: See the `names.field` and `names.delim` argument of `CreateSeuratObect`, or use `AddMetaData`). Verify the `ident` is set correctly by printing the idents for the first 5 cells (e.g `Idents(your_seurat_object)`).


```{r}
library(Seurat)

so <- CreateSeuratObject(...) # Fill in the ...
```

Q2B 2 points)

Next calculate the % of UMIs that are derived from mitochondrial genes and plot this percentage against the # of UMIs per cell. 

```{r}
# Remeber the pattern for human mitochondrial genes is "^MT-"
so <- PercentageFeatureSet(...) # Fill in ...
```

Q2C 2 points) Plot the # of UMIs and # of genes detected as violin plots. You may need to reduce the size of the points to more clearly see the distribution. 

```{r}
# code here
```

Q2D 2-points) Based on these plots, filter your Seurat object to exclude low quality cells. Report the # of cells pre-filtering and post-filtering for each sample. 


```{r}
so_filt <- subset(...) # Fill in code here
```

