---
title: "MOLB 7950 â€” Final Projects"
---

## Overview

* Final projects can involve groups of 1-3 people.

* Projects are choose your own adventure:

    1. You can start with a template. Each of these have a set of questions,
       that you can build off of:
      
        - `dna-block/` explores an MNase-seq data set in
        S. cerevisiae.  Sub-nucleosomal fragments provide a DNA-based signal to
        understand chromatin transactions that lead to transcription.
        
        - `rna-block/published-data/` is the material Kent presented, examining the
        role of A-to-I editing using RNA-seq.
      
        - `rna-block/rna-localization/`: is an exploration of neuronal RNA localization
        by RNA-seq.

    1. You could find a data set on [NCBI GEO](https://www.ncbi.nlm.nih.gov/geo/)
       of interest (e.g., relevant to your thesis work), and work it up with
       salmon, DEseq, and exploratory analysis.
       
    1. You can start with your own sequencing data (bulk/single-cell RNA seq,
       DNA sequencing).

## Deliverables     

* An Rmarkdown document with code, plots, interpretations, and next steps.

* If you work in a group, list the members of the group at the top of the document,
  and make it clear which parts are your work by adding your initials to code chunks.

* Short presentations (8-10 minutes) by the groups the week of Nov 28. Presentations
  should include 1-2 slides of background, a hypothesis for the approach, code
  output (table or graph) that addresses the hypothesis, and one or more tests
  of the statistical significance of the observation.

## Grading and rubric

The final project will be worth 20% of your grade and we will use the grading
scheme outlined in the [grading rubric](syllabus.qmd#problem-set-rubric).

Each individual in a group will be evaluated separately, so contributions must
be clearly marked in the document, using e.g. using chunk labels:

``` r
#| label: plotting-code-by-Jay
ggplot(mtcars, aes(hp, mpg)) + geom_point()
```